// @shopify/hydrogen-react@2026.4.2/dist/umd/hydrogen-react.dev.js
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports, require("react")) : typeof define === "function" && define.amd ? define(["exports", "react"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.hydrogenreact = {}, global.React));
})(this, function(exports, react) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region \0rolldown/runtime.js
	var __create = Object.create;
	var __defProp = Object.defineProperty;
	var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
	var __getOwnPropNames = Object.getOwnPropertyNames;
	var __getProtoOf = Object.getPrototypeOf;
	var __hasOwnProp = Object.prototype.hasOwnProperty;
	var __copyProps = (to, from, except, desc) => {
		if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
				get: ((k) => from[k]).bind(null, key),
				enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
			});
		}
		return to;
	};
	var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
		value: mod,
		enumerable: true
	}) : target, mod));
	//#endregion
	react = __toESM(react);
	//#region ../../node_modules/.pnpm/@xstate+fsm@2.0.0/node_modules/@xstate/fsm/es/index.js
	/*! *****************************************************************************
	Copyright (c) Microsoft Corporation.
	
	Permission to use, copy, modify, and/or distribute this software for any
	purpose with or without fee is hereby granted.
	
	THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
	REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
	AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
	INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
	LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
	OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
	PERFORMANCE OF THIS SOFTWARE.
	***************************************************************************** */
	function t(t, n) {
		var e = "function" == typeof Symbol && t[Symbol.iterator];
		if (!e) return t;
		var r, i, o = e.call(t), a = [];
		try {
			for (; (void 0 === n || n-- > 0) && !(r = o.next()).done;) a.push(r.value);
		} catch (t) {
			i = { error: t };
		} finally {
			try {
				r && !r.done && (e = o.return) && e.call(o);
			} finally {
				if (i) throw i.error;
			}
		}
		return a;
	}
	var n;
	(function(t) {
		t[t.NotStarted = 0] = "NotStarted", t[t.Running = 1] = "Running", t[t.Stopped = 2] = "Stopped";
	})(n || (n = {}));
	var e = { type: "xstate.init" };
	function r(t) {
		return void 0 === t ? [] : [].concat(t);
	}
	function i(t) {
		return {
			type: "xstate.assign",
			assignment: t
		};
	}
	function o(t, n) {
		return "string" == typeof (t = "string" == typeof t && n && n[t] ? n[t] : t) ? { type: t } : "function" == typeof t ? {
			type: t.name,
			exec: t
		} : t;
	}
	function a(t) {
		return function(n) {
			return t === n;
		};
	}
	function u(t) {
		return "string" == typeof t ? { type: t } : t;
	}
	function c(t, n) {
		return {
			value: t,
			context: n,
			actions: [],
			changed: !1,
			matches: a(t)
		};
	}
	function f(t, n, e) {
		var r = n, i = !1;
		return [
			t.filter((function(t) {
				if ("xstate.assign" === t.type) {
					i = !0;
					var n = Object.assign({}, r);
					return "function" == typeof t.assignment ? n = t.assignment(r, e) : Object.keys(t.assignment).forEach((function(i) {
						n[i] = "function" == typeof t.assignment[i] ? t.assignment[i](r, e) : t.assignment[i];
					})), r = n, !1;
				}
				return !0;
			})),
			r,
			i
		];
	}
	function s(n, i) {
		void 0 === i && (i = {});
		var s = t(f(r(n.states[n.initial].entry).map((function(t) {
			return o(t, i.actions);
		})), n.context, e), 2), l = s[0], v = s[1], y = {
			config: n,
			_options: i,
			initialState: {
				value: n.initial,
				actions: l,
				context: v,
				matches: a(n.initial)
			},
			transition: function(e, i) {
				var s, l, v = "string" == typeof e ? {
					value: e,
					context: n.context
				} : e, p = v.value, g = v.context, d = u(i), x = n.states[p];
				if (x.on) {
					var m = r(x.on[d.type]);
					try {
						for (var h = function(t) {
							var n = "function" == typeof Symbol && Symbol.iterator, e = n && t[n], r = 0;
							if (e) return e.call(t);
							if (t && "number" == typeof t.length) return { next: function() {
								return t && r >= t.length && (t = void 0), {
									value: t && t[r++],
									done: !t
								};
							} };
							throw new TypeError(n ? "Object is not iterable." : "Symbol.iterator is not defined.");
						}(m), b = h.next(); !b.done; b = h.next()) {
							var S = b.value;
							if (void 0 === S) return c(p, g);
							var w = "string" == typeof S ? { target: S } : S, j = w.target, E = w.actions, R = void 0 === E ? [] : E, N = w.cond, O = void 0 === N ? function() {
								return !0;
							} : N, _ = void 0 === j, k = null != j ? j : p, T = n.states[k];
							if (O(g, d)) {
								var q = t(f((_ ? r(R) : [].concat(x.exit, R, T.entry).filter((function(t) {
									return t;
								}))).map((function(t) {
									return o(t, y._options.actions);
								})), g, d), 3), z = q[0], A = q[1], B = q[2], C = null != j ? j : p;
								return {
									value: C,
									context: A,
									actions: z,
									changed: j !== p || z.length > 0 || B,
									matches: a(C)
								};
							}
						}
					} catch (t) {
						s = { error: t };
					} finally {
						try {
							b && !b.done && (l = h.return) && l.call(h);
						} finally {
							if (s) throw s.error;
						}
					}
				}
				return c(p, g);
			}
		};
		return y;
	}
	var l$1 = function(t, n) {
		return t.actions.forEach((function(e) {
			var r = e.exec;
			return r && r(t.context, n);
		}));
	};
	function v(t) {
		var r = t.initialState, i = n.NotStarted, o = /* @__PURE__ */ new Set(), c = {
			_machine: t,
			send: function(e) {
				i === n.Running && (r = t.transition(r, e), l$1(r, u(e)), o.forEach((function(t) {
					return t(r);
				})));
			},
			subscribe: function(t) {
				return o.add(t), t(r), { unsubscribe: function() {
					return o.delete(t);
				} };
			},
			start: function(o) {
				if (o) {
					var u = "object" == typeof o ? o : {
						context: t.config.context,
						value: o
					};
					r = {
						value: u.value,
						actions: [],
						context: u.context,
						matches: a(u.value)
					};
				} else r = t.initialState;
				return i = n.Running, l$1(r, e), c;
			},
			stop: function() {
				return i = n.Stopped, o.clear(), c;
			},
			get state() {
				return r;
			},
			get status() {
				return i;
			}
		};
		return c;
	}
	//#endregion
	//#region src/useMachine.ts
	var useIsomorphicLayoutEffect = typeof window !== "undefined" ? react.useLayoutEffect : react.useEffect;
	function useConstant(fn) {
		const ref = (0, react.useRef)();
		if (!ref.current) ref.current = { v: fn() };
		return ref.current.v;
	}
	function getServiceState(service) {
		let currentValue;
		service.subscribe((state) => currentValue = state).unsubscribe();
		return currentValue;
	}
	function useMachine(stateMachine, options) {
		const persistedStateRef = (0, react.useRef)();
		const [service, queue] = useConstant(() => {
			const eventQueue = [];
			const svc = v(s(stateMachine.config, options ? options : stateMachine._options));
			const originalSend = svc.send;
			svc.send = (event) => {
				if (svc.status === n.NotStarted) {
					eventQueue.push(event);
					return;
				}
				originalSend(event);
				persistedStateRef.current = svc.state;
			};
			return [svc, eventQueue];
		});
		useIsomorphicLayoutEffect(() => {
			if (options) service._machine._options = options;
		});
		const getSnapshot = (0, react.useCallback)(() => getServiceState(service), [service]);
		const storeSnapshot = (0, react.useSyncExternalStore)((0, react.useCallback)((handleStoreChange) => {
			const { unsubscribe } = service.subscribe(handleStoreChange);
			return unsubscribe;
		}, [service]), getSnapshot, getSnapshot);
		(0, react.useEffect)(() => {
			service.start(persistedStateRef.current);
			queue.forEach(service.send);
			persistedStateRef.current = service.state;
			return () => {
				service.stop();
			};
		}, []);
		return [
			storeSnapshot,
			service.send,
			service
		];
	}
	//#endregion
	//#region src/flatten-connection.ts
	/**
	* The `flattenConnection` utility transforms a connection object from the Storefront API (for example, [Product-related connections](https://shopify.dev/api/storefront/reference/products/product)) into a flat array of nodes.
	* The utility works with either `nodes` or `edges.node`.
	*
	* If `connection` is null or undefined, will return an empty array instead in production. In development, an error will be thrown.
	* @publicDocs
	*/
	function flattenConnection(connection) {
		if (!connection) {
			const noConnectionErr = `flattenConnection(): needs a 'connection' to flatten, but received '${connection ?? ""}' instead.`;
			throw new Error(noConnectionErr);
		}
		if ("nodes" in connection) return connection.nodes;
		if ("edges" in connection && Array.isArray(connection.edges)) return connection.edges.map((edge) => {
			if (!edge?.node) throw new Error("flattenConnection(): Connection edges must contain nodes");
			return edge.node;
		});
		console.warn(`flattenConnection(): The connection did not contain either "nodes" or "edges.node". Returning an empty array.`);
		return [];
	}
	//#endregion
	//#region src/cart-queries.ts
	function getInContextVariables(includeVisitorConsent) {
		const base = `$country: CountryCode = ZZ
    $language: LanguageCode`;
		return includeVisitorConsent ? `${base}
    $visitorConsent: VisitorConsent` : base;
	}
	function getInContextDirective(includeVisitorConsent) {
		return includeVisitorConsent ? `@inContext(
    country: $country
    language: $language
    visitorConsent: $visitorConsent
  )` : `@inContext(
    country: $country
    language: $language
  )`;
	}
	var CartLineAdd = (cartFragment, options = {}) => `
  mutation CartLineAdd(
    $cartId: ID!
    $lines: [CartLineInput!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesAdd(cartId: $cartId, lines: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartCreate = (cartFragment, options = {}) => `
  mutation CartCreate(
    $input: CartInput!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartCreate(input: $input) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartLineRemove = (cartFragment, options = {}) => `
  mutation CartLineRemove(
    $cartId: ID!
    $lines: [ID!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesRemove(cartId: $cartId, lineIds: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartLineUpdate = (cartFragment, options = {}) => `
  mutation CartLineUpdate(
    $cartId: ID!
    $lines: [CartLineUpdateInput!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesUpdate(cartId: $cartId, lines: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartNoteUpdate = (cartFragment, options = {}) => `
  mutation CartNoteUpdate(
    $cartId: ID!
    $note: String!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartNoteUpdate(cartId: $cartId, note: $note) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartBuyerIdentityUpdate = (cartFragment, options = {}) => `
  mutation CartBuyerIdentityUpdate(
    $cartId: ID!
    $buyerIdentity: CartBuyerIdentityInput!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartBuyerIdentityUpdate(cartId: $cartId, buyerIdentity: $buyerIdentity) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartAttributesUpdate = (cartFragment, options = {}) => `
  mutation CartAttributesUpdate(
    $attributes: [AttributeInput!]!
    $cartId: ID!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartAttributesUpdate(attributes: $attributes, cartId: $cartId) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartDiscountCodesUpdate = (cartFragment, options = {}) => `
  mutation CartDiscountCodesUpdate(
    $cartId: ID!
    $discountCodes: [String!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartDiscountCodesUpdate(cartId: $cartId, discountCodes: $discountCodes) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
	var CartQuery = (cartFragment, options = {}) => `
  query CartQuery(
    $id: ID!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cart(id: $id) {
      ...CartFragment
    }
  }

  ${cartFragment}
`;
	var defaultCartFragment = `
  fragment CartFragment on Cart {
    id
    checkoutUrl
    totalQuantity
    buyerIdentity {
      countryCode
      customer {
        id
        email
        firstName
        lastName
        displayName
      }
      email
      phone
    }
    lines(first: $numCartLines) {
      edges {
        node {
          id
          quantity
          attributes {
            key
            value
          }
          cost {
            totalAmount {
              amount
              currencyCode
            }
            compareAtAmountPerQuantity {
              amount
              currencyCode
            }
          }
          merchandise {
            ... on ProductVariant {
              id
              availableForSale
              compareAtPrice {
                ...MoneyFragment
              }
              price {
                ...MoneyFragment
              }
              requiresShipping
              title
              image {
                ...ImageFragment
              }
              product {
                handle
                title
                id
              }
              selectedOptions {
                name
                value
              }
            }
          }
        }
      }
    }
    cost {
      subtotalAmount {
        ...MoneyFragment
      }
      totalAmount {
        ...MoneyFragment
      }
      totalDutyAmount {
        ...MoneyFragment
      }
      totalTaxAmount {
        ...MoneyFragment
      }
    }
    note
    attributes {
      key
      value
    }
    discountCodes {
      code
      applicable
    }
  }

  fragment MoneyFragment on MoneyV2 {
    currencyCode
    amount
  }
  fragment ImageFragment on Image {
    id
    url
    altText
    width
    height
  }
`;
	//#endregion
	//#region src/storefront-api-constants.ts
	var SFAPI_VERSION = "2026-04";
	//#endregion
	//#region src/storefront-client.ts
	var MOCK_SHOP_DOMAIN = "mock.shop";
	var isMockShop = (domain) => domain.includes(MOCK_SHOP_DOMAIN);
	/**
	* The `createStorefrontClient()` function creates helpers that enable you to quickly query the Shopify Storefront API.
	*
	* When used on the server, it is recommended to use the `privateStorefrontToken` prop. When used on the client, it is recommended to use the `publicStorefrontToken` prop.
	* @publicDocs
	*/
	function createStorefrontClient({ storeDomain, privateStorefrontToken, publicStorefrontToken, storefrontApiVersion = SFAPI_VERSION, contentType }) {
		if (!storeDomain) {
			storeDomain = MOCK_SHOP_DOMAIN;
			warnOnce(`storeDomain missing, defaulting to ${MOCK_SHOP_DOMAIN}`, "info");
		}
		if (storefrontApiVersion !== "2026-04") warnOnce(`The Storefront API version that you're using is different than the version this build of Hydrogen React is targeting.\nYou may run into unexpected errors if these versions don't match. Received version: "${storefrontApiVersion}"; expected version "${SFAPI_VERSION}"`);
		if (!privateStorefrontToken && !globalThis.document && !isMockShop(storeDomain)) warnOnce("Using a private storefront token is recommended for server environments.\nRefer to the authentication https://shopify.dev/api/storefront#authentication documentation for more details.");
		if (privateStorefrontToken && globalThis.document) warnOnce("You are attempting to use a private token in an environment where it can be easily accessed by anyone.\nThis is a security risk; please use the public token and the `publicStorefrontToken` prop");
		const getShopifyDomain = (overrideProps) => {
			const domain = overrideProps?.storeDomain ?? storeDomain;
			return domain.includes("://") ? domain : `https://${domain}`;
		};
		return {
			getShopifyDomain,
			getStorefrontApiUrl(overrideProps) {
				const domain = getShopifyDomain(overrideProps);
				return `${domain + (domain.endsWith("/") ? "api" : "/api")}/${overrideProps?.storefrontApiVersion ?? storefrontApiVersion}/graphql.json`;
			},
			getPrivateTokenHeaders(overrideProps) {
				if (!privateStorefrontToken && !overrideProps?.privateStorefrontToken && !isMockShop(storeDomain)) throw new Error(H2_PREFIX_ERROR + "You did not pass in a `privateStorefrontToken` while using `createStorefrontClient()` or `getPrivateTokenHeaders()`");
				if (!overrideProps?.buyerIp) warnOnce("It is recommended to pass in the `buyerIp` property which improves analytics and data in the admin.");
				return {
					"content-type": (overrideProps?.contentType ?? contentType) === "graphql" ? "application/graphql" : "application/json",
					"X-SDK-Variant": "hydrogen-react",
					"X-SDK-Variant-Source": "react",
					"X-SDK-Version": storefrontApiVersion,
					"Shopify-Storefront-Private-Token": overrideProps?.privateStorefrontToken ?? privateStorefrontToken ?? "",
					...overrideProps?.buyerIp ? { "Shopify-Storefront-Buyer-IP": overrideProps.buyerIp } : {}
				};
			},
			getPublicTokenHeaders(overrideProps) {
				if (!publicStorefrontToken && !overrideProps?.publicStorefrontToken && !isMockShop(storeDomain)) throw new Error(H2_PREFIX_ERROR + "You did not pass in a `publicStorefrontToken` while using `createStorefrontClient()` or `getPublicTokenHeaders()`");
				return getPublicTokenHeadersRaw(overrideProps?.contentType ?? contentType ?? "json", storefrontApiVersion, overrideProps?.publicStorefrontToken ?? publicStorefrontToken ?? "");
			}
		};
	}
	function getPublicTokenHeadersRaw(contentType, storefrontApiVersion, accessToken) {
		return {
			"content-type": contentType === "graphql" ? "application/graphql" : "application/json",
			"X-SDK-Variant": "hydrogen-react",
			"X-SDK-Variant-Source": "react",
			"X-SDK-Version": storefrontApiVersion,
			"X-Shopify-Storefront-Access-Token": accessToken
		};
	}
	var warnings = /* @__PURE__ */ new Set();
	var H2_PREFIX_ERROR = "[h2:error:createStorefrontClient] ";
	var warnOnce = (string, type = "warn") => {
		if (!warnings.has(string)) {
			console[type](`[h2:${type}:createStorefrontClient] ` + string);
			warnings.add(string);
		}
	};
	//#endregion
	//#region src/ShopifyProvider.tsx
	var defaultShopifyContext = {
		storeDomain: "test",
		storefrontToken: "abc123",
		storefrontApiVersion: SFAPI_VERSION,
		countryIsoCode: "US",
		languageIsoCode: "EN",
		getStorefrontApiUrl() {
			return "";
		},
		getPublicTokenHeaders() {
			return {};
		},
		getShopifyDomain() {
			return "";
		}
	};
	var ShopifyContext = (0, react.createContext)(defaultShopifyContext);
	/**
	* Hydrogen server sets this server timing key when the SFAPI proxy is enabled.
	* Read it automatically in the browser for apps using frontend cart in full-stack Hydrogen,
	* but don't export this utility yet since we don't want to make this a public convention.
	*/
	function isSfapiProxyEnabled() {
		if (typeof window === "undefined") return false;
		try {
			return !!(window.performance?.getEntriesByType?.("navigation")[0])?.serverTiming?.some((entry) => entry.name === "_sfapi_proxy");
		} catch (e) {
			return false;
		}
	}
	/**
	* The `<ShopifyProvider/>` component enables use of the `useShop()` hook. The component should wrap your app.
	* @publicDocs
	*/
	function ShopifyProvider({ children, ...shopifyConfig }) {
		if (!shopifyConfig.countryIsoCode || !shopifyConfig.languageIsoCode || !shopifyConfig.storeDomain || !shopifyConfig.storefrontToken || !shopifyConfig.storefrontApiVersion) throw new Error(`Please provide the necessary props to '<ShopifyProvider/>'`);
		if (shopifyConfig.storefrontApiVersion !== "2026-04") console.warn(`<ShopifyProvider/>: This version of Hydrogen React is built for Shopify's Storefront API version ${SFAPI_VERSION}, but it looks like you're using version ${shopifyConfig.storefrontApiVersion}. There may be issues or bugs if you use a mismatched version of Hydrogen React and the Storefront API.`);
		const finalConfig = (0, react.useMemo)(() => {
			const sameDomainForStorefrontApi = shopifyConfig.sameDomainForStorefrontApi ?? isSfapiProxyEnabled();
			function getShopifyDomain(overrideProps) {
				const domain = overrideProps?.storeDomain ?? shopifyConfig.storeDomain;
				return domain.includes("://") ? domain : `https://${domain}`;
			}
			return {
				...shopifyConfig,
				sameDomainForStorefrontApi,
				getPublicTokenHeaders(overrideProps) {
					return getPublicTokenHeadersRaw(overrideProps.contentType, shopifyConfig.storefrontApiVersion, overrideProps.storefrontToken ?? shopifyConfig.storefrontToken);
				},
				getShopifyDomain,
				getStorefrontApiUrl(overrideProps) {
					const finalDomainUrl = sameDomainForStorefrontApi && typeof window !== "undefined" ? window.location.origin : getShopifyDomain({ storeDomain: overrideProps?.storeDomain ?? shopifyConfig.storeDomain });
					return `${finalDomainUrl}${finalDomainUrl.endsWith("/") ? "" : "/"}api/${overrideProps?.storefrontApiVersion ?? shopifyConfig.storefrontApiVersion}/graphql.json`;
				}
			};
		}, [shopifyConfig]);
		return /* @__PURE__ */ React.createElement(ShopifyContext.Provider, { value: finalConfig }, children);
	}
	/**
	* Provides access to the `shopifyConfig` prop of `<ShopifyProvider/>`. Must be a descendent of `<ShopifyProvider/>`.
	* @publicDocs
	*/
	function useShop() {
		const shopContext = (0, react.useContext)(ShopifyContext);
		if (!shopContext) throw new Error(`'useShop()' must be a descendent of <ShopifyProvider/>`);
		return shopContext;
	}
	//#endregion
	//#region src/cart-constants.ts
	var CART_ID_STORAGE_KEY = "shopifyCartId";
	var SHOPIFY_STOREFRONT_ID_HEADER = "Shopify-Storefront-Id";
	var SHOPIFY_STOREFRONT_Y_HEADER = "Shopify-Storefront-Y";
	var SHOPIFY_STOREFRONT_S_HEADER = "Shopify-Storefront-S";
	var SHOPIFY_Y = "_shopify_y";
	var SHOPIFY_S = "_shopify_s";
	//#endregion
	//#region src/tracking-utils.ts
	/** Storefront API header for VisitToken */
	var SHOPIFY_VISIT_TOKEN_HEADER = "X-Shopify-VisitToken";
	/** Storefront API header for UniqueToken */
	var SHOPIFY_UNIQUE_TOKEN_HEADER = "X-Shopify-UniqueToken";
	var cachedTrackingValues = { current: null };
	/**
	* Retrieves user session tracking values for analytics and marketing from the browser environment.
	* @publicDocs
	*/
	function getTrackingValues() {
		let trackingValues;
		if (typeof window !== "undefined" && typeof window.performance !== "undefined") try {
			const resourceRE = /^https?:\/\/([^/]+)(\/api\/(?:unstable|2\d{3}-\d{2})\/graphql\.json(?=$|\?))?/;
			const entries = performance.getEntriesByType("resource");
			let matchedValues;
			for (let i = entries.length - 1; i >= 0; i--) {
				const entry = entries[i];
				if (entry.initiatorType !== "fetch") continue;
				const currentHost = window.location.host;
				const match = entry.name.match(resourceRE);
				if (!match) continue;
				const [, matchedHost, sfapiPath] = match;
				if (matchedHost === currentHost || sfapiPath && matchedHost?.endsWith(`.${currentHost}`)) {
					const values = extractFromPerformanceEntry(entry);
					if (values) {
						matchedValues = values;
						break;
					}
				}
			}
			if (matchedValues) trackingValues = matchedValues;
			if (trackingValues) cachedTrackingValues.current = trackingValues;
			else if (cachedTrackingValues.current) trackingValues = cachedTrackingValues.current;
			if (!trackingValues) {
				const navigationEntries = performance.getEntriesByType("navigation")[0];
				trackingValues = extractFromPerformanceEntry(navigationEntries, false);
			}
		} catch {}
		if (!trackingValues) {
			const cookie = typeof arguments[0] === "string" ? arguments[0] : typeof document !== "undefined" ? document.cookie : "";
			trackingValues = {
				uniqueToken: cookie.match(/\b_shopify_y=([^;]+)/)?.[1] || "",
				visitToken: cookie.match(/\b_shopify_s=([^;]+)/)?.[1] || "",
				consent: cookie.match(/\b_tracking_consent=([^;]+)/)?.[1] || ""
			};
		}
		return trackingValues;
	}
	function extractFromPerformanceEntry(entry, isConsentRequired = true) {
		let uniqueToken = "";
		let visitToken = "";
		let consent = "";
		const serverTiming = entry.serverTiming;
		if (serverTiming && serverTiming.length >= 3) for (let i = serverTiming.length - 1; i >= 0; i--) {
			const { name, description } = serverTiming[i];
			if (!name || !description) continue;
			if (name === "_y") uniqueToken = description;
			else if (name === "_s") visitToken = description;
			else if (name === "_cmp") consent = description;
			if (uniqueToken && visitToken && consent) break;
		}
		return uniqueToken && visitToken && (isConsentRequired ? consent : true) ? {
			uniqueToken,
			visitToken,
			consent
		} : void 0;
	}
	//#endregion
	//#region src/cart-hooks.tsx
	function useCartFetch() {
		const { storefrontId, getPublicTokenHeaders, getStorefrontApiUrl, sameDomainForStorefrontApi } = useShop();
		return (0, react.useCallback)(({ query, variables }) => {
			const headers = getPublicTokenHeaders({ contentType: "json" });
			if (storefrontId) headers[SHOPIFY_STOREFRONT_ID_HEADER] = storefrontId;
			if (!sameDomainForStorefrontApi) {
				const { uniqueToken, visitToken } = getTrackingValues();
				if (uniqueToken) {
					headers[SHOPIFY_STOREFRONT_Y_HEADER] = uniqueToken;
					headers[SHOPIFY_UNIQUE_TOKEN_HEADER] = uniqueToken;
				}
				if (visitToken) {
					headers[SHOPIFY_STOREFRONT_S_HEADER] = visitToken;
					headers[SHOPIFY_VISIT_TOKEN_HEADER] = visitToken;
				}
			}
			return fetch(getStorefrontApiUrl(), {
				method: "POST",
				headers,
				body: JSON.stringify({
					query: query.toString(),
					variables
				})
			}).then((res) => res.json()).catch((error) => {
				return {
					data: void 0,
					errors: error?.toString()
				};
			});
		}, [
			getPublicTokenHeaders,
			storefrontId,
			getStorefrontApiUrl,
			sameDomainForStorefrontApi
		]);
	}
	//#endregion
	//#region src/useCartActions.tsx
	/**
	* The `useCartActions` hook returns helper graphql functions for Storefront Cart API
	*
	* See [cart API graphql mutations](https://shopify.dev/api/storefront/2026-04/objects/Cart)
	*/
	function useCartActions({ numCartLines, cartFragment, countryCode = "US", languageCode = "EN" }) {
		const fetchCart = useCartFetch();
		const cartFetch = (0, react.useCallback)((cartId) => {
			return fetchCart({
				query: CartQuery(cartFragment),
				variables: {
					id: cartId,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			fetchCart,
			cartFragment,
			numCartLines,
			countryCode,
			languageCode
		]);
		const cartCreate = (0, react.useCallback)((cart) => {
			return fetchCart({
				query: CartCreate(cartFragment),
				variables: {
					input: cart,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const cartLineAdd = (0, react.useCallback)((cartId, lines) => {
			return fetchCart({
				query: CartLineAdd(cartFragment),
				variables: {
					cartId,
					lines,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const cartLineUpdate = (0, react.useCallback)((cartId, lines) => {
			return fetchCart({
				query: CartLineUpdate(cartFragment),
				variables: {
					cartId,
					lines,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const cartLineRemove = (0, react.useCallback)((cartId, lines) => {
			return fetchCart({
				query: CartLineRemove(cartFragment),
				variables: {
					cartId,
					lines,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const noteUpdate = (0, react.useCallback)((cartId, note) => {
			return fetchCart({
				query: CartNoteUpdate(cartFragment),
				variables: {
					cartId,
					note,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			fetchCart,
			cartFragment,
			numCartLines,
			countryCode,
			languageCode
		]);
		const buyerIdentityUpdate = (0, react.useCallback)((cartId, buyerIdentity) => {
			return fetchCart({
				query: CartBuyerIdentityUpdate(cartFragment),
				variables: {
					cartId,
					buyerIdentity,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const cartAttributesUpdate = (0, react.useCallback)((cartId, attributes) => {
			return fetchCart({
				query: CartAttributesUpdate(cartFragment),
				variables: {
					cartId,
					attributes,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		const discountCodesUpdate = (0, react.useCallback)((cartId, discountCodes) => {
			return fetchCart({
				query: CartDiscountCodesUpdate(cartFragment),
				variables: {
					cartId,
					discountCodes,
					numCartLines,
					country: countryCode,
					language: languageCode
				}
			});
		}, [
			cartFragment,
			countryCode,
			fetchCart,
			numCartLines,
			languageCode
		]);
		return (0, react.useMemo)(() => ({
			cartFetch,
			cartCreate,
			cartLineAdd,
			cartLineUpdate,
			cartLineRemove,
			noteUpdate,
			buyerIdentityUpdate,
			cartAttributesUpdate,
			discountCodesUpdate,
			cartFragment
		}), [
			cartFetch,
			cartCreate,
			cartLineAdd,
			cartLineUpdate,
			cartLineRemove,
			noteUpdate,
			buyerIdentityUpdate,
			cartAttributesUpdate,
			discountCodesUpdate,
			cartFragment
		]);
	}
	//#endregion
	//#region src/useCartAPIStateMachine.tsx
	function invokeCart(action, options) {
		return {
			entry: [
				...options?.entryActions || [],
				i({ lastValidCart: (context) => context?.cart }),
				"onCartActionEntry",
				"onCartActionOptimisticUI",
				action
			],
			on: {
				RESOLVE: {
					target: options?.resolveTarget || "idle",
					actions: [i({
						prevCart: (context) => context?.lastValidCart,
						cart: (_, event) => event?.payload?.cart,
						rawCartResult: (_, event) => event?.payload?.rawCartResult,
						errors: (_) => void 0
					})]
				},
				ERROR: {
					target: options?.errorTarget || "error",
					actions: [i({
						prevCart: (context) => context?.lastValidCart,
						cart: (context) => context?.lastValidCart,
						errors: (_, event) => event?.payload?.errors
					})]
				},
				CART_COMPLETED: {
					target: "cartCompleted",
					actions: i({
						prevCart: (_) => void 0,
						cart: (_) => void 0,
						lastValidCart: (_) => void 0,
						rawCartResult: (_) => void 0,
						errors: (_) => void 0
					})
				}
			},
			exit: ["onCartActionComplete", ...options?.exitActions || []]
		};
	}
	var INITIALIZING_CART_EVENTS = {
		CART_FETCH: { target: "cartFetching" },
		CART_CREATE: { target: "cartCreating" },
		CART_SET: {
			target: "idle",
			actions: [i({
				rawCartResult: (_, event) => event.payload.cart,
				cart: (_, event) => cartFromGraphQL(event.payload.cart)
			})]
		}
	};
	var UPDATING_CART_EVENTS = {
		CARTLINE_ADD: { target: "cartLineAdding" },
		CARTLINE_UPDATE: { target: "cartLineUpdating" },
		CARTLINE_REMOVE: { target: "cartLineRemoving" },
		NOTE_UPDATE: { target: "noteUpdating" },
		BUYER_IDENTITY_UPDATE: { target: "buyerIdentityUpdating" },
		CART_ATTRIBUTES_UPDATE: { target: "cartAttributesUpdating" },
		DISCOUNT_CODES_UPDATE: { target: "discountCodesUpdating" }
	};
	function createCartMachine(initialCart) {
		return s({
			id: "Cart",
			initial: initialCart ? "idle" : "uninitialized",
			context: { cart: initialCart && cartFromGraphQL(initialCart) },
			states: {
				uninitialized: { on: INITIALIZING_CART_EVENTS },
				cartCompleted: { on: INITIALIZING_CART_EVENTS },
				initializationError: { on: INITIALIZING_CART_EVENTS },
				idle: { on: {
					...INITIALIZING_CART_EVENTS,
					...UPDATING_CART_EVENTS
				} },
				error: { on: {
					...INITIALIZING_CART_EVENTS,
					...UPDATING_CART_EVENTS
				} },
				cartFetching: invokeCart("cartFetchAction", { errorTarget: "initializationError" }),
				cartCreating: invokeCart("cartCreateAction", { errorTarget: "initializationError" }),
				cartLineRemoving: invokeCart("cartLineRemoveAction"),
				cartLineUpdating: invokeCart("cartLineUpdateAction"),
				cartLineAdding: invokeCart("cartLineAddAction"),
				noteUpdating: invokeCart("noteUpdateAction"),
				buyerIdentityUpdating: invokeCart("buyerIdentityUpdateAction"),
				cartAttributesUpdating: invokeCart("cartAttributesUpdateAction"),
				discountCodesUpdating: invokeCart("discountCodesUpdateAction")
			}
		});
	}
	function useCartAPIStateMachine({ numCartLines, onCartActionEntry, onCartActionOptimisticUI, onCartActionComplete, data: cart, cartFragment, countryCode, languageCode }) {
		const { cartFetch, cartCreate, cartLineAdd, cartLineUpdate, cartLineRemove, noteUpdate, buyerIdentityUpdate, cartAttributesUpdate, discountCodesUpdate } = useCartActions({
			numCartLines,
			cartFragment,
			countryCode,
			languageCode
		});
		const [state, send, service] = useMachine((0, react.useMemo)(() => createCartMachine(cart), [cart]), { actions: {
			cartFetchAction: async (_, event) => {
				if (event.type !== "CART_FETCH") return;
				const { data, errors } = await cartFetch(event?.payload?.cartId);
				send(eventFromFetchResult(event, data?.cart, errors));
			},
			cartCreateAction: async (_, event) => {
				if (event.type !== "CART_CREATE") return;
				const { data, errors } = await cartCreate(event?.payload);
				send(eventFromFetchResult(event, data?.cartCreate?.cart, errors));
			},
			cartLineAddAction: async (context, event) => {
				if (event.type !== "CARTLINE_ADD" || !context?.cart?.id) return;
				const { data, errors } = await cartLineAdd(context.cart.id, event.payload.lines);
				send(eventFromFetchResult(event, data?.cartLinesAdd?.cart, errors));
			},
			cartLineUpdateAction: async (context, event) => {
				if (event.type !== "CARTLINE_UPDATE" || !context?.cart?.id) return;
				const { data, errors } = await cartLineUpdate(context.cart.id, event.payload.lines);
				send(eventFromFetchResult(event, data?.cartLinesUpdate?.cart, errors));
			},
			cartLineRemoveAction: async (context, event) => {
				if (event.type !== "CARTLINE_REMOVE" || !context?.cart?.id) return;
				const { data, errors } = await cartLineRemove(context.cart.id, event.payload.lines);
				send(eventFromFetchResult(event, data?.cartLinesRemove?.cart, errors));
			},
			noteUpdateAction: async (context, event) => {
				if (event.type !== "NOTE_UPDATE" || !context?.cart?.id) return;
				const { data, errors } = await noteUpdate(context.cart.id, event.payload.note);
				send(eventFromFetchResult(event, data?.cartNoteUpdate?.cart, errors));
			},
			buyerIdentityUpdateAction: async (context, event) => {
				if (event.type !== "BUYER_IDENTITY_UPDATE" || !context?.cart?.id) return;
				const { data, errors } = await buyerIdentityUpdate(context.cart.id, event.payload.buyerIdentity);
				send(eventFromFetchResult(event, data?.cartBuyerIdentityUpdate?.cart, errors));
			},
			cartAttributesUpdateAction: async (context, event) => {
				if (event.type !== "CART_ATTRIBUTES_UPDATE" || !context?.cart?.id) return;
				const { data, errors } = await cartAttributesUpdate(context.cart.id, event.payload.attributes);
				send(eventFromFetchResult(event, data?.cartAttributesUpdate?.cart, errors));
			},
			discountCodesUpdateAction: async (context, event) => {
				if (event.type !== "DISCOUNT_CODES_UPDATE" || !context?.cart?.id) return;
				const { data, errors } = await discountCodesUpdate(context.cart.id, event.payload.discountCodes);
				send(eventFromFetchResult(event, data?.cartDiscountCodesUpdate?.cart, errors));
			},
			...onCartActionEntry && { onCartActionEntry: (context, event) => {
				if (isCartActionEvent(event)) onCartActionEntry(context, event);
			} },
			...onCartActionOptimisticUI && { onCartActionOptimisticUI: i((context, event) => {
				return onCartActionOptimisticUI(context, event);
			}) },
			...onCartActionComplete && { onCartActionComplete: (context, event) => {
				if (isCartFetchResultEvent(event)) onCartActionComplete(context, event);
			} }
		} });
		return (0, react.useMemo)(() => [
			state,
			send,
			service
		], [
			state,
			send,
			service
		]);
	}
	function cartFromGraphQL(cart) {
		return {
			...cart,
			lines: flattenConnection(cart?.lines),
			note: cart.note ?? void 0
		};
	}
	function eventFromFetchResult(cartActionEvent, cart, errors) {
		if (errors) return {
			type: "ERROR",
			payload: {
				errors,
				cartActionEvent
			}
		};
		if (!cart) return {
			type: "CART_COMPLETED",
			payload: { cartActionEvent }
		};
		return {
			type: "RESOLVE",
			payload: {
				cart: cartFromGraphQL(cart),
				rawCartResult: cart,
				cartActionEvent
			}
		};
	}
	function isCartActionEvent(event) {
		return event.type === "CART_CREATE" || event.type === "CARTLINE_ADD" || event.type === "CARTLINE_UPDATE" || event.type === "CARTLINE_REMOVE" || event.type === "NOTE_UPDATE" || event.type === "BUYER_IDENTITY_UPDATE" || event.type === "CART_ATTRIBUTES_UPDATE" || event.type === "DISCOUNT_CODES_UPDATE";
	}
	function isCartFetchResultEvent(event) {
		return event.type === "RESOLVE" || event.type === "ERROR" || event.type === "CART_COMPLETED";
	}
	//#endregion
	//#region src/CartProvider.tsx
	var CartContext = (0, react.createContext)(null);
	/**
	* The `useCart` hook provides access to the cart object. It must be a descendent of a `CartProvider` component.
	* @publicDocs
	*/
	function useCart() {
		const context = (0, react.useContext)(CartContext);
		if (!context) throw new Error("Expected a Cart Context, but no Cart Context was found");
		return context;
	}
	/**
	* The `CartProvider` component synchronizes the state of the Storefront API Cart and a customer's cart,
	* and allows you to more easily manipulate the cart by adding, removing, and updating it.
	* It could be placed at the root of your app so that your whole app is able to use the `useCart()` hook anywhere.
	*
	* There are props that trigger when a call to the Storefront API is made, such as `onLineAdd={}` when a line is added to the cart.
	* There are also props that trigger when a call to the Storefront API is completed, such as `onLineAddComplete={}` when the fetch request for adding a line to the cart completes.
	*
	* The `CartProvider` component must be a descendant of the `ShopifyProvider` component.
	* @publicDocs
	*/
	function CartProvider({ children, numCartLines, onCreate, onLineAdd, onLineRemove, onLineUpdate, onNoteUpdate, onBuyerIdentityUpdate, onAttributesUpdate, onDiscountCodesUpdate, onCreateComplete, onLineAddComplete, onLineRemoveComplete, onLineUpdateComplete, onNoteUpdateComplete, onBuyerIdentityUpdateComplete, onAttributesUpdateComplete, onDiscountCodesUpdateComplete, data: cart, cartFragment = defaultCartFragment, customerAccessToken, countryCode, languageCode }) {
		const shop = useShop();
		if (!shop) throw new Error("<CartProvider> needs to be a descendant of <ShopifyProvider>");
		countryCode = (countryCode ?? shop.countryIsoCode ?? "US").toUpperCase();
		languageCode = (languageCode ?? shop.languageIsoCode ?? "EN").toUpperCase();
		if (countryCode) countryCode = countryCode.toUpperCase();
		const [prevCountryCode, setPrevCountryCode] = (0, react.useState)(countryCode);
		const [prevCustomerAccessToken, setPrevCustomerAccessToken] = (0, react.useState)(customerAccessToken);
		const customerOverridesCountryCode = (0, react.useRef)(false);
		if (prevCountryCode !== countryCode || prevCustomerAccessToken !== customerAccessToken) {
			setPrevCountryCode(countryCode);
			setPrevCustomerAccessToken(customerAccessToken);
			customerOverridesCountryCode.current = false;
		}
		const [cartState, cartSend] = useCartAPIStateMachine({
			numCartLines,
			data: cart,
			cartFragment,
			countryCode,
			languageCode,
			onCartActionEntry(_, event) {
				try {
					switch (event.type) {
						case "CART_CREATE": return onCreate?.();
						case "CARTLINE_ADD": return onLineAdd?.();
						case "CARTLINE_REMOVE": return onLineRemove?.();
						case "CARTLINE_UPDATE": return onLineUpdate?.();
						case "NOTE_UPDATE": return onNoteUpdate?.();
						case "BUYER_IDENTITY_UPDATE": return onBuyerIdentityUpdate?.();
						case "CART_ATTRIBUTES_UPDATE": return onAttributesUpdate?.();
						case "DISCOUNT_CODES_UPDATE": return onDiscountCodesUpdate?.();
					}
				} catch (error) {
					console.error("Cart entry action failed", error);
				}
			},
			onCartActionOptimisticUI(context, event) {
				if (!context.cart) return { ...context };
				switch (event.type) {
					case "CARTLINE_REMOVE": return {
						...context,
						cart: {
							...context.cart,
							lines: context?.cart?.lines?.filter((line) => line?.id && !event.payload.lines.includes(line?.id))
						}
					};
					case "CARTLINE_UPDATE": return {
						...context,
						cart: {
							...context.cart,
							lines: context?.cart?.lines?.map((line) => {
								const updatedLine = event.payload.lines.find(({ id }) => id === line?.id);
								if (updatedLine && updatedLine.quantity) return {
									...line,
									quantity: updatedLine.quantity
								};
								return line;
							})
						}
					};
				}
				return { ...context };
			},
			onCartActionComplete(context, event) {
				const cartActionEvent = event.payload.cartActionEvent;
				try {
					switch (event.type) {
						case "RESOLVE": switch (cartActionEvent.type) {
							case "CART_CREATE": return onCreateComplete?.();
							case "CARTLINE_ADD": return onLineAddComplete?.();
							case "CARTLINE_REMOVE": return onLineRemoveComplete?.();
							case "CARTLINE_UPDATE": return onLineUpdateComplete?.();
							case "NOTE_UPDATE": return onNoteUpdateComplete?.();
							case "BUYER_IDENTITY_UPDATE":
								if (countryCodeNotUpdated(context, cartActionEvent)) customerOverridesCountryCode.current = true;
								return onBuyerIdentityUpdateComplete?.();
							case "CART_ATTRIBUTES_UPDATE": return onAttributesUpdateComplete?.();
							case "DISCOUNT_CODES_UPDATE": return onDiscountCodesUpdateComplete?.();
						}
					}
				} catch (error) {
					console.error("onCartActionComplete failed", error);
				}
			}
		});
		const cartReady = (0, react.useRef)(false);
		const [isCartReady, setIsCartReady] = (0, react.useState)(false);
		const cartCompleted = cartState.matches("cartCompleted");
		const countryChanged = (cartState.value === "idle" || cartState.value === "error" || cartState.value === "cartCompleted") && countryCode !== cartState?.context?.cart?.buyerIdentity?.countryCode && !cartState.context.errors;
		const fetchingFromStorage = (0, react.useRef)(false);
		/**
		* Initializes cart with priority in this order:
		* 1. cart props
		* 2. localStorage cartId
		*/
		(0, react.useEffect)(() => {
			if (!cartReady.current && !fetchingFromStorage.current) {
				if (!cart && storageAvailable("localStorage")) {
					fetchingFromStorage.current = true;
					try {
						const cartId = window.localStorage.getItem(CART_ID_STORAGE_KEY);
						if (cartId) cartSend({
							type: "CART_FETCH",
							payload: { cartId }
						});
					} catch (error) {
						console.warn("error fetching cartId");
						console.warn(error);
					}
				}
				cartReady.current = true;
				setIsCartReady(true);
			}
		}, [
			cart,
			cartReady,
			cartSend
		]);
		(0, react.useEffect)(() => {
			if (!countryChanged || customerOverridesCountryCode.current) return;
			cartSend({
				type: "BUYER_IDENTITY_UPDATE",
				payload: { buyerIdentity: {
					countryCode,
					customerAccessToken
				} }
			});
		}, [
			countryCode,
			customerAccessToken,
			countryChanged,
			customerOverridesCountryCode,
			cartSend
		]);
		const onCartReadySend = (0, react.useCallback)((cartEvent) => {
			if (!cartReady.current) return console.warn("Cart isn't ready yet");
			cartSend(cartEvent);
		}, [cartSend]);
		(0, react.useEffect)(() => {
			if (cartState?.context?.cart?.id && storageAvailable("localStorage")) try {
				window.localStorage.setItem(CART_ID_STORAGE_KEY, cartState.context.cart?.id);
			} catch (error) {
				console.warn("Failed to save cartId to localStorage", error);
			}
		}, [cartState?.context?.cart?.id]);
		(0, react.useEffect)(() => {
			if (cartCompleted && storageAvailable("localStorage")) try {
				window.localStorage.removeItem(CART_ID_STORAGE_KEY);
			} catch (error) {
				console.warn("Failed to delete cartId from localStorage", error);
			}
		}, [cartCompleted]);
		const cartCreate = (0, react.useCallback)((cartInput) => {
			if (countryCode && !cartInput.buyerIdentity?.countryCode) {
				if (cartInput.buyerIdentity == null) cartInput.buyerIdentity = {};
				cartInput.buyerIdentity.countryCode = countryCode;
			}
			if (customerAccessToken && !cartInput.buyerIdentity?.customerAccessToken) {
				if (cartInput.buyerIdentity == null) cartInput.buyerIdentity = {};
				cartInput.buyerIdentity.customerAccessToken = customerAccessToken;
			}
			onCartReadySend({
				type: "CART_CREATE",
				payload: cartInput
			});
		}, [
			countryCode,
			customerAccessToken,
			onCartReadySend
		]);
		const cartDisplayState = useDelayedStateUntilHydration(cartState);
		const cartContextValue = (0, react.useMemo)(() => {
			return {
				...cartDisplayState?.context?.cart ?? {
					lines: [],
					attributes: []
				},
				status: transposeStatus(cartDisplayState.value),
				error: cartDisplayState?.context?.errors,
				totalQuantity: cartDisplayState?.context?.cart?.totalQuantity ?? 0,
				cartCreate,
				cartReady: isCartReady,
				linesAdd(lines) {
					if (cartDisplayState?.context?.cart?.id) onCartReadySend({
						type: "CARTLINE_ADD",
						payload: { lines }
					});
					else cartCreate({ lines });
				},
				linesRemove(lines) {
					onCartReadySend({
						type: "CARTLINE_REMOVE",
						payload: { lines }
					});
				},
				linesUpdate(lines) {
					onCartReadySend({
						type: "CARTLINE_UPDATE",
						payload: { lines }
					});
				},
				noteUpdate(note) {
					onCartReadySend({
						type: "NOTE_UPDATE",
						payload: { note }
					});
				},
				buyerIdentityUpdate(buyerIdentity) {
					onCartReadySend({
						type: "BUYER_IDENTITY_UPDATE",
						payload: { buyerIdentity }
					});
				},
				cartAttributesUpdate(attributes) {
					onCartReadySend({
						type: "CART_ATTRIBUTES_UPDATE",
						payload: { attributes }
					});
				},
				discountCodesUpdate(discountCodes) {
					onCartReadySend({
						type: "DISCOUNT_CODES_UPDATE",
						payload: { discountCodes }
					});
				},
				cartFragment
			};
		}, [
			cartCreate,
			isCartReady,
			cartDisplayState?.context?.cart,
			cartDisplayState?.context?.errors,
			cartDisplayState.value,
			cartFragment,
			onCartReadySend
		]);
		return /* @__PURE__ */ React.createElement(CartContext.Provider, { value: cartContextValue }, children);
	}
	function transposeStatus(status) {
		switch (status) {
			case "uninitialized":
			case "initializationError": return "uninitialized";
			case "idle":
			case "cartCompleted":
			case "error": return "idle";
			case "cartFetching": return "fetching";
			case "cartCreating": return "creating";
			case "cartLineAdding":
			case "cartLineRemoving":
			case "cartLineUpdating":
			case "noteUpdating":
			case "buyerIdentityUpdating":
			case "cartAttributesUpdating":
			case "discountCodesUpdating": return "updating";
		}
	}
	/**
	* Delays a state update until hydration finishes. Useful for preventing suspense boundaries errors when updating a context
	* @remarks this uses startTransition and waits for it to finish.
	*/
	function useDelayedStateUntilHydration(state) {
		const [isPending, startTransition] = (0, react.useTransition)();
		const [delayedState, setDelayedState] = (0, react.useState)(state);
		const firstTimePending = (0, react.useRef)(false);
		if (isPending) firstTimePending.current = true;
		const firstTimePendingFinished = (0, react.useRef)(false);
		if (!isPending && firstTimePending.current) firstTimePendingFinished.current = true;
		(0, react.useEffect)(() => {
			startTransition(() => {
				if (!firstTimePendingFinished.current) setDelayedState(state);
			});
		}, [state]);
		return firstTimePendingFinished.current ? state : delayedState;
	}
	/**
	* Check for storage availability function obtained from
	* @see https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
	*/
	function storageAvailable(type) {
		let storage;
		try {
			storage = window[type];
			const x = "__storage_test__";
			storage.setItem(x, x);
			storage.removeItem(x);
			return true;
		} catch (e) {
			return !!(e instanceof DOMException && (e.code === 22 || e.code === 1014 || e.name === "QuotaExceededError" || e.name === "NS_ERROR_DOM_QUOTA_REACHED") && storage && storage.length !== 0);
		}
	}
	function countryCodeNotUpdated(context, event) {
		return !!(event.payload.buyerIdentity.countryCode && context.cart?.buyerIdentity?.countryCode !== event.payload.buyerIdentity.countryCode);
	}
	//#endregion
	//#region src/ProductProvider.tsx
	var ProductOptionsContext = (0, react.createContext)(null);
	/**
	* `<ProductProvider />` is a context provider that enables use of the `useProduct()` hook.
	*
	* It helps manage selected options and variants for a product.
	* @publicDocs
	*/
	function ProductProvider({ children, data: product, initialVariantId: explicitVariantId }) {
		const variants = (0, react.useMemo)(() => flattenConnection(product.variants ?? {}), [product.variants]);
		if (!isProductVariantArray(variants)) throw new Error(`<ProductProvider/> requires 'product.variants.nodes' or 'product.variants.edges'`);
		const options = (0, react.useMemo)(() => getOptions(variants), [variants]);
		/**
		* Track the selectedVariant within the provider.
		*/
		const [selectedVariant, setSelectedVariant] = (0, react.useState)(() => getVariantBasedOnIdProp(explicitVariantId, variants));
		/**
		* Track the selectedOptions within the provider. If a `initialVariantId`
		* is passed, use that to select initial options.
		*/
		const [selectedOptions, setSelectedOptions] = (0, react.useState)(() => getSelectedOptions(selectedVariant));
		/**
		* When the initialVariantId changes, we need to make sure we
		* update the selected variant and selected options. If not,
		* then the selected variant and options will reference incorrect
		* values.
		*/
		(0, react.useEffect)(() => {
			const newSelectedVariant = getVariantBasedOnIdProp(explicitVariantId, variants);
			setSelectedVariant(newSelectedVariant);
			setSelectedOptions(getSelectedOptions(newSelectedVariant));
		}, [explicitVariantId, variants]);
		/**
		* Allow the developer to select an option.
		*/
		const setSelectedOption = (0, react.useCallback)((name, value) => {
			setSelectedOptions((selectedOptions) => {
				const opts = {
					...selectedOptions,
					[name]: value
				};
				setSelectedVariant(getSelectedVariant(variants, opts));
				return opts;
			});
		}, [setSelectedOptions, variants]);
		const isOptionInStock = (0, react.useCallback)((option, value) => {
			return getSelectedVariant(variants, {
				...selectedOptions,
				[option]: value
			})?.availableForSale ?? true;
		}, [selectedOptions, variants]);
		const sellingPlanGroups = (0, react.useMemo)(() => flattenConnection(product.sellingPlanGroups ?? {}).map((sellingPlanGroup) => ({
			...sellingPlanGroup,
			sellingPlans: flattenConnection(sellingPlanGroup?.sellingPlans ?? {})
		})), [product.sellingPlanGroups]);
		/**
		* Track the selectedSellingPlan within the hook. If `initialSellingPlanId`
		* is passed, use that as an initial value. Look it up from the `selectedVariant`, since
		* that is also a requirement.
		*/
		const [selectedSellingPlan, setSelectedSellingPlan] = (0, react.useState)(void 0);
		const selectedSellingPlanAllocation = (0, react.useMemo)(() => {
			if (!selectedVariant || !selectedSellingPlan) return;
			if (!selectedVariant.sellingPlanAllocations?.nodes && !selectedVariant.sellingPlanAllocations?.edges) throw new Error(`<ProductProvider/>: You must include 'sellingPlanAllocations.nodes' or 'sellingPlanAllocations.edges' in your variants in order to calculate selectedSellingPlanAllocation`);
			return flattenConnection(selectedVariant.sellingPlanAllocations).find((allocation) => allocation?.sellingPlan?.id === selectedSellingPlan.id);
		}, [selectedVariant, selectedSellingPlan]);
		const value = (0, react.useMemo)(() => ({
			product,
			variants,
			variantsConnection: product.variants,
			options,
			selectedVariant,
			setSelectedVariant,
			selectedOptions,
			setSelectedOption,
			setSelectedOptions,
			isOptionInStock,
			selectedSellingPlan,
			setSelectedSellingPlan,
			selectedSellingPlanAllocation,
			sellingPlanGroups,
			sellingPlanGroupsConnection: product.sellingPlanGroups
		}), [
			product,
			isOptionInStock,
			options,
			selectedOptions,
			selectedSellingPlan,
			selectedSellingPlanAllocation,
			selectedVariant,
			sellingPlanGroups,
			setSelectedOption,
			variants
		]);
		return /* @__PURE__ */ React.createElement(ProductOptionsContext.Provider, { value }, children);
	}
	/**
	* Provides access to the context value provided by `<ProductProvider />`. Must be a descendent of `<ProductProvider />`.
	* @publicDocs
	*/
	function useProduct() {
		const context = (0, react.useContext)(ProductOptionsContext);
		if (!context) throw new Error(`'useProduct' must be a child of <ProductProvider />`);
		return context;
	}
	function getSelectedVariant(variants, choices) {
		/**
		* Ensure the user has selected all the required options, not just some.
		*/
		if (!variants.length || variants?.[0]?.selectedOptions?.length !== Object.keys(choices).length) return;
		return variants?.find((variant) => {
			return Object.entries(choices).every(([name, value]) => {
				return variant?.selectedOptions?.some((option) => option?.name === name && option?.value === value);
			});
		});
	}
	function getOptions(variants) {
		const map = variants.reduce((memo, variant) => {
			if (!variant.selectedOptions) throw new Error(`'getOptions' requires 'variant.selectedOptions'`);
			variant?.selectedOptions?.forEach((opt) => {
				memo[opt?.name ?? ""] = memo[opt?.name ?? ""] || /* @__PURE__ */ new Set();
				memo[opt?.name ?? ""].add(opt?.value ?? "");
			});
			return memo;
		}, {});
		return Object.keys(map).map((option) => {
			return {
				name: option,
				values: Array.from(map[option])
			};
		});
	}
	function getVariantBasedOnIdProp(explicitVariantId, variants) {
		if (explicitVariantId) {
			const foundVariant = variants.find((variant) => variant?.id === explicitVariantId);
			if (!foundVariant) console.warn(`<ProductProvider/> received a 'initialVariantId' prop, but could not actually find a variant with that ID`);
			return foundVariant;
		}
		if (explicitVariantId === null) return null;
		if (explicitVariantId === void 0) return variants.find((variant) => variant?.availableForSale) || variants[0];
	}
	function getSelectedOptions(selectedVariant) {
		return selectedVariant?.selectedOptions ? selectedVariant.selectedOptions.reduce((memo, optionSet) => {
			memo[optionSet?.name ?? ""] = optionSet?.value ?? "";
			return memo;
		}, {}) : {};
	}
	function isProductVariantArray(maybeVariantArray) {
		if (!maybeVariantArray || !Array.isArray(maybeVariantArray)) return false;
		return true;
	}
	//#endregion
	//#region src/BaseButton.tsx
	function BaseButton(props) {
		const { as, onClick, defaultOnClick, children, buttonRef, ...passthroughProps } = props;
		const handleOnClick = (0, react.useCallback)((event) => {
			if (onClick) {
				const clickShouldContinue = onClick(event);
				if (typeof clickShouldContinue === "boolean" && clickShouldContinue === false || event?.defaultPrevented) return;
			}
			defaultOnClick?.(event);
		}, [defaultOnClick, onClick]);
		const Component = as || "button";
		return /* @__PURE__ */ React.createElement(Component, {
			ref: buttonRef,
			onClick: handleOnClick,
			...passthroughProps
		}, children);
	}
	//#endregion
	//#region src/AddToCartButton.tsx
	/**
	* The `AddToCartButton` component renders a button that adds an item to the cart when pressed.
	* It must be a descendent of the `CartProvider` component.
	* @publicDocs
	*/
	function AddToCartButton(props) {
		const [addingItem, setAddingItem] = (0, react.useState)(false);
		const { variantId: explicitVariantId, quantity = 1, attributes, sellingPlanId, onClick, children, accessibleAddingToCartLabel, parent, ...passthroughProps } = props;
		const { status, linesAdd } = useCart();
		const { selectedVariant } = useProduct();
		const variantId = explicitVariantId ?? selectedVariant?.id ?? "";
		const disabled = explicitVariantId === null || variantId === "" || selectedVariant === null || addingItem || passthroughProps.disabled;
		(0, react.useEffect)(() => {
			if (addingItem && status === "idle") setAddingItem(false);
		}, [status, addingItem]);
		const handleAddItem = (0, react.useCallback)(() => {
			setAddingItem(true);
			linesAdd([{
				quantity,
				merchandiseId: variantId || "",
				attributes,
				parent,
				sellingPlanId
			}]);
		}, [
			linesAdd,
			quantity,
			variantId,
			attributes,
			sellingPlanId,
			parent
		]);
		return /* @__PURE__ */ react.createElement(react.Fragment, null, /* @__PURE__ */ react.createElement(BaseButton, {
			...passthroughProps,
			disabled,
			onClick,
			defaultOnClick: handleAddItem
		}, children), accessibleAddingToCartLabel ? /* @__PURE__ */ react.createElement("p", {
			style: {
				position: "absolute",
				width: "1px",
				height: "1px",
				padding: "0",
				margin: "-1px",
				overflow: "hidden",
				clip: "rect(0, 0, 0, 0)",
				whiteSpace: "nowrap",
				borderWidth: "0"
			},
			role: "alert",
			"aria-live": "assertive"
		}, addingItem ? accessibleAddingToCartLabel : null) : null);
	}
	//#endregion
	//#region src/analytics-constants.ts
	var AnalyticsEventName = {
		PAGE_VIEW: "PAGE_VIEW",
		ADD_TO_CART: "ADD_TO_CART",
		PAGE_VIEW_2: "PAGE_VIEW_2",
		COLLECTION_VIEW: "COLLECTION_VIEW",
		PRODUCT_VIEW: "PRODUCT_VIEW",
		SEARCH_VIEW: "SEARCH_VIEW"
	};
	var AnalyticsPageType = {
		article: "article",
		blog: "blog",
		captcha: "captcha",
		cart: "cart",
		collection: "collection",
		customersAccount: "customers/account",
		customersActivateAccount: "customers/activate_account",
		customersAddresses: "customers/addresses",
		customersLogin: "customers/login",
		customersOrder: "customers/order",
		customersRegister: "customers/register",
		customersResetPassword: "customers/reset_password",
		giftCard: "gift_card",
		home: "index",
		listCollections: "list-collections",
		forbidden: "403",
		notFound: "404",
		page: "page",
		password: "password",
		product: "product",
		policy: "policy",
		search: "search"
	};
	var ShopifySalesChannel = {
		hydrogen: "hydrogen",
		headless: "headless"
	};
	var ShopifyAppId = {
		hydrogen: "6167201",
		headless: "12875497473"
	};
	//#endregion
	//#region src/analytics-utils.ts
	/**
	* Builds a Shopify Monorail event from a Shopify Monorail payload and a schema ID.
	* @param payload - The Monorail payload
	* @param schemaId - The schema ID to use
	* @returns The formatted payload
	**/
	function schemaWrapper(schemaId, payload) {
		return {
			schema_id: schemaId,
			payload,
			metadata: { event_created_at_ms: Date.now() }
		};
	}
	/**
	* Parses global id (gid) and returns the resource type and id.
	* @see https://shopify.dev/api/usage/gids
	* @param gid - A shopify GID (string)
	*
	* @example
	* ```ts
	* const {id, resource} = parseGid('gid://shopify/Order/123')
	* // => id = "123", resource = 'Order'
	*
	*  * const {id, resource} = parseGid('gid://shopify/Cart/abc123')
	* // => id = "abc123", resource = 'Cart'
	* ```
	* @publicDocs
	**/
	function parseGid(gid) {
		const defaultReturn = {
			id: "",
			resource: null,
			resourceId: null,
			search: "",
			searchParams: new URLSearchParams(),
			hash: ""
		};
		if (typeof gid !== "string") return defaultReturn;
		try {
			const { search, searchParams, pathname, hash } = new URL(gid);
			const pathnameParts = pathname.split("/");
			const lastPathnamePart = pathnameParts[pathnameParts.length - 1];
			const resourcePart = pathnameParts[pathnameParts.length - 2];
			if (!lastPathnamePart || !resourcePart) return defaultReturn;
			return {
				id: `${lastPathnamePart}${search}${hash}` || "",
				resource: resourcePart ?? null,
				resourceId: lastPathnamePart || null,
				search,
				searchParams,
				hash
			};
		} catch {
			return defaultReturn;
		}
	}
	/**
	* Filters properties from an object and returns a new object with only the properties that have a truthy value.
	* @param keyValuePairs - An object of key-value pairs
	* @param formattedData - An object which will hold the truthy values
	* @returns The formatted object
	**/
	function addDataIf(keyValuePairs, formattedData) {
		if (typeof keyValuePairs !== "object") return {};
		Object.entries(keyValuePairs).forEach(([key, value]) => {
			if (value) formattedData[key] = value;
		});
		return formattedData;
	}
	/**
	* Utility that errors if a function is called on the server.
	* @param fnName - The name of the function
	* @returns A boolean
	**/
	function errorIfServer(fnName) {
		if (typeof document === "undefined") {
			console.error(`${fnName} should only be used within the useEffect callback or event handlers`);
			return true;
		}
		return false;
	}
	//#endregion
	//#region src/cookies-utils.tsx
	var tokenHash = "xxxx-4xxx-xxxx-xxxxxxxxxxxx";
	function buildUUID() {
		let hash = "";
		try {
			const crypto = window.crypto;
			const randomValuesArray = new Uint16Array(31);
			crypto.getRandomValues(randomValuesArray);
			let i = 0;
			hash = tokenHash.replace(/[x]/g, (c) => {
				const r = randomValuesArray[i] % 16;
				const v = c === "x" ? r : r & 3 | 8;
				i++;
				return v.toString(16);
			}).toUpperCase();
		} catch (err) {
			hash = tokenHash.replace(/[x]/g, (c) => {
				const r = Math.random() * 16 | 0;
				return (c === "x" ? r : r & 3 | 8).toString(16);
			}).toUpperCase();
		}
		return `${hexTime()}-${hash}`;
	}
	function hexTime() {
		let dateNumber = 0;
		let perfNumber = 0;
		dateNumber = (/* @__PURE__ */ new Date()).getTime() >>> 0;
		try {
			perfNumber = performance.now() >>> 0;
		} catch (err) {
			perfNumber = 0;
		}
		return Math.abs(dateNumber + perfNumber).toString(16).toLowerCase().padStart(8, "0");
	}
	/**
	* Gets the values of _shopify_y and _shopify_s cookies from the provided cookie string. If the Shopify cookies doesn't exist, this method will return an empty string for each missing cookie.
	* @deprecated Use getTrackingValues instead.
	* @publicDocs
	*/
	function getShopifyCookies(cookies) {
		const trackingValues = getTrackingValues(cookies);
		return {
			[SHOPIFY_Y]: trackingValues.uniqueToken,
			[SHOPIFY_S]: trackingValues.visitToken
		};
	}
	//#endregion
	//#region src/analytics-schema-trekkie-storefront-page-view.ts
	var SCHEMA_ID$1 = "trekkie_storefront_page_view/1.4";
	var OXYGEN_DOMAIN = "myshopify.dev";
	function pageView$1(payload) {
		const pageViewPayload = payload;
		const { id, resource } = parseGid(pageViewPayload.resourceId);
		const resourceType = resource ? resource.toLowerCase() : void 0;
		return [schemaWrapper(SCHEMA_ID$1, addDataIf({
			pageType: pageViewPayload.pageType,
			customerId: parseInt(parseGid(pageViewPayload.customerId).id || "0"),
			resourceType,
			resourceId: parseInt(id)
		}, formatPayload$1(pageViewPayload)))];
	}
	function formatPayload$1(payload) {
		return {
			appClientId: payload.shopifySalesChannel ? ShopifyAppId[payload.shopifySalesChannel] : ShopifyAppId.headless,
			isMerchantRequest: isMerchantRequest(payload.url),
			hydrogenSubchannelId: payload.storefrontId || payload.hydrogenSubchannelId || "0",
			isPersistentCookie: payload.hasUserConsent,
			uniqToken: payload.uniqueToken,
			visitToken: payload.visitToken,
			microSessionId: buildUUID(),
			microSessionCount: 1,
			url: payload.url,
			path: payload.path,
			search: payload.search,
			referrer: payload.referrer,
			title: payload.title,
			shopId: parseInt(parseGid(payload.shopId).id),
			currency: payload.currency,
			contentLanguage: payload.acceptedLanguage || "en"
		};
	}
	function isMerchantRequest(url) {
		if (typeof url !== "string") return false;
		const hostname = new URL(url).hostname;
		if (hostname.indexOf(OXYGEN_DOMAIN) !== -1 || hostname === "localhost") return true;
		return false;
	}
	//#endregion
	//#region src/analytics-schema-custom-storefront-customer-tracking.ts
	var SCHEMA_ID = "custom_storefront_customer_tracking/1.2";
	var PAGE_RENDERED_EVENT_NAME = "page_rendered";
	var COLLECTION_PAGE_RENDERED_EVENT_NAME = "collection_page_rendered";
	var PRODUCT_PAGE_RENDERED_EVENT_NAME = "product_page_rendered";
	var PRODUCT_ADDED_TO_CART_EVENT_NAME = "product_added_to_cart";
	var SEARCH_SUBMITTED_EVENT_NAME = "search_submitted";
	function prepareAdditionalPayload(payload) {
		return {
			canonical_url: payload.canonicalUrl || payload.url,
			customer_id: parseInt(parseGid(payload.customerId).id || "0")
		};
	}
	function pageView(payload) {
		const pageViewPayload = payload;
		const additionalPayload = prepareAdditionalPayload(pageViewPayload);
		const pageType = pageViewPayload.pageType;
		const pageViewEvents = [];
		pageViewEvents.push(schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: PAGE_RENDERED_EVENT_NAME,
			...additionalPayload
		}, formatPayload(pageViewPayload))));
		switch (pageType) {
			case AnalyticsPageType.collection:
				pageViewEvents.push(schemaWrapper(SCHEMA_ID, addDataIf({
					event_name: COLLECTION_PAGE_RENDERED_EVENT_NAME,
					...additionalPayload,
					collection_name: pageViewPayload.collectionHandle,
					collection_id: parseInt(parseGid(pageViewPayload.collectionId).id)
				}, formatPayload(pageViewPayload))));
				break;
			case AnalyticsPageType.product:
				pageViewEvents.push(schemaWrapper(SCHEMA_ID, addDataIf({
					event_name: PRODUCT_PAGE_RENDERED_EVENT_NAME,
					...additionalPayload,
					products: formatProductPayload(pageViewPayload.products),
					total_value: pageViewPayload.totalValue
				}, formatPayload(pageViewPayload))));
				break;
			case AnalyticsPageType.search:
				pageViewEvents.push(schemaWrapper(SCHEMA_ID, addDataIf({
					event_name: SEARCH_SUBMITTED_EVENT_NAME,
					...additionalPayload,
					search_string: pageViewPayload.searchString
				}, formatPayload(pageViewPayload))));
				break;
		}
		return pageViewEvents;
	}
	function pageView2(payload) {
		const pageViewPayload = payload;
		return [schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: PAGE_RENDERED_EVENT_NAME,
			...prepareAdditionalPayload(pageViewPayload)
		}, formatPayload(pageViewPayload)))];
	}
	function collectionView(payload) {
		const pageViewPayload = payload;
		return [schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: COLLECTION_PAGE_RENDERED_EVENT_NAME,
			...prepareAdditionalPayload(pageViewPayload),
			collection_name: pageViewPayload.collectionHandle,
			collection_id: parseInt(parseGid(pageViewPayload.collectionId).id)
		}, formatPayload(pageViewPayload)))];
	}
	function productView(payload) {
		const pageViewPayload = payload;
		return [schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: PRODUCT_PAGE_RENDERED_EVENT_NAME,
			...prepareAdditionalPayload(pageViewPayload),
			products: formatProductPayload(pageViewPayload.products),
			total_value: pageViewPayload.totalValue
		}, formatPayload(pageViewPayload)))];
	}
	function searchView(payload) {
		const pageViewPayload = payload;
		return [schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: SEARCH_SUBMITTED_EVENT_NAME,
			...prepareAdditionalPayload(pageViewPayload),
			search_string: pageViewPayload.searchString
		}, formatPayload(pageViewPayload)))];
	}
	function addToCart(payload) {
		const addToCartPayload = payload;
		const cartToken = parseGid(addToCartPayload.cartId);
		return [schemaWrapper(SCHEMA_ID, addDataIf({
			event_name: PRODUCT_ADDED_TO_CART_EVENT_NAME,
			customerId: addToCartPayload.customerId,
			cart_token: cartToken?.id ? `${cartToken.id}` : null,
			total_value: addToCartPayload.totalValue,
			products: formatProductPayload(addToCartPayload.products),
			customer_id: parseInt(parseGid(addToCartPayload.customerId).id || "0")
		}, formatPayload(addToCartPayload)))];
	}
	function formatPayload(payload) {
		const payloadWithPrivacy = payload;
		return {
			source: payload.shopifySalesChannel || ShopifySalesChannel.headless,
			asset_version_id: payload.assetVersionId || "2026.4.2",
			hydrogenSubchannelId: payload.storefrontId || payload.hydrogenSubchannelId || "0",
			is_persistent_cookie: payload.hasUserConsent,
			deprecated_visit_token: payload.visitToken,
			unique_token: payload.uniqueToken,
			event_time: Date.now(),
			event_id: buildUUID(),
			event_source_url: payload.url,
			referrer: payload.referrer,
			user_agent: payload.userAgent,
			navigation_type: payload.navigationType,
			navigation_api: payload.navigationApi,
			shop_id: parseInt(parseGid(payload.shopId).id),
			currency: payload.currency,
			ccpa_enforced: payloadWithPrivacy.ccpaEnforced || false,
			gdpr_enforced: payloadWithPrivacy.gdprEnforced || false,
			gdpr_enforced_as_string: payloadWithPrivacy.gdprEnforced ? "true" : "false",
			analytics_allowed: payload.analyticsAllowed || false,
			marketing_allowed: payload.marketingAllowed || false,
			sale_of_data_allowed: payload.saleOfDataAllowed || false
		};
	}
	function formatProductPayload(products) {
		return products ? products.map((p) => {
			const product = addDataIf({
				variant_gid: p.variantGid,
				category: p.category,
				sku: p.sku,
				product_id: parseInt(parseGid(p.productGid).id),
				variant_id: parseInt(parseGid(p.variantGid).id)
			}, {
				product_gid: p.productGid,
				name: p.name,
				variant: p.variantName || "",
				brand: p.brand,
				price: parseFloat(p.price),
				quantity: Number(p.quantity || 0)
			});
			return JSON.stringify(product);
		}) : [];
	}
	//#endregion
	//#region src/analytics.ts
	/**
	* Set user and session cookies and refresh the expiry time. If `event.payload.hasUserConsent` is false, no analytics event will happen.
	* @param event - The analytics event.
	* @param shopDomain - The Online Store domain to sent Shopify analytics under the same
	*   top level domain.
	* @publicDocs
	*/
	function sendShopifyAnalytics(event, shopDomain) {
		const { eventName, payload } = event;
		if (!payload.hasUserConsent) return Promise.resolve();
		let events = [];
		const pageViewPayload = payload;
		if (eventName === AnalyticsEventName.PAGE_VIEW) events = events.concat(pageView$1(pageViewPayload), pageView(pageViewPayload));
		else if (eventName === AnalyticsEventName.ADD_TO_CART) events = events.concat(addToCart(payload));
		else if (eventName === AnalyticsEventName.PAGE_VIEW_2) events = events.concat(pageView$1(pageViewPayload), pageView2(pageViewPayload));
		else if (eventName === AnalyticsEventName.COLLECTION_VIEW) events = events.concat(collectionView(pageViewPayload));
		else if (eventName === AnalyticsEventName.PRODUCT_VIEW) events = events.concat(productView(pageViewPayload));
		else if (eventName === AnalyticsEventName.SEARCH_VIEW) events = events.concat(searchView(pageViewPayload));
		if (events.length) return sendToShopify(events, shopDomain);
		else return Promise.resolve();
	}
	function isLighthouseUserAgent() {
		if (typeof window === "undefined" || !window.navigator) return false;
		return /Chrome-Lighthouse/.test(window.navigator.userAgent);
	}
	var ERROR_MESSAGE = "sendShopifyAnalytics request is unsuccessful";
	function sendToShopify(events, shopDomain) {
		if (isLighthouseUserAgent()) return Promise.resolve();
		const eventsToBeSent = {
			events,
			metadata: { event_sent_at_ms: Date.now() }
		};
		try {
			return fetch(shopDomain ? `https://${shopDomain}/.well-known/shopify/monorail/unstable/produce_batch` : "https://monorail-edge.shopifysvc.com/unstable/produce_batch", {
				method: "post",
				headers: { "content-type": "text/plain" },
				body: JSON.stringify(eventsToBeSent)
			}).then((response) => {
				if (!response.ok) throw new Error("Response failed");
				return response.text();
			}).then((data) => {
				if (data) JSON.parse(data).result.forEach((eventResponse) => {
					if (eventResponse.status !== 200) console.error(ERROR_MESSAGE, "\n\n", eventResponse.message);
				});
			}).catch((err) => {
				console.error(ERROR_MESSAGE, err);
				throw new Error(ERROR_MESSAGE);
			});
		} catch (error) {
			return Promise.resolve();
		}
	}
	/**
	* If executed on server, this method will return empty string for each field.
	* @publicDocs
	*/
	function getClientBrowserParameters() {
		if (errorIfServer("getClientBrowserParameters")) return {
			uniqueToken: "",
			visitToken: "",
			url: "",
			path: "",
			search: "",
			referrer: "",
			title: "",
			userAgent: "",
			navigationType: "",
			navigationApi: ""
		};
		const [navigationType, navigationApi] = getNavigationType();
		const trackingValues = getTrackingValues();
		return {
			uniqueToken: trackingValues.uniqueToken,
			visitToken: trackingValues.visitToken,
			url: location.href,
			path: location.pathname,
			search: location.search,
			referrer: document.referrer,
			title: document.title,
			userAgent: navigator.userAgent,
			navigationType,
			navigationApi
		};
	}
	function getNavigationTypeExperimental() {
		try {
			const navigationEntries = performance?.getEntriesByType && performance?.getEntriesByType("navigation");
			if (navigationEntries && navigationEntries[0]) {
				const rawType = window.performance.getEntriesByType("navigation")[0]["type"];
				return rawType && rawType.toString();
			}
		} catch (err) {}
	}
	function getNavigationTypeLegacy() {
		try {
			if (PerformanceNavigation && performance?.navigation?.type !== null && performance?.navigation?.type !== void 0) {
				const rawType = performance.navigation.type;
				switch (rawType) {
					case PerformanceNavigation.TYPE_NAVIGATE: return "navigate";
					case PerformanceNavigation.TYPE_RELOAD: return "reload";
					case PerformanceNavigation.TYPE_BACK_FORWARD: return "back_forward";
					default: return `unknown: ${rawType}`;
				}
			}
		} catch (err) {}
	}
	function getNavigationType() {
		try {
			let navApi = "PerformanceNavigationTiming";
			let navType = getNavigationTypeExperimental();
			if (!navType) {
				navType = getNavigationTypeLegacy();
				navApi = "performance.navigation";
			}
			if (navType) return [navType, navApi];
			else return ["unknown", "unknown"];
		} catch (err) {}
		return ["error", "error"];
	}
	//#endregion
	//#region src/BuyNowButton.tsx
	/**
	* The `BuyNowButton` component renders a button that adds an item to the cart and redirects the customer to checkout.
	* Must be a child of a `CartProvider` component.
	* @publicDocs
	*/
	function BuyNowButton(props) {
		const { cartCreate, checkoutUrl } = useCart();
		const [loading, setLoading] = (0, react.useState)(false);
		const { quantity, variantId, sellingPlanId, onClick, attributes, children, ...passthroughProps } = props;
		(0, react.useEffect)(() => {
			if (loading && checkoutUrl) window.location.href = checkoutUrl;
		}, [loading, checkoutUrl]);
		const handleBuyNow = (0, react.useCallback)(() => {
			setLoading(true);
			cartCreate({ lines: [{
				quantity: quantity ?? 1,
				merchandiseId: variantId,
				attributes,
				sellingPlanId
			}] });
		}, [
			cartCreate,
			quantity,
			variantId,
			attributes,
			sellingPlanId
		]);
		return /* @__PURE__ */ React.createElement(BaseButton, {
			disabled: loading ?? passthroughProps.disabled,
			...passthroughProps,
			onClick,
			defaultOnClick: handleBuyNow
		}, children);
	}
	//#endregion
	//#region src/CartCheckoutButton.tsx
	/**
	* The `CartCheckoutButton` component renders a button that redirects to the checkout URL for the cart.
	* It must be a descendent of a `CartProvider` component.
	* @publicDocs
	*/
	function CartCheckoutButton(props) {
		const [requestedCheckout, setRequestedCheckout] = (0, react.useState)(false);
		const { status, checkoutUrl } = useCart();
		const { children, ...passthroughProps } = props;
		(0, react.useEffect)(() => {
			if (requestedCheckout && checkoutUrl && status === "idle") window.location.href = checkoutUrl;
		}, [
			requestedCheckout,
			status,
			checkoutUrl
		]);
		return /* @__PURE__ */ React.createElement(BaseButton, {
			...passthroughProps,
			disabled: requestedCheckout || passthroughProps.disabled,
			onClick: () => setRequestedCheckout(true)
		}, children);
	}
	//#endregion
	//#region src/useMoney.tsx
	/**
	* The `useMoney` hook takes a [MoneyV2 object from the Storefront API](https://shopify.dev/docs/api/storefront/2026-04/objects/MoneyV2)
	* or a [MoneyV2 object from the Customer Account API](https://shopify.dev/docs/api/customer/2026-04/objects/moneyv2) and returns a
	* default-formatted string of the amount with the correct currency indicator, along with some of the parts provided by
	* [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat).
	* Uses `locale` from `ShopifyProvider`
	* &nbsp;
	* @see {@link https://shopify.dev/api/hydrogen/hooks/usemoney}
	* @example initialize the money object
	* ```ts
	* const money = useMoney({
	*   amount: '100.00',
	*   currencyCode: 'USD'
	* })
	* ```
	* &nbsp;
	*
	* @example basic usage, outputs: $100.00
	* ```ts
	* money.localizedString
	* ```
	* &nbsp;
	*
	* @example without currency, outputs: 100.00
	* ```ts
	* money.amount
	* ```
	* &nbsp;
	*
	* @example without trailing zeros, outputs: $100
	* ```ts
	* money.withoutTrailingZeros
	* ```
	* &nbsp;
	*
	* @example currency name, outputs: US dollars
	* ```ts
	* money.currencyCode
	* ```
	* &nbsp;
	*
	* @example currency symbol, outputs: $
	* ```ts
	* money.currencySymbol
	* ```
	* &nbsp;
	*
	* @example without currency and without trailing zeros, outputs: 100
	* ```ts
	* money.withoutTrailingZerosAndCurrency
	* ```
	* @publicDocs
	*/
	function useMoney(money) {
		const { countryIsoCode, languageIsoCode } = useShop();
		const locale = languageIsoCode.includes("_") ? languageIsoCode.replace("_", "-") : `${languageIsoCode}-${countryIsoCode}`;
		if (!locale) throw new Error(`useMoney(): Unable to get 'locale' from 'useShop()', which means that 'locale' was not passed to '<ShopifyProvider/>'. 'locale' is required for 'useMoney()' to work`);
		const amount = parseFloat(money.amount);
		let isCurrencySupported = true;
		try {
			new Intl.NumberFormat(locale, {
				style: "currency",
				currency: money.currencyCode
			});
		} catch (e) {
			if (e instanceof RangeError && e.message.includes("currency")) isCurrencySupported = false;
		}
		const { defaultFormatter, nameFormatter, narrowSymbolFormatter, withoutTrailingZerosFormatter, withoutCurrencyFormatter, withoutTrailingZerosOrCurrencyFormatter } = (0, react.useMemo)(() => {
			const options = isCurrencySupported ? {
				style: "currency",
				currency: money.currencyCode
			} : {
				style: "decimal",
				minimumFractionDigits: 2,
				maximumFractionDigits: 2
			};
			return {
				defaultFormatter: getLazyFormatter(locale, options),
				nameFormatter: getLazyFormatter(locale, {
					...options,
					currencyDisplay: "name"
				}),
				narrowSymbolFormatter: getLazyFormatter(locale, {
					...options,
					currencyDisplay: "narrowSymbol"
				}),
				withoutTrailingZerosFormatter: getLazyFormatter(locale, {
					...options,
					minimumFractionDigits: 0,
					maximumFractionDigits: 0
				}),
				withoutCurrencyFormatter: getLazyFormatter(locale, {
					minimumFractionDigits: 2,
					maximumFractionDigits: 2
				}),
				withoutTrailingZerosOrCurrencyFormatter: getLazyFormatter(locale, {
					minimumFractionDigits: 0,
					maximumFractionDigits: 0
				})
			};
		}, [
			money.currencyCode,
			locale,
			isCurrencySupported
		]);
		const isPartCurrency = (part) => part.type === "currency";
		const lazyFormatters = (0, react.useMemo)(() => ({
			original: () => money,
			currencyCode: () => money.currencyCode,
			localizedString: () => {
				const formatted = defaultFormatter().format(amount);
				return isCurrencySupported ? formatted : `${formatted} ${money.currencyCode}`;
			},
			parts: () => {
				const parts = defaultFormatter().formatToParts(amount);
				if (!isCurrencySupported) parts.push({
					type: "literal",
					value: " "
				}, {
					type: "currency",
					value: money.currencyCode
				});
				return parts;
			},
			withoutTrailingZeros: () => {
				const formatted = amount % 1 === 0 ? withoutTrailingZerosFormatter().format(amount) : defaultFormatter().format(amount);
				return isCurrencySupported ? formatted : `${formatted} ${money.currencyCode}`;
			},
			withoutTrailingZerosAndCurrency: () => amount % 1 === 0 ? withoutTrailingZerosOrCurrencyFormatter().format(amount) : withoutCurrencyFormatter().format(amount),
			currencyName: () => nameFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? money.currencyCode,
			currencySymbol: () => defaultFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? money.currencyCode,
			currencyNarrowSymbol: () => narrowSymbolFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? "",
			amount: () => defaultFormatter().formatToParts(amount).filter((part) => [
				"decimal",
				"fraction",
				"group",
				"integer",
				"literal"
			].includes(part.type)).map((part) => part.value).join("")
		}), [
			money,
			amount,
			isCurrencySupported,
			nameFormatter,
			defaultFormatter,
			narrowSymbolFormatter,
			withoutCurrencyFormatter,
			withoutTrailingZerosFormatter,
			withoutTrailingZerosOrCurrencyFormatter
		]);
		return (0, react.useMemo)(() => new Proxy(lazyFormatters, { get: (target, key) => Reflect.get(target, key)?.call(null) }), [lazyFormatters]);
	}
	var formatterCache = /* @__PURE__ */ new Map();
	function getLazyFormatter(locale, options) {
		const key = JSON.stringify([locale, options]);
		return function() {
			let formatter = formatterCache.get(key);
			if (!formatter) {
				try {
					formatter = new Intl.NumberFormat(locale, options);
				} catch (error) {
					if (error instanceof RangeError && error.message.includes("currency")) {
						const fallbackOptions = { ...options };
						delete fallbackOptions.currency;
						delete fallbackOptions.currencyDisplay;
						delete fallbackOptions.currencySign;
						formatter = new Intl.NumberFormat(locale, fallbackOptions);
					} else throw error;
				}
				formatterCache.set(key, formatter);
			}
			return formatter;
		};
	}
	//#endregion
	//#region src/Money.tsx
	/**
	* The `Money` component renders a string of the [Storefront API's MoneyV2 object](https://shopify.dev/docs/api/storefront/2026-04/objects/MoneyV2)
	* or the [Customer Account API's MoneyV2 object](https://shopify.dev/docs/api/customer/2026-04/objects/moneyv2)
	* according to the `locale` in the `ShopifyProvider` component.
	* &nbsp;
	* @see {@link https://shopify.dev/api/hydrogen/components/money}
	* @example basic usage, outputs: $100.00
	* ```ts
	* <Money data={{amount: '100.00', currencyCode: 'USD'}} />
	* ```
	* &nbsp;
	*
	* @example without currency, outputs: 100.00
	* ```ts
	* <Money data={{amount: '100.00', currencyCode: 'USD'}} withoutCurrency />
	* ```
	* &nbsp;
	*
	* @example without trailing zeros, outputs: $100
	* ```ts
	* <Money data={{amount: '100.00', currencyCode: 'USD'}} withoutTrailingZeros />
	* ```
	* &nbsp;
	*
	* @example with per-unit measurement, outputs: $100.00 per G
	* ```ts
	* <Money
	*   data={{amount: '100.00', currencyCode: 'USD'}}
	*   measurement={{referenceUnit: 'G'}}
	*   measurementSeparator=" per "
	* />
	* ```
	* @publicDocs
	*/
	function Money({ data, as, withoutCurrency, withoutTrailingZeros, measurement, measurementSeparator = "/", ...passthroughProps }) {
		if (!isMoney(data)) throw new Error(`<Money/> needs a valid 'data' prop that has 'amount' and 'currencyCode'`);
		const moneyObject = useMoney(data);
		const Wrapper = as ?? "div";
		let output = moneyObject.localizedString;
		if (withoutCurrency || withoutTrailingZeros) if (withoutCurrency && !withoutTrailingZeros) output = moneyObject.amount;
		else if (!withoutCurrency && withoutTrailingZeros) output = moneyObject.withoutTrailingZeros;
		else output = moneyObject.withoutTrailingZerosAndCurrency;
		return /* @__PURE__ */ React.createElement(Wrapper, passthroughProps, output, measurement && measurement.referenceUnit && /* @__PURE__ */ React.createElement(React.Fragment, null, measurementSeparator, measurement.referenceUnit));
	}
	function isMoney(maybeMoney) {
		return typeof maybeMoney.amount === "string" && !!maybeMoney.amount && typeof maybeMoney.currencyCode === "string" && !!maybeMoney.currencyCode;
	}
	//#endregion
	//#region src/CartCost.tsx
	/**
	* The `CartCost` component renders a `Money` component with the cost associated with the `amountType` prop.
	* If no `amountType` prop is specified, then it defaults to `totalAmount`.
	* Depends on `useCart()` and must be a child of `<CartProvider/>`
	* @publicDocs
	*/
	function CartCost(props) {
		const { cost } = useCart();
		const { amountType = "total", children, ...passthroughProps } = props;
		let amount;
		if (amountType == "total") amount = cost?.totalAmount;
		else if (amountType == "subtotal") amount = cost?.subtotalAmount;
		else if (amountType == "tax") amount = cost?.totalTaxAmount;
		else if (amountType == "duty") amount = cost?.totalDutyAmount;
		if (amount == null) return null;
		return /* @__PURE__ */ React.createElement(Money, {
			...passthroughProps,
			data: amount
		}, children);
	}
	//#endregion
	//#region src/CartLineProvider.tsx
	var CartLineContext = (0, react.createContext)(null);
	/**
	* The `useCartLine` hook provides access to the [CartLine object](https://shopify.dev/api/storefront/2026-04/objects/cartline) from the Storefront API. It must be a descendent of a `CartProvider` component.
	* @publicDocs
	*/
	function useCartLine() {
		const context = (0, react.useContext)(CartLineContext);
		if (context == null) throw new Error("Expected a cart line context but none was found");
		return context;
	}
	/**
	* The `CartLineProvider` component creates a context for using a cart line.
	* @publicDocs
	*/
	function CartLineProvider({ children, line }) {
		return /* @__PURE__ */ React.createElement(CartLineContext.Provider, { value: line }, children);
	}
	//#endregion
	//#region src/CartLineQuantity.tsx
	/**
	* The `<CartLineQuantity/>` component renders a `span` (or another element / component that can be customized by the `as` prop) with the cart line's quantity.
	*
	* It must be a descendent of a `<CartLineProvider/>` component, and uses the `useCartLine()` hook internally.
	* @publicDocs
	*/
	function CartLineQuantity(props) {
		const cartLine = useCartLine();
		const { as, ...passthroughProps } = props;
		const Wrapper = as ? as : "span";
		return /* @__PURE__ */ React.createElement(Wrapper, passthroughProps, cartLine.quantity);
	}
	//#endregion
	//#region src/CartLineQuantityAdjustButton.tsx
	/**
	* The `<CartLineQuantityAdjustButton />` component renders a button that adjusts the cart line's quantity when pressed.
	*
	* It must be a descendent of `<CartLineProvider/>` and `<CartProvider/>`.
	* @publicDocs
	*/
	function CartLineQuantityAdjustButton(props) {
		const { status, linesRemove, linesUpdate } = useCart();
		const cartLine = useCartLine();
		const { children, adjust, onClick, ...passthroughProps } = props;
		const handleAdjust = (0, react.useCallback)(() => {
			if (adjust === "remove") {
				linesRemove([cartLine?.id ?? ""]);
				return;
			}
			const quantity = adjust === "decrease" ? (cartLine?.quantity ?? 0) - 1 : (cartLine?.quantity ?? 0) + 1;
			if (quantity <= 0) {
				linesRemove([cartLine?.id ?? ""]);
				return;
			}
			linesUpdate([{
				id: cartLine?.id ?? "",
				quantity,
				attributes: cartLine?.attributes ?? []
			}]);
		}, [
			adjust,
			cartLine?.attributes,
			cartLine?.id,
			cartLine?.quantity,
			linesRemove,
			linesUpdate
		]);
		const disabledAttr = passthroughProps.disabled;
		return /* @__PURE__ */ React.createElement(BaseButton, {
			...passthroughProps,
			onClick,
			defaultOnClick: handleAdjust,
			disabled: typeof disabledAttr !== "undefined" ? disabledAttr : status !== "idle"
		}, children);
	}
	//#endregion
	//#region src/codegen.helpers.ts
	/**
	* Meant to be used with GraphQL CodeGen to type the Storefront API's custom scalars correctly.
	* Reference for the GraphQL types: https://shopify.dev/docs/api/storefront/2026-04/scalars/HTML
	* Note: JSON is generated as 'unknown' by default.
	* @publicDocs
	*/
	var storefrontApiCustomScalars = {
		DateTime: "string",
		Decimal: "string",
		HTML: "string",
		URL: "string",
		Color: "string",
		UnsignedInt64: "string"
	};
	/**
	* Meant to be used with GraphQL CodeGen to type the Storefront API's custom scalars correctly.
	* Reference for the GraphQL types: https://shopify.dev/docs/api/customer/2026-04/scalars/HTML
	* Note: JSON is generated as 'unknown' by default.
	* @publicDocs
	*/
	var customerAccountApiCustomScalars = {
		DateTime: "string",
		Decimal: "string",
		HTML: "string",
		ISO8601DateTime: "string",
		URL: "string",
		UnsignedInt64: "string"
	};
	//#endregion
	//#region src/ExternalVideo.tsx
	/**
	* The `ExternalVideo` component renders an embedded video for the Storefront
	* API's [ExternalVideo object](https://shopify.dev/api/storefront/reference/products/externalvideo).
	* @publicDocs
	*/
	var ExternalVideo = (0, react.forwardRef)((props, ref) => {
		const { data, options, id = data.id, frameBorder = "0", allow = "accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture", allowFullScreen = true, loading = "lazy", ...passthroughProps } = props;
		if (!data.embedUrl) throw new Error(`<ExternalVideo/> requires the 'embedUrl' property`);
		let finalUrl = data.embedUrl;
		if (options) {
			const urlObject = new URL(data.embedUrl);
			for (const [key, value] of Object.entries(options)) {
				if (typeof value === "undefined") continue;
				urlObject.searchParams.set(key, value.toString());
			}
			finalUrl = urlObject.toString();
		}
		return /* @__PURE__ */ React.createElement("iframe", {
			...passthroughProps,
			id: id ?? data.embedUrl,
			title: data.alt ?? data.id ?? "external video",
			frameBorder,
			allow,
			allowFullScreen,
			src: finalUrl,
			loading,
			ref
		});
	});
	//#endregion
	//#region src/optionValueDecoder.ts
	var OPTION_VALUE_SEPARATOR = ",";
	var V1_CONTROL_CHARS = {
		OPTION: ":",
		END_OF_PREFIX: ",",
		SEQUENCE_GAP: " ",
		RANGE: "-"
	};
	/**
	* Determine whether an option value combination is present in an encoded option value string. Function is memoized by encodedVariantField.
	*
	* @param targetOptionValueCombination - Indices of option values to look up in the encoded option value string. A partial set of indices may be passed to determine whether a node or any children is present. For example, if a product has 3 options, passing [0] will return true if any option value combination for the first option's option value is present in the encoded string.
	* @param encodedVariantField - Encoded option value string from the Storefront API, e.g. [product.encodedVariantExistence](/docs/api/storefront/2026-04/objects/Product#field-encodedvariantexistence) or [product.encodedVariantAvailability](/docs/api/storefront/2026-04/objects/Product#field-encodedvariantavailability)
	* @returns - True if a full or partial targetOptionValueIndices is present in the encoded option value string, false otherwise.
	* @publicDocs
	*/
	var isOptionValueCombinationInEncodedVariant = (() => {
		const decodedOptionValues = /* @__PURE__ */ new Map();
		return function(targetOptionValueCombination, encodedVariantField) {
			if (targetOptionValueCombination.length === 0) return false;
			if (!decodedOptionValues.has(encodedVariantField)) {
				const decodedOptionValuesSet = /* @__PURE__ */ new Set();
				for (const optionValue of decodeEncodedVariant(encodedVariantField)) {
					decodedOptionValuesSet.add(optionValue.join(OPTION_VALUE_SEPARATOR));
					for (let i = 0; i < optionValue.length; i++) decodedOptionValuesSet.add(optionValue.slice(0, i + 1).join(OPTION_VALUE_SEPARATOR));
				}
				decodedOptionValues.set(encodedVariantField, decodedOptionValuesSet);
			}
			return Boolean(decodedOptionValues.get(encodedVariantField)?.has(targetOptionValueCombination.join(OPTION_VALUE_SEPARATOR)));
		};
	})();
	/**
	* For an encoded option value string, decode into option value combinations. Entries represent a valid combination formatted as an array of option value positions.
	* @param encodedVariantField - Encoded option value string from the Storefront API, e.g. [product.encodedVariantExistence](/docs/api/storefront/2026-04/objects/Product#field-encodedvariantexistence) or [product.encodedVariantAvailability](/docs/api/storefront/2026-04/objects/Product#field-encodedvariantavailability)
	* @returns Decoded option value combinations
	* @publicDocs
	*/
	function decodeEncodedVariant(encodedVariantField) {
		if (!encodedVariantField) return [];
		if (encodedVariantField.startsWith("v1_")) return v1Decoder(stripVersion(encodedVariantField));
		throw new Error("Unsupported option value encoding");
	}
	var stripVersion = (encodedVariantField) => encodedVariantField.replace(/^v1_/, "");
	/**
	* We encode an array of arrays representing variants, expressed in terms of options and option values, as a trie.
	*
	* This encoding strategy allows extremely large numbers of variants to be expressed in an extremely compact data structure.
	*
	* Integers represent option and values, so [0,0,0] represents option_value at array index 0 for the options at array indexes 0, 1 and 2
	*
	* `:`, `,`, ` ` and `-` are control characters.
	* `:` indicates a new option
	* `,` indicates the end of a repeated prefix, mulitple consecutive commas indicate the end of multiple repeated prefixes.
	* ` ` indicates a gap in the sequence of option values
	* `-` indicates a continuous range of option values
	*
	* Encoding process:
	*
	* example input array: [[0,0,0], [0,1,0], [0,1,1], [1,0,0], [1,0,1], [1,1,1], [2,0,1], [2,1,0]]
	*
	* step 1: encode as string: "0:0:0,0:1:0,0:1:1,1:0:0,1:0:1,1:1:1,2:0:1,2:1:0,"
	* step 2: combine nodes that share a prefix: "0:0:0,0:1:0 1,1:0:0 1,1:1:1,2:0:1,2:1:0,"
	* step 3: encode data as a trie so no prefixes need to be repeated: "0:0:0,1:0 1,,1:0:0 1,1:1,,2:0:1,1:0,,"
	* step 4: since the options are sorted, use a dash to express ranges: "0:0:0,1:0-1,,1:0:0-1,1:1,,2:0:1,1:0,,"
	*/
	function v1Decoder(encodedVariantField) {
		const tokenizer = /[ :,-]/g;
		let index = 0;
		let token;
		const options = [];
		const currentOptionValue = [];
		let depth = 0;
		let rangeStart = null;
		while (token = tokenizer.exec(encodedVariantField)) {
			const operation = token[0];
			const optionValueIndex = Number.parseInt(encodedVariantField.slice(index, token.index)) || 0;
			if (rangeStart !== null) {
				for (; rangeStart < optionValueIndex; rangeStart++) {
					currentOptionValue[depth] = rangeStart;
					options.push([...currentOptionValue]);
				}
				rangeStart = null;
			}
			currentOptionValue[depth] = optionValueIndex;
			if (operation === V1_CONTROL_CHARS.RANGE) rangeStart = optionValueIndex;
			else if (operation === V1_CONTROL_CHARS.OPTION) depth++;
			else {
				if (operation === V1_CONTROL_CHARS.SEQUENCE_GAP || operation === V1_CONTROL_CHARS.END_OF_PREFIX && encodedVariantField[token.index - 1] !== V1_CONTROL_CHARS.END_OF_PREFIX) options.push([...currentOptionValue]);
				if (operation === V1_CONTROL_CHARS.END_OF_PREFIX) {
					currentOptionValue.pop();
					depth--;
				}
			}
			index = tokenizer.lastIndex;
		}
		const encodingEndsWithIndex = encodedVariantField.match(/\d+$/g);
		if (encodingEndsWithIndex) {
			const finalValueIndex = parseInt(encodingEndsWithIndex[0]);
			if (rangeStart != null) for (; rangeStart <= finalValueIndex; rangeStart++) {
				currentOptionValue[depth] = rangeStart;
				options.push([...currentOptionValue]);
			}
			else options.push([finalValueIndex]);
		}
		return options;
	}
	//#endregion
	//#region src/getProductOptions.ts
	/**
	* Creates a mapping of product options to their index for matching encoded values
	* For example, a product option of
	*  [
	*    \{
	*      name: 'Color',
	*      optionValues: [\{name: 'Red'\}, \{name: 'Blue'\}]
	*    \},
	*    \{
	*      name: 'Size',
	*      optionValues: [\{name: 'Small'\}, \{name: 'Medium'\}, \{name: 'Large'\}]
	*    \}
	*  ]
	* Would return
	*  \{
	*    'Color': \{Red: 0, Blue: 1\},
	*    'Size': \{Small: 0, Medium: 1, Large: 2\}
	*  \}
	*/
	function mapProductOptions(options) {
		return Object.assign({}, ...options.map((option) => {
			return { [option.name]: Object.assign({}, ...option?.optionValues ? option.optionValues.map((value, index) => {
				return { [value.name]: index };
			}) : []) };
		}));
	}
	/**
	* Converts the product option into an Object\<key, value\> for building query params
	* For example, a selected product option of
	*  [
	*    \{
	*      name: 'Color',
	*      value: 'Red',
	*    \},
	*    \{
	*      name: 'Size',
	*      value: 'Medium',
	*    \}
	*  ]
	* Would return
	*  \{
	*    Color: 'Red',
	*    Size: 'Medium',
	*  \}
	* @publicDocs
	*/
	function mapSelectedProductOptionToObject(options) {
		return Object.assign({}, ...options.map((key) => {
			return { [key.name]: key.value };
		}));
	}
	/**
	* Returns the JSON stringify result of mapSelectedProductOptionToObject
	*/
	function mapSelectedProductOptionToObjectAsString(options) {
		return JSON.stringify(mapSelectedProductOptionToObject(options));
	}
	/**
	* Encode the selected product option as a key for mapping to the encoded variants
	* For example, a selected product option of
	*  [
	*    \{
	*      name: 'Color',
	*      value: 'Red',
	*    \},
	*    \{
	*      name: 'Size',
	*      value: 'Medium',
	*    \}
	*  ]
	* Would return
	*  JSON.stringify(\{
	*    Color: 'Red',
	*    Size: 'Medium',
	*  \})
	*/
	function encodeSelectedProductOptionAsKey(selectedOption) {
		if (Array.isArray(selectedOption)) return JSON.stringify(Object.assign({}, ...selectedOption.map((option) => ({ [option.name]: option.value }))));
		else return JSON.stringify(selectedOption);
	}
	/**
	* Build the encoding array for the given selected options. For example, if we have
	* the following productOptionMappings:
	*
	*  \{
	*    'Color': \{Red: 0, Blue: 1\},
	*    'Size': \{Small: 0, Medium: 1, Large: 2\}
	*  \}
	*
	* A selectedOption of
	*
	* \{
	*    Color: 'Red',
	*    Size: 'Medium',
	* \}
	*
	* `buildEncodingArrayFromSelectedOptions` will produce
	*
	* [0,1]
	*
	* If in the case where a selected option doesn't exists in the mapping array, for example:
	*
	* \{
	*    Color: 'Red',
	*    Fabric: 'Cotton',
	*    Size: 'Medium',
	* \}
	*
	* `buildEncodingArrayFromSelectedOptions` will still produce
	*
	*  [0,1]
	*
	* This can be caused by when we do not have all the product
	* option information for the loading optimistic variant
	*/
	function buildEncodingArrayFromSelectedOptions(selectedOption, productOptionMappings) {
		return Object.keys(selectedOption).map((key) => {
			return productOptionMappings[key] ? productOptionMappings[key][selectedOption[key]] : null;
		}).filter((code) => code !== null);
	}
	/**
	* Takes an array of product variants and maps them to an object with the encoded selected option values as the key.
	* For example, a product variant of
	* [
	*  \{
	*    id: 1,
	*    selectedOptions: [
	*      \{name: 'Color', value: 'Red'\},
	*      \{name: 'Size', value: 'Small'\},
	*    ],
	*  \},
	*  \{
	*    id: 2,
	*    selectedOptions: [
	*      \{name: 'Color', value: 'Red'\},
	*      \{name: 'Size', value: 'Medium'\},
	*    ],
	*  \}
	* ]
	* Would return
	* \{
	*    '[0,0]': \{id: 1, selectedOptions: [\{name: 'Color', value: 'Red'\}, \{name: 'Size', value: 'Small'\}]\},
	*    '[0,1]': \{id: 2, selectedOptions: [\{name: 'Color', value: 'Red'\}, \{name: 'Size', value: 'Medium'\}]\},
	* \}
	*/
	function mapVariants(variants) {
		return Object.assign({}, ...variants.map((variant) => {
			return { [encodeSelectedProductOptionAsKey(variant.selectedOptions || [])]: variant };
		}));
	}
	var PRODUCT_INPUTS = [
		"options",
		"selectedOrFirstAvailableVariant",
		"adjacentVariants"
	];
	var PRODUCT_INPUTS_EXTRA = [
		"handle",
		"encodedVariantExistence",
		"encodedVariantAvailability"
	];
	function logErrorAndReturnFalse(key) {
		console.error(`[h2:error:getProductOptions] product.${key} is missing. Make sure you query for this field from the Storefront API.`);
		return false;
	}
	function checkProductParam(product, checkAll = false) {
		let validParam = true;
		const productKeys = Object.keys(product);
		(checkAll ? [...PRODUCT_INPUTS, ...PRODUCT_INPUTS_EXTRA] : PRODUCT_INPUTS).forEach((key) => {
			if (!productKeys.includes(key)) validParam = logErrorAndReturnFalse(key);
		});
		if (product.options) {
			const firstOption = product?.options[0];
			if (checkAll && !firstOption?.name) validParam = logErrorAndReturnFalse("options.name");
			if (product?.options[0]?.optionValues) {
				let firstOptionValues = product.options[0].optionValues[0];
				if (checkAll && !firstOptionValues?.name) validParam = logErrorAndReturnFalse("options.optionValues.name");
				firstOptionValues = product.options[0].optionValues.filter((value) => !!value?.firstSelectableVariant)[0];
				if (firstOptionValues?.firstSelectableVariant) validParam = checkProductVariantParam(firstOptionValues.firstSelectableVariant, "options.optionValues.firstSelectableVariant", validParam, checkAll);
			} else validParam = logErrorAndReturnFalse("options.optionValues");
		}
		if (product.selectedOrFirstAvailableVariant) validParam = checkProductVariantParam(product.selectedOrFirstAvailableVariant, "selectedOrFirstAvailableVariant", validParam, checkAll);
		if (!!product.adjacentVariants && product.adjacentVariants[0]) validParam = checkProductVariantParam(product.adjacentVariants[0], "adjacentVariants", validParam, checkAll);
		return validParam ? product : {};
	}
	function checkProductVariantParam(variant, key, currentValidParamState, checkAll) {
		let validParam = currentValidParamState;
		if (checkAll && !variant.product?.handle) validParam = logErrorAndReturnFalse(`${key}.product.handle`);
		if (variant.selectedOptions) {
			const firstSelectedOption = variant.selectedOptions[0];
			if (!firstSelectedOption?.name) validParam = logErrorAndReturnFalse(`${key}.selectedOptions.name`);
			if (!firstSelectedOption?.value) validParam = logErrorAndReturnFalse(`${key}.selectedOptions.value`);
		} else validParam = logErrorAndReturnFalse(`${key}.selectedOptions`);
		return validParam;
	}
	/**
	* Finds all the variants provided by adjacentVariants, options.optionValues.firstAvailableVariant,
	* and selectedOrFirstAvailableVariant and return them in a single array
	* @publicDocs
	*/
	function getAdjacentAndFirstAvailableVariants(product) {
		const checkedProduct = checkProductParam(product);
		if (!checkedProduct.options) return [];
		const availableVariants = {};
		checkedProduct.options.map((option) => {
			option.optionValues?.map((value) => {
				if (value.firstSelectableVariant) {
					const variantKey = mapSelectedProductOptionToObjectAsString(value.firstSelectableVariant.selectedOptions);
					availableVariants[variantKey] = value.firstSelectableVariant;
				}
			});
		});
		checkedProduct.adjacentVariants.map((variant) => {
			const variantKey = mapSelectedProductOptionToObjectAsString(variant.selectedOptions);
			availableVariants[variantKey] = variant;
		});
		const selectedVariant = checkedProduct.selectedOrFirstAvailableVariant;
		if (selectedVariant) {
			const variantKey = mapSelectedProductOptionToObjectAsString(selectedVariant.selectedOptions);
			availableVariants[variantKey] = selectedVariant;
		}
		return Object.values(availableVariants);
	}
	/**
	* Returns a product options array with its relevant information
	* about the variant
	* @publicDocs
	*/
	function getProductOptions(product) {
		const checkedProduct = checkProductParam(product, true);
		if (!checkedProduct.options) return [];
		const { options, selectedOrFirstAvailableVariant: selectedVariant, adjacentVariants, encodedVariantExistence, encodedVariantAvailability, handle: productHandle } = checkedProduct;
		const selectedOptionKeys = selectedVariant?.selectedOptions.map((option) => option.name);
		const filteredOptions = options.filter((option) => {
			return selectedOptionKeys && selectedOptionKeys.indexOf(option.name) >= 0;
		});
		const productOptionMappings = mapProductOptions(options);
		const variants = mapVariants(selectedVariant ? [selectedVariant, ...adjacentVariants] : adjacentVariants);
		const selectedOptions = mapSelectedProductOptionToObject(selectedVariant ? selectedVariant.selectedOptions : []);
		return filteredOptions.map((option, optionIndex) => {
			return {
				...option,
				optionValues: option.optionValues.map((value) => {
					const targetOptionParams = { ...selectedOptions };
					targetOptionParams[option.name] = value.name;
					const targetKey = encodeSelectedProductOptionAsKey(targetOptionParams || []);
					const topDownKey = buildEncodingArrayFromSelectedOptions(targetOptionParams || [], productOptionMappings).slice(0, optionIndex + 1);
					const exists = isOptionValueCombinationInEncodedVariant(topDownKey, encodedVariantExistence || "");
					const available = isOptionValueCombinationInEncodedVariant(topDownKey, encodedVariantAvailability || "");
					const variant = variants[targetKey] || value.firstSelectableVariant;
					let variantOptionParam = {};
					if (variant) variantOptionParam = mapSelectedProductOptionToObject(variant.selectedOptions || []);
					const searchParams = new URLSearchParams(variantOptionParam);
					const handle = variant?.product?.handle || productHandle;
					return {
						...value,
						variant,
						handle,
						variantUriQuery: searchParams.toString(),
						selected: selectedOptions[option.name] === value.name,
						exists,
						available,
						isDifferentProduct: handle !== productHandle
					};
				})
			};
		});
	}
	//#endregion
	//#region src/Image.tsx
	/**
	* A Storefront API GraphQL fragment that can be used to query for an image.
	*/
	var IMAGE_FRAGMENT = `#graphql
  fragment Image on Image {
    altText
    url
    width
    height
  }
`;
	/**
	* Hydrogen’s Image component is a wrapper around the HTML image element.
	* It supports the same props as the HTML `img` element, but automatically
	* generates the srcSet and sizes attributes for you. For most use cases,
	* you’ll want to set the `aspectRatio` prop to ensure the image is sized
	* correctly.
	*
	* @remarks
	* - `decoding` is set to `async` by default.
	* - `loading` is set to `lazy` by default.
	* - `alt` will automatically be set to the `altText` from the Storefront API if passed in the `data` prop
	* - `src` will automatically be set to the `url` from the Storefront API if passed in the `data` prop
	*
	* @example
	* A responsive image with a 4:5 aspect ratio:
	* ```
	* <Image
	*   data={product.featuredImage}
	*   aspectRatio="4/5"
	*   sizes="(min-width: 45em) 40vw, 100vw"
	* />
	* ```
	* @example
	* A fixed size image:
	* ```
	* <Image
	*   data={product.featuredImage}
	*   width={100}
	*   height={100}
	* />
	* ```
	*
	* {@link https://shopify.dev/docs/api/hydrogen-react/components/image}
	* @publicDocs
	*/
	var Image = react.forwardRef(({ alt, aspectRatio, crop = "center", data, decoding = "async", height = "auto", loader = shopifyLoader, loading = "lazy", sizes, src, srcSetOptions = {
		intervals: 15,
		startingWidth: 200,
		incrementSize: 200,
		placeholderWidth: 100
	}, width = "100%", ...passthroughProps }, ref) => {
		const normalizedData = react.useMemo(() => {
			const dataWidth = data?.width && data?.height ? data?.width : void 0;
			const dataHeight = data?.width && data?.height ? data?.height : void 0;
			return {
				width: dataWidth,
				height: dataHeight,
				unitsMatch: Boolean(unitsMatch(dataWidth, dataHeight))
			};
		}, [data]);
		const normalizedProps = react.useMemo(() => {
			const widthParts = getUnitValueParts((width || "100%").toString());
			const nWidth = `${widthParts.number}${widthParts.unit}`;
			const autoHeight = height === void 0 || height === null;
			const heightParts = autoHeight ? null : getUnitValueParts(height.toString());
			const fixedHeight = heightParts ? `${heightParts.number}${heightParts.unit}` : "";
			const nHeight = autoHeight ? "auto" : fixedHeight;
			const nSrc = src || data?.url;
			if (!nSrc) console.warn(`No src or data.url provided to Image component.`, passthroughProps?.key || "");
			return {
				width: nWidth,
				height: nHeight,
				src: nSrc,
				alt: data?.altText && !alt ? data?.altText : alt || "",
				aspectRatio: aspectRatio ? aspectRatio : normalizedData.unitsMatch ? [getNormalizedFixedUnit(normalizedData.width), getNormalizedFixedUnit(normalizedData.height)].join("/") : void 0
			};
		}, [
			width,
			height,
			src,
			data,
			alt,
			aspectRatio,
			normalizedData,
			passthroughProps?.key
		]);
		const { intervals, startingWidth, incrementSize, placeholderWidth } = srcSetOptions;
		const imageWidths = react.useMemo(() => {
			return generateImageWidths(width, intervals, startingWidth, incrementSize);
		}, [
			width,
			intervals,
			startingWidth,
			incrementSize
		]);
		const fixedWidth = isFixedWidth(normalizedProps.width);
		if (!sizes && !fixedWidth) console.warn([
			"No sizes prop provided to Image component,",
			"you may be loading unnecessarily large images.",
			`Image used is ${src || data?.url || passthroughProps?.key || "unknown"}`
		].join(" "));
		if (fixedWidth) return /* @__PURE__ */ react.createElement(FixedWidthImage, {
			aspectRatio,
			crop,
			decoding,
			height,
			imageWidths,
			loader,
			loading,
			normalizedProps,
			passthroughProps,
			ref,
			width,
			data
		});
		else return /* @__PURE__ */ react.createElement(FluidImage, {
			aspectRatio,
			crop,
			decoding,
			imageWidths,
			loader,
			loading,
			normalizedProps,
			passthroughProps,
			placeholderWidth,
			ref,
			sizes,
			data
		});
	});
	var FixedWidthImage = react.forwardRef(({ aspectRatio, crop, decoding, height, imageWidths, loader = shopifyLoader, loading, normalizedProps, passthroughProps, width, data }, ref) => {
		const fixed = react.useMemo(() => {
			const intWidth = getNormalizedFixedUnit(width);
			const intHeight = getNormalizedFixedUnit(height);
			const fixedAspectRatio = aspectRatio ? aspectRatio : unitsMatch(normalizedProps.width, normalizedProps.height) ? [intWidth, intHeight].join("/") : normalizedProps.aspectRatio ? normalizedProps.aspectRatio : void 0;
			const sizesArray = imageWidths === void 0 ? void 0 : generateSizes(imageWidths, fixedAspectRatio, crop, {
				width: data?.width ?? void 0,
				height: data?.height ?? void 0
			});
			const fixedHeight = intHeight ? intHeight : fixedAspectRatio && intWidth ? intWidth * (parseAspectRatio(fixedAspectRatio) ?? 1) : void 0;
			return {
				width: intWidth,
				aspectRatio: fixedAspectRatio,
				height: fixedHeight,
				srcSet: generateSrcSet(normalizedProps.src, sizesArray, loader),
				src: loader({
					src: normalizedProps.src,
					width: intWidth,
					height: fixedHeight,
					crop: normalizedProps.height === "auto" ? void 0 : crop
				})
			};
		}, [
			aspectRatio,
			crop,
			data,
			height,
			imageWidths,
			loader,
			normalizedProps,
			width
		]);
		return /* @__PURE__ */ react.createElement("img", {
			ref,
			alt: normalizedProps.alt,
			decoding,
			height: fixed.height,
			loading,
			src: fixed.src,
			srcSet: fixed.srcSet,
			width: fixed.width,
			style: {
				aspectRatio: fixed.aspectRatio,
				...passthroughProps.style
			},
			...passthroughProps
		});
	});
	var FluidImage = react.forwardRef(({ crop, decoding, imageWidths, loader = shopifyLoader, loading, normalizedProps, passthroughProps, placeholderWidth, sizes, data }, ref) => {
		const fluid = react.useMemo(() => {
			const sizesArray = imageWidths === void 0 ? void 0 : generateSizes(imageWidths, normalizedProps.aspectRatio, crop, {
				width: data?.width ?? void 0,
				height: data?.height ?? void 0
			});
			const placeholderHeight = normalizedProps.aspectRatio && placeholderWidth ? placeholderWidth * (parseAspectRatio(normalizedProps.aspectRatio) ?? 1) : void 0;
			return {
				placeholderHeight,
				srcSet: generateSrcSet(normalizedProps.src, sizesArray, loader),
				src: loader({
					src: normalizedProps.src,
					width: placeholderWidth,
					height: placeholderHeight,
					crop
				})
			};
		}, [
			crop,
			data,
			imageWidths,
			loader,
			normalizedProps,
			placeholderWidth
		]);
		return /* @__PURE__ */ react.createElement("img", {
			ref,
			alt: normalizedProps.alt,
			decoding,
			height: fluid.placeholderHeight,
			loading,
			sizes,
			src: fluid.src,
			srcSet: fluid.srcSet,
			width: placeholderWidth,
			...passthroughProps,
			style: {
				width: normalizedProps.width,
				aspectRatio: normalizedProps.aspectRatio,
				...passthroughProps.style
			}
		});
	});
	/**
	* The shopifyLoader function is a simple utility function that takes a src, width,
	* height, and crop and returns a string that can be used as the src for an image.
	* It can be used with the Hydrogen Image component or with the next/image component.
	* (or any others that accept equivalent configuration)
	* @param src - The source URL of the image, e.g. `https://cdn.shopify.com/static/sample-images/garnished.jpeg`
	* @param width - The width of the image, e.g. `100`
	* @param height - The height of the image, e.g. `100`
	* @param crop - The crop of the image, e.g. `center`
	* @returns A Shopify image URL with the correct query parameters, e.g. `https://cdn.shopify.com/static/sample-images/garnished.jpeg?width=100&height=100&crop=center`
	*
	* @example
	* ```
	* shopifyLoader({
	*   src: 'https://cdn.shopify.com/static/sample-images/garnished.jpeg',
	*   width: 100,
	*   height: 100,
	*   crop: 'center',
	* })
	* ```
	*/
	var PLACEHOLDER_DOMAIN = "https://placeholder.shopify.com";
	function shopifyLoader({ src, width, height, crop }) {
		if (!src) return "";
		const url = new URL(src, PLACEHOLDER_DOMAIN);
		if (width) url.searchParams.append("width", Math.round(width).toString());
		if (height) url.searchParams.append("height", Math.round(height).toString());
		if (crop) url.searchParams.append("crop", crop);
		return url.href.replace(PLACEHOLDER_DOMAIN, "");
	}
	/**
	* Checks whether the width and height share the same unit type
	* @param width - The width of the image, e.g. 100% | 10px
	* @param height - The height of the image, e.g. auto | 100px
	* @returns Whether the width and height share the same unit type (boolean)
	*/
	function unitsMatch(width = "100%", height = "auto") {
		return getUnitValueParts(width.toString()).unit === getUnitValueParts(height.toString()).unit;
	}
	/**
	* Given a CSS size, returns the unit and number parts of the value
	* @param value - The CSS size, e.g. 100px
	* @returns The unit and number parts of the value, e.g. \{unit: 'px', number: 100\}
	*/
	function getUnitValueParts(value) {
		const unit = value.replace(/[0-9.]/g, "");
		const number = parseFloat(value.replace(unit, ""));
		return {
			unit: unit === "" ? number === void 0 ? "auto" : "px" : unit,
			number
		};
	}
	/**
	* Given a value, returns the width of the image as an integer in pixels
	* @param value - The width of the image, e.g. 16px | 1rem | 1em | 16
	* @returns The width of the image in pixels, e.g. 16, or undefined if the value is not a fixed unit
	*/
	function getNormalizedFixedUnit(value) {
		if (value === void 0) return;
		const { unit, number } = getUnitValueParts(value.toString());
		switch (unit) {
			case "em": return number * 16;
			case "rem": return number * 16;
			case "px": return number;
			case "": return number;
			default: return;
		}
	}
	/**
	* This function checks whether a width is fixed or not.
	* @param width - The width of the image, e.g. 100 | '100px' | '100em' | '100rem'
	* @returns Whether the width is fixed or not
	*/
	function isFixedWidth(width) {
		return typeof width === "number" || /\d(px|em|rem)$/.test(width);
	}
	/**
	* This function generates a srcSet for Shopify images.
	* @param src - The source URL of the image, e.g. https://cdn.shopify.com/static/sample-images/garnished.jpeg
	* @param sizesArray - An array of objects containing the `width`, `height`, and `crop` of the image, e.g. [\{width: 200, height: 200, crop: 'center'\}, \{width: 400, height: 400, crop: 'center'\}]
	* @param loader - A function that takes a Shopify image URL and returns a Shopify image URL with the correct query parameters
	* @returns A srcSet for Shopify images, e.g. 'https://cdn.shopify.com/static/sample-images/garnished.jpeg?width=200&height=200&crop=center 200w, https://cdn.shopify.com/static/sample-images/garnished.jpeg?width=400&height=400&crop=center 400w'
	*/
	function generateSrcSet(src, sizesArray, loader = shopifyLoader) {
		if (!src) return "";
		if (sizesArray?.length === 0 || !sizesArray) return src;
		return sizesArray.map((size, i) => `${loader({
			src,
			width: size.width,
			height: size.height,
			crop: size.crop
		})} ${sizesArray.length === 3 ? `${i + 1}x` : `${size.width ?? 0}w`}`).join(`, `);
	}
	/**
	* This function generates an array of sizes for Shopify images, for both fixed and responsive images.
	* @param width - The CSS width of the image
	* @param intervals - The number of intervals to generate
	* @param startingWidth - The starting width of the image
	* @param incrementSize - The size of each interval
	* @returns An array of widths
	*/
	function generateImageWidths(width = "100%", intervals, startingWidth, incrementSize) {
		const responsive = Array.from({ length: intervals }, (_, i) => i * incrementSize + startingWidth);
		const fixed = Array.from({ length: 3 }, (_, i) => (i + 1) * (getNormalizedFixedUnit(width) ?? 0));
		return isFixedWidth(width) ? fixed : responsive;
	}
	/**
	* Simple utility function to convert an aspect ratio CSS string to a decimal, currently only supports values like `1/1`, not `0.5`, or `auto`
	* @param aspectRatio - The aspect ratio of the image, e.g. `1/1`
	* @returns The aspect ratio as a number, e.g. `0.5`
	*
	* {@link https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio}
	*/
	function parseAspectRatio(aspectRatio) {
		if (!aspectRatio) return;
		const [width, height] = aspectRatio.split("/");
		return 1 / (Number(width) / Number(height));
	}
	function generateSizes(imageWidths, aspectRatio, crop = "center", sourceDimensions) {
		if (!imageWidths) return;
		return imageWidths.map((width) => {
			return {
				width,
				height: aspectRatio ? width * (parseAspectRatio(aspectRatio) ?? 1) : void 0,
				crop
			};
		}).filter(({ width, height }) => {
			if (sourceDimensions?.width && width > sourceDimensions.width) return false;
			if (sourceDimensions?.height && height && height > sourceDimensions.height) return false;
			return true;
		});
	}
	//#endregion
	//#region src/load-script.tsx
	var SCRIPTS_LOADED = {};
	function loadScript(src, options) {
		const isScriptLoaded = SCRIPTS_LOADED[src];
		if (isScriptLoaded) return isScriptLoaded;
		const promise = new Promise((resolve, reject) => {
			const script = document.createElement("script");
			if (options?.module) script.type = "module";
			else script.type = "text/javascript";
			script.src = src;
			script.onload = () => {
				resolve(true);
			};
			script.onerror = () => {
				reject(false);
			};
			if (options?.in === "head") document.head.appendChild(script);
			else document.body.appendChild(script);
			const attributes = options?.attributes;
			if (attributes) Object.keys(attributes).forEach((key) => {
				script.setAttribute(key, attributes[key]);
			});
		});
		SCRIPTS_LOADED[src] = promise;
		return promise;
	}
	/**
	* The `useLoadScript` hook loads an external script tag in the browser. It allows React components to lazy-load large third-party dependencies.
	* @publicDocs
	*/
	function useLoadScript(url, options) {
		const [status, setStatus] = (0, react.useState)("loading");
		(0, react.useEffect)(() => {
			loadScript(url, options).then(() => setStatus("done")).catch(() => setStatus("error"));
		}, [url]);
		return status;
	}
	//#endregion
	//#region src/Video.tsx
	/**
	* The `Video` component renders a `video` for the Storefront API's [Video object](https://shopify.dev/api/storefront/reference/products/video).
	* @publicDocs
	*/
	var Video = (0, react.forwardRef)((props, ref) => {
		const { data, previewImageOptions, id = data.id, playsInline = true, controls = true, sourceProps = {}, ...passthroughProps } = props;
		const posterUrl = shopifyLoader({
			src: data.previewImage?.url ?? "",
			...previewImageOptions
		});
		if (!data.sources) throw new Error(`<Video/> requires a 'data.sources' array`);
		return /* @__PURE__ */ React.createElement("video", {
			...passthroughProps,
			id,
			playsInline,
			controls,
			poster: posterUrl,
			ref
		}, data.sources.map((source) => {
			if (!(source?.url && source?.mimeType)) throw new Error(`<Video/> needs 'source.url' and 'source.mimeType'`);
			return /* @__PURE__ */ React.createElement("source", {
				...sourceProps,
				key: source.url,
				src: source.url,
				type: source.mimeType
			});
		}));
	});
	//#endregion
	//#region src/ModelViewer.tsx
	/**
	* The `ModelViewer` component renders a 3D model (with the `model-viewer` custom element) for
	* the Storefront API's [Model3d object](https://shopify.dev/api/storefront/reference/products/model3d).
	*
	* The `model-viewer` custom element is lazily downloaded through a dynamically-injected `<script type="module">` tag when the `<ModelViewer />` component is rendered
	*
	* ModelViewer is using version `1.21.1` of the `@google/model-viewer` library.
	* @publicDocs
	*/
	function ModelViewer(props) {
		const [modelViewer, setModelViewer] = (0, react.useState)(void 0);
		const callbackRef = (0, react.useCallback)((node) => {
			setModelViewer(node);
		}, []);
		const { data, children, className, ...passthroughProps } = props;
		const modelViewerLoadedStatus = useLoadScript("https://unpkg.com/@google/model-viewer@v1.12.1/dist/model-viewer.min.js", { module: true });
		(0, react.useEffect)(() => {
			const hydrogenEventListener = {
				error: passthroughProps.onError,
				load: passthroughProps.onLoad,
				preload: passthroughProps.onPreload,
				"model-visibility": passthroughProps.onModelVisibility,
				progress: passthroughProps.onProgress,
				"ar-status": passthroughProps.onArStatus,
				"ar-tracking": passthroughProps.onArTracking,
				"quick-look-button-tapped": passthroughProps.onQuickLookButtonTapped,
				"camera-change": passthroughProps.onCameraChange,
				"environment-change": passthroughProps.onEnvironmentChange,
				play: passthroughProps.onPlay,
				pause: passthroughProps.onPause,
				"scene-graph-ready": passthroughProps.onSceneGraphReady
			};
			if (!modelViewer) return;
			Object.entries(hydrogenEventListener).forEach(([eventName, callbackFunc]) => {
				if (callbackFunc) modelViewer.addEventListener(eventName, callbackFunc);
			});
			return () => {
				if (modelViewer == null) return;
				Object.entries(hydrogenEventListener).forEach(([eventName, callbackFunc]) => {
					if (callbackFunc) modelViewer.removeEventListener(eventName, callbackFunc);
				});
			};
		}, [
			modelViewer,
			passthroughProps.onArStatus,
			passthroughProps.onArTracking,
			passthroughProps.onCameraChange,
			passthroughProps.onEnvironmentChange,
			passthroughProps.onError,
			passthroughProps.onLoad,
			passthroughProps.onModelVisibility,
			passthroughProps.onPause,
			passthroughProps.onPlay,
			passthroughProps.onPreload,
			passthroughProps.onProgress,
			passthroughProps.onQuickLookButtonTapped,
			passthroughProps.onSceneGraphReady
		]);
		if (modelViewerLoadedStatus !== "done") return null;
		if (!data.sources?.[0]?.url) throw new Error(`<ModelViewer/> requires 'data.sources' prop to be an array, with an object that has a property 'url' on it. Rendering 'null'`);
		if (!data.alt) console.warn(`<ModelViewer/> requires the 'data.alt' prop for accessibility`);
		return /* @__PURE__ */ React.createElement("model-viewer", {
			ref: callbackRef,
			...passthroughProps,
			class: className,
			id: passthroughProps.id ?? data.id,
			src: data.sources[0].url,
			alt: data.alt ?? null,
			"camera-controls": passthroughProps.cameraControls ?? true,
			poster: (passthroughProps.poster || data.previewImage?.url) ?? null,
			autoplay: passthroughProps.autoplay ?? true,
			loading: passthroughProps.loading,
			reveal: passthroughProps.reveal,
			ar: passthroughProps.ar,
			"ar-modes": passthroughProps.arModes,
			"ar-scale": passthroughProps.arScale,
			"ar-placement": passthroughProps.arPlacement,
			"ios-src": passthroughProps.iosSrc,
			"touch-action": passthroughProps.touchAction,
			"disable-zoom": passthroughProps.disableZoom,
			"orbit-sensitivity": passthroughProps.orbitSensitivity,
			"auto-rotate": passthroughProps.autoRotate,
			"auto-rotate-delay": passthroughProps.autoRotateDelay,
			"rotation-per-second": passthroughProps.rotationPerSecond,
			"interaction-policy": passthroughProps.interactionPolicy,
			"interaction-prompt": passthroughProps.interactionPrompt,
			"interaction-prompt-style": passthroughProps.interactionPromptStyle,
			"interaction-prompt-threshold": passthroughProps.interactionPromptThreshold,
			"camera-orbit": passthroughProps.cameraOrbit,
			"camera-target": passthroughProps.cameraTarget,
			"field-of-view": passthroughProps.fieldOfView,
			"max-camera-orbit": passthroughProps.maxCameraOrbit,
			"min-camera-orbit": passthroughProps.minCameraOrbit,
			"max-field-of-view": passthroughProps.maxFieldOfView,
			"min-field-of-view": passthroughProps.minFieldOfView,
			bounds: passthroughProps.bounds,
			"interpolation-decay": passthroughProps.interpolationDecay ?? 100,
			"skybox-image": passthroughProps.skyboxImage,
			"environment-image": passthroughProps.environmentImage,
			exposure: passthroughProps.exposure,
			"shadow-intensity": passthroughProps.shadowIntensity ?? 0,
			"shadow-softness": passthroughProps.shadowSoftness ?? 0,
			"animation-name": passthroughProps.animationName,
			"animation-crossfade-duration": passthroughProps.animationCrossfadeDuration,
			"variant-name": passthroughProps.variantName,
			orientation: passthroughProps.orientation,
			scale: passthroughProps.scale
		}, children);
	}
	//#endregion
	//#region src/MediaFile.tsx
	/**
	* The `MediaFile` component renders the media for the Storefront API's
	* [Media object](https://shopify.dev/api/storefront/reference/products/media). It renders an `Image`, a
	* `Video`, an `ExternalVideo`, or a `ModelViewer` depending on the `__typename` of the `data` prop.
	* @publicDocs
	*/
	function MediaFile({ data, mediaOptions, ...passthroughProps }) {
		switch (data.__typename) {
			case "MediaImage":
				if (!data.image) throw new Error(`<MediaFile/>: 'data.image' does not exist for __typename of 'MediaImage'; rendering 'null' by default.`);
				return /* @__PURE__ */ React.createElement(Image, {
					...passthroughProps,
					...mediaOptions?.image,
					data: data.image
				});
			case "Video": return /* @__PURE__ */ React.createElement(Video, {
				...passthroughProps,
				...mediaOptions?.video,
				data
			});
			case "ExternalVideo": return /* @__PURE__ */ React.createElement(ExternalVideo, {
				...passthroughProps,
				...mediaOptions?.externalVideo,
				data
			});
			case "Model3d": return /* @__PURE__ */ React.createElement(ModelViewer, {
				...passthroughProps,
				...mediaOptions?.modelViewer,
				data
			});
			default: throw new Error(`<MediaFile /> requires the '__typename' property to exist on the 'data' prop in order to render the matching sub-component for this type of media.`);
		}
	}
	//#endregion
	//#region src/parse-metafield.ts
	/**
	* A function that uses `metafield.type` to parse the Metafield's `value` or `reference` or `references` (depending on the `metafield.type`) and places the result in `metafield.parsedValue`
	*
	* TypeScript developers can use the type `ParsedMetafields` from this package to get the returned object's type correct. For example:
	*
	* ```
	* parseMetafield<ParsedMetafields['boolean']>({type: 'boolean', value: 'false'}
	* ```
	* @publicDocs
	*/
	function parseMetafield(metafield) {
		if (!metafield.type) throw new Error(`parseMetafield(): The 'type' field is required in order to parse the Metafield.`);
		switch (metafield.type) {
			case "boolean": return {
				...metafield,
				parsedValue: metafield.value === "true"
			};
			case "article_reference":
			case "collection_reference":
			case "file_reference":
			case "page_reference":
			case "product_reference":
			case "variant_reference": return {
				...metafield,
				parsedValue: metafield.reference
			};
			case "color":
			case "multi_line_text_field":
			case "single_line_text_field":
			case "url": return {
				...metafield,
				parsedValue: metafield.value
			};
			case "money": {
				let parsedValue = null;
				try {
					const parsed = parseJSON(metafield.value ?? "");
					if (parsed && typeof parsed === "object" && "currency_code" in parsed) {
						const moneyData = parsed;
						parsedValue = {
							amount: moneyData.amount,
							currencyCode: moneyData.currency_code
						};
					} else if (parsed && typeof parsed === "object" && "currencyCode" in parsed) parsedValue = parsed;
					else parsedValue = parsed;
				} catch (err) {
					throw new Error(`parseMetafield(): attempted to JSON.parse the 'metafield.value' property, but failed.`);
				}
				return {
					...metafield,
					parsedValue
				};
			}
			case "dimension":
			case "json":
			case "rating":
			case "volume":
			case "weight":
			case "rich_text_field":
			case "list.color":
			case "list.dimension":
			case "list.number_integer":
			case "list.number_decimal":
			case "list.rating":
			case "list.single_line_text_field":
			case "list.url":
			case "list.volume":
			case "list.weight": {
				let parsedValue = null;
				try {
					parsedValue = parseJSON(metafield.value ?? "");
				} catch (err) {
					throw new Error(`parseMetafield(): attempted to JSON.parse the 'metafield.value' property, but failed.`);
				}
				return {
					...metafield,
					parsedValue
				};
			}
			case "date":
			case "date_time": return {
				...metafield,
				parsedValue: new Date(metafield.value ?? "")
			};
			case "list.date":
			case "list.date_time": {
				const jsonParseValue = parseJSON(metafield?.value ?? "");
				return {
					...metafield,
					parsedValue: jsonParseValue.map((dateString) => new Date(dateString))
				};
			}
			case "number_decimal":
			case "number_integer": return {
				...metafield,
				parsedValue: Number(metafield.value)
			};
			case "list.article_reference":
			case "list.collection_reference":
			case "list.file_reference":
			case "list.page_reference":
			case "list.product_reference":
			case "list.variant_reference": return {
				...metafield,
				parsedValue: flattenConnection(metafield.references ?? void 0)
			};
			default: {
				const typeNotFoundError = `parseMetafield(): the 'metafield.type' you passed in is not supported. Your type: "${metafield.type}". If you believe this is an error, please open an issue on GitHub.`;
				throw new Error(typeNotFoundError);
			}
		}
	}
	/**
	* Parses a JSON string while preventing prototype injection attacks.
	*/
	function parseJSON(json) {
		if (String(json).includes("__proto__")) return JSON.parse(json, (k, v) => {
			if (k !== "__proto__") return v;
		});
		return JSON.parse(json);
	}
	//#endregion
	//#region src/ProductPrice.tsx
	/**
	* The `ProductPrice` component renders a `Money` component with the product
	* [`priceRange`](https://shopify.dev/api/storefront/reference/products/productpricerange)'s `maxVariantPrice` or `minVariantPrice`, for either the regular price or compare at price range.
	* @publicDocs
	*/
	function ProductPrice(props) {
		const { priceType = "regular", variantId, valueType = "min", data: product, ...passthroughProps } = props;
		if (product == null) throw new Error(`<ProductPrice/> requires a product as the 'data' prop`);
		let price;
		let measurement;
		const variant = variantId ? flattenConnection(product?.variants ?? {}).find((variant) => variant?.id === variantId) ?? null : null;
		const variantPriceProperty = valueType === "max" ? "maxVariantPrice" : "minVariantPrice";
		if (priceType === "compareAt") {
			if (variantId && variant) price = variant.compareAtPrice;
			else price = product?.compareAtPriceRange?.[variantPriceProperty];
			let priceAsNumber;
			if (variantId && variant) priceAsNumber = parseFloat(variant.price?.amount ?? "0");
			else priceAsNumber = parseFloat(product?.priceRange?.[variantPriceProperty]?.amount ?? "0");
			const compareAtPriceAsNumber = parseFloat(price?.amount ?? "0");
			if (priceAsNumber >= compareAtPriceAsNumber) return null;
		} else if (variantId && variant) {
			price = variant.price;
			if (valueType === "unit") {
				price = variant.unitPrice;
				measurement = variant.unitPriceMeasurement;
			}
		} else if (valueType === "max") price = product.priceRange?.maxVariantPrice;
		else price = product.priceRange?.minVariantPrice;
		if (!price) return null;
		if (measurement) return /* @__PURE__ */ React.createElement(Money, {
			...passthroughProps,
			data: price,
			measurement
		});
		return /* @__PURE__ */ React.createElement(Money, {
			...passthroughProps,
			data: price
		});
	}
	//#endregion
	//#region src/RichText.components.tsx
	var RichTextComponents = {
		root: Root,
		heading: Heading,
		paragraph: Paragraph,
		text: Text,
		link: RichTextLink,
		list: List,
		"list-item": ListItem
	};
	function Root({ node }) {
		return /* @__PURE__ */ React.createElement("div", null, node.children);
	}
	function Heading({ node }) {
		return (0, react.createElement)(`h${node.level ?? "1"}`, null, node.children);
	}
	function Paragraph({ node }) {
		return /* @__PURE__ */ React.createElement("p", null, node.children);
	}
	function Text({ node }) {
		if (node.bold && node.italic) return /* @__PURE__ */ React.createElement("em", null, /* @__PURE__ */ React.createElement("strong", null, node.value));
		if (node.bold) return /* @__PURE__ */ React.createElement("strong", null, node.value);
		if (node.italic) return /* @__PURE__ */ React.createElement("em", null, node.value);
		return node.value;
	}
	function RichTextLink({ node }) {
		return /* @__PURE__ */ React.createElement("a", {
			href: node.url,
			title: node.title,
			target: node.target
		}, node.children);
	}
	function List({ node }) {
		const List = node.listType === "unordered" ? "ul" : "ol";
		return /* @__PURE__ */ React.createElement(List, null, node.children);
	}
	function ListItem({ node }) {
		return /* @__PURE__ */ React.createElement("li", null, node.children);
	}
	//#endregion
	//#region src/RichText.tsx
	/** @publicDocs */
	function RichText({ as, data, plain, components, ...passthroughProps }) {
		try {
			const Wrapper = as ?? "div";
			const parsedData = (0, react.useMemo)(() => JSON.parse(data), [data]);
			return /* @__PURE__ */ React.createElement(Wrapper, passthroughProps, plain ? richTextToString(parsedData) : serializeRichTextASTNode(components, parsedData));
		} catch (e) {
			throw new Error("[h2:error:RichText] Parsing error. Make sure to pass a JSON string of rich text metafield", { cause: e });
		}
	}
	function serializeRichTextASTNode(components = {}, node, index = 0) {
		let children;
		if ("children" in node) children = node.children.map((child, childIndex) => serializeRichTextASTNode(components, child, childIndex));
		const Component = components[node.type === "list-item" ? "listItem" : node.type] ?? RichTextComponents[node.type];
		switch (node.type) {
			case "root": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "root",
					children
				}
			});
			case "heading": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "heading",
					level: node.level,
					children
				}
			});
			case "paragraph": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "paragraph",
					children
				}
			});
			case "text": {
				const elements = (node.value ?? "").split("\n").flatMap((value, subindex) => {
					const key = `${index}-${value}-${subindex}`;
					const textElement = (0, react.createElement)(Component, {
						key,
						node: {
							type: "text",
							italic: node.italic,
							bold: node.bold,
							value
						}
					});
					return subindex === 0 ? textElement : [(0, react.createElement)("br", { key: `${key}-br` }), textElement];
				});
				return elements.length > 1 ? (0, react.createElement)(react.Fragment, { key: index }, elements) : elements[0];
			}
			case "link": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "link",
					url: node.url,
					title: node.title,
					target: node.target,
					children
				}
			});
			case "list": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "list",
					listType: node.listType,
					children
				}
			});
			case "list-item": return (0, react.createElement)(Component, {
				key: index,
				node: {
					type: "list-item",
					children
				}
			});
		}
	}
	function richTextToString(node, result = []) {
		switch (node.type) {
			case "root":
				node.children.forEach((child) => richTextToString(child, result));
				break;
			case "heading":
			case "paragraph":
				node.children.forEach((child) => richTextToString(child, result));
				result.push(" ");
				break;
			case "text":
				result.push(node.value || "");
				break;
			case "link":
				node.children.forEach((child) => richTextToString(child, result));
				break;
			case "list":
				node.children.forEach((item) => {
					if (item.children) item.children.forEach((child) => richTextToString(child, result));
					result.push(" ");
				});
				break;
			default: throw new Error(`Unknown node encountered ${node.type}`);
		}
		return result.join("").trim();
	}
	//#endregion
	//#region src/ShopPayButton.tsx
	var SHOPJS_URL = "https://cdn.shopify.com/shopifycloud/shop-js/v1.0/client.js";
	function isChannel(channel) {
		return channel === "headless" || channel === "hydrogen";
	}
	/**
	* The `ShopPayButton` component renders a button that redirects to the Shop Pay checkout.
	* It renders a [`<shop-pay-button>`](https://shopify.dev/custom-storefronts/tools/web-components) custom element, for which it will lazy-load the source code automatically.
	* It relies on the `<ShopProvider>` context provider.
	* @publicDocs
	*/
	function ShopPayButton({ channel, variantIds, className, variantIdsAndQuantities, width, storeDomain: _storeDomain }) {
		const shop = useShop();
		const storeDomain = _storeDomain || shop?.storeDomain;
		const shopPayLoadedStatus = useLoadScript(SHOPJS_URL);
		let ids = [];
		let channelAttribution;
		if (!storeDomain || storeDomain === defaultShopifyContext.storeDomain) throw new Error(MissingStoreDomainErrorMessage);
		if (variantIds && variantIdsAndQuantities) throw new Error(DoublePropsErrorMessage);
		if (!variantIds && !variantIdsAndQuantities) throw new Error(MissingPropsErrorMessage);
		if (channel) if (isChannel(channel)) channelAttribution = channel;
		else throw new Error(InvalidChannelErrorMessage);
		if (variantIds) ids = variantIds.reduce((prev, curr) => {
			const bareId = parseGid(curr).id;
			if (bareId) prev.push(bareId);
			return prev;
		}, []);
		else if (variantIdsAndQuantities) ids = variantIdsAndQuantities.reduce((prev, curr) => {
			const bareId = parseGid(curr?.id).id;
			if (bareId) prev.push(`${bareId}:${curr?.quantity ?? 1}`);
			return prev;
		}, []);
		else throw new Error(MissingPropsErrorMessage);
		if (ids.length === 0) throw new Error(InvalidPropsErrorMessage);
		const style = width ? { "--shop-pay-button-width": width } : void 0;
		return /* @__PURE__ */ React.createElement("div", {
			className,
			style
		}, shopPayLoadedStatus === "done" && /* @__PURE__ */ React.createElement("shop-pay-button", {
			...channelAttribution ? { channel: channelAttribution } : {},
			"store-url": storeDomain,
			variants: ids.join(",")
		}));
	}
	var MissingStoreDomainErrorMessage = "You must pass a \"storeDomain\" prop to the \"ShopPayButton\" component, or wrap it in a \"ShopifyProvider\" component.";
	var InvalidPropsErrorMessage = `You must pass in "variantIds" in the form of ["gid://shopify/ProductVariant/1"]`;
	var MissingPropsErrorMessage = `You must pass in either "variantIds" or "variantIdsAndQuantities" to ShopPayButton`;
	var DoublePropsErrorMessage = `You must provide either a variantIds or variantIdsAndQuantities prop, but not both in the ShopPayButton component`;
	var InvalidChannelErrorMessage = `Invalid channel attribution value. Must be either "headless" or "hydrogen"`;
	//#endregion
	//#region src/useSelectedOptionInUrlParam.tsx
	/** @publicDocs */
	function useSelectedOptionInUrlParam(selectedOptions) {
		(0, react.useEffect)(() => {
			const optionsSearchParams = new URLSearchParams(mapSelectedProductOptionToObject(selectedOptions || []));
			const currentSearchParams = new URLSearchParams(window.location.search);
			const combinedSearchParams = new URLSearchParams({
				...Object.fromEntries(currentSearchParams),
				...Object.fromEntries(optionsSearchParams)
			});
			if (combinedSearchParams.size > 0) window.history.replaceState({}, "", `${window.location.pathname}?${combinedSearchParams.toString()}`);
		}, [JSON.stringify(selectedOptions)]);
		return null;
	}
	//#endregion
	//#region ../../node_modules/.pnpm/worktop@0.7.3/node_modules/worktop/cookie/index.mjs
	function l(a, r, e = {}) {
		let t = a + "=" + encodeURIComponent(r);
		return e.expires && (t += "; Expires=" + new Date(e.expires).toUTCString()), e.maxage != null && e.maxage >= 0 && (t += "; Max-Age=" + (e.maxage | 0)), e.domain && (t += "; Domain=" + e.domain), e.path && (t += "; Path=" + e.path), e.samesite && (t += "; SameSite=" + e.samesite), (e.secure || e.samesite === "None") && (t += "; Secure"), e.httponly && (t += "; HttpOnly"), t;
	}
	//#endregion
	//#region src/useShopifyCookies.tsx
	var longTermLength = 3600 * 24 * 360 * 1;
	var shortTermLength = 1800;
	/**
	* Sets the `shopify_y` and `shopify_s` cookies in the browser based on user consent
	* for backward compatibility support.
	*
	* If `fetchTrackingValues` is true, it makes a request to Storefront API
	* to fetch or refresh Shopiy analytics and marketing cookies and tracking values.
	* Generally speaking, this should only be needed if you're not using Hydrogen's
	* built-in analytics components and hooks that already handle this automatically.
	* For example, set it to `true` if you are using `hydrogen-react` only with
	* a different framework and still need to make a same-domain request to
	* Storefront API to set cookies.
	*
	* If `ignoreDeprecatedCookies` is true, it skips setting the deprecated cookies entirely.
	* Useful when you only want to use the newer tracking values and not rely on the deprecated ones.
	*
	* @returns `true` when cookies are set and ready.
	* @publicDocs
	*/
	function useShopifyCookies(options) {
		const { hasUserConsent, domain = "", checkoutDomain = "", storefrontAccessToken, fetchTrackingValues, ignoreDeprecatedCookies = false } = options || {};
		const coreCookiesReady = useCoreShopifyCookies({
			storefrontAccessToken,
			fetchTrackingValues,
			checkoutDomain
		});
		(0, react.useEffect)(() => {
			if (ignoreDeprecatedCookies || !coreCookiesReady) return;
			/**
			* Setting cookie with domain
			*
			* If no domain is provided, the cookie will be set for the current host.
			* For Shopify, we need to ensure this domain is set with a leading dot.
			*/
			let currentDomain = domain || window.location.host;
			if (checkoutDomain) {
				const checkoutDomainParts = checkoutDomain.split(".").reverse();
				const currentDomainParts = currentDomain.split(".").reverse();
				const sameDomainParts = [];
				checkoutDomainParts.forEach((part, index) => {
					if (part === currentDomainParts[index]) sameDomainParts.push(part);
				});
				currentDomain = sameDomainParts.reverse().join(".");
			}
			if (/^localhost/.test(currentDomain)) currentDomain = "";
			const domainWithLeadingDot = currentDomain ? /^\./.test(currentDomain) ? currentDomain : `.${currentDomain}` : "";
			/**
			* Set user and session cookies and refresh the expiry time
			*/
			if (hasUserConsent) {
				const trackingValues = getTrackingValues();
				if ((trackingValues.uniqueToken || trackingValues.visitToken || "").startsWith("00000000-")) return;
				setCookie(SHOPIFY_Y, trackingValues.uniqueToken || buildUUID(), longTermLength, domainWithLeadingDot);
				setCookie(SHOPIFY_S, trackingValues.visitToken || buildUUID(), shortTermLength, domainWithLeadingDot);
			} else {
				setCookie(SHOPIFY_Y, "", 0, domainWithLeadingDot);
				setCookie(SHOPIFY_S, "", 0, domainWithLeadingDot);
			}
		}, [
			coreCookiesReady,
			hasUserConsent,
			domain,
			checkoutDomain,
			ignoreDeprecatedCookies
		]);
		return coreCookiesReady;
	}
	function setCookie(name, value, maxage, domain) {
		document.cookie = l(name, value, {
			maxage,
			domain,
			samesite: "Lax",
			path: "/"
		});
	}
	async function fetchTrackingValuesFromBrowser(storefrontAccessToken, storefrontApiDomain = "") {
		const { uniqueToken, visitToken } = getTrackingValues();
		const response = await fetch(`${storefrontApiDomain.replace(/\/+$/, "")}/api/unstable/graphql.json`, {
			method: "POST",
			headers: {
				"Content-Type": "application/json",
				...storefrontAccessToken && { "X-Shopify-Storefront-Access-Token": storefrontAccessToken },
				...visitToken || uniqueToken ? {
					[SHOPIFY_VISIT_TOKEN_HEADER]: visitToken,
					[SHOPIFY_UNIQUE_TOKEN_HEADER]: uniqueToken
				} : void 0
			},
			body: JSON.stringify({ query: "query ensureCookies { consentManagement { cookies(visitorConsent:{}) { cookieDomain } } }" })
		});
		if (!response.ok) throw new Error(`Failed to fetch consent from browser: ${response.status} ${response.statusText}`);
		await response.json();
		getTrackingValues();
	}
	/**
	* Gets http-only cookies from Storefront API via same-origin fetch request.
	* Falls back to checkout domain if provided to at least obtain the tracking
	* values via server-timing headers.
	*/
	function useCoreShopifyCookies({ checkoutDomain, storefrontAccessToken, fetchTrackingValues = false }) {
		const [cookiesReady, setCookiesReady] = (0, react.useState)(!fetchTrackingValues);
		const hasFetchedTrackingValues = (0, react.useRef)(false);
		(0, react.useEffect)(() => {
			if (!fetchTrackingValues) {
				setCookiesReady(true);
				return;
			}
			if (hasFetchedTrackingValues.current) return;
			hasFetchedTrackingValues.current = true;
			fetchTrackingValuesFromBrowser(storefrontAccessToken).catch((error) => checkoutDomain ? fetchTrackingValuesFromBrowser(storefrontAccessToken, checkoutDomain) : Promise.reject(error)).catch((error) => {
				console.warn("[h2:warn:useShopifyCookies] Failed to fetch tracking values from browser: " + (error instanceof Error ? error.message : String(error)));
			}).finally(() => {
				setCookiesReady(true);
			});
		}, [
			checkoutDomain,
			fetchTrackingValues,
			storefrontAccessToken
		]);
		return cookiesReady;
	}
	//#endregion
	exports.AddToCartButton = AddToCartButton;
	exports.AnalyticsEventName = AnalyticsEventName;
	exports.AnalyticsPageType = AnalyticsPageType;
	exports.BuyNowButton = BuyNowButton;
	exports.CartCheckoutButton = CartCheckoutButton;
	exports.CartCost = CartCost;
	exports.CartLineProvider = CartLineProvider;
	exports.CartLineQuantity = CartLineQuantity;
	exports.CartLineQuantityAdjustButton = CartLineQuantityAdjustButton;
	exports.CartProvider = CartProvider;
	exports.ExternalVideo = ExternalVideo;
	exports.IMAGE_FRAGMENT = IMAGE_FRAGMENT;
	exports.Image = Image;
	exports.MediaFile = MediaFile;
	exports.ModelViewer = ModelViewer;
	exports.Money = Money;
	exports.ProductPrice = ProductPrice;
	exports.ProductProvider = ProductProvider;
	exports.RichText = RichText;
	exports.SHOPIFY_S = SHOPIFY_S;
	exports.SHOPIFY_STOREFRONT_ID_HEADER = SHOPIFY_STOREFRONT_ID_HEADER;
	exports.SHOPIFY_STOREFRONT_S_HEADER = SHOPIFY_STOREFRONT_S_HEADER;
	exports.SHOPIFY_STOREFRONT_Y_HEADER = SHOPIFY_STOREFRONT_Y_HEADER;
	exports.SHOPIFY_UNIQUE_TOKEN_HEADER = SHOPIFY_UNIQUE_TOKEN_HEADER;
	exports.SHOPIFY_VISIT_TOKEN_HEADER = SHOPIFY_VISIT_TOKEN_HEADER;
	exports.SHOPIFY_Y = SHOPIFY_Y;
	exports.ShopPayButton = ShopPayButton;
	exports.ShopifyProvider = ShopifyProvider;
	exports.ShopifySalesChannel = ShopifySalesChannel;
	exports.Video = Video;
	exports.createStorefrontClient = createStorefrontClient;
	exports.customerAccountApiCustomScalars = customerAccountApiCustomScalars;
	exports.decodeEncodedVariant = decodeEncodedVariant;
	exports.flattenConnection = flattenConnection;
	exports.getAdjacentAndFirstAvailableVariants = getAdjacentAndFirstAvailableVariants;
	exports.getClientBrowserParameters = getClientBrowserParameters;
	exports.getProductOptions = getProductOptions;
	exports.getShopifyCookies = getShopifyCookies;
	exports.getTrackingValues = getTrackingValues;
	exports.isOptionValueCombinationInEncodedVariant = isOptionValueCombinationInEncodedVariant;
	exports.mapSelectedProductOptionToObject = mapSelectedProductOptionToObject;
	exports.parseGid = parseGid;
	exports.parseMetafield = parseMetafield;
	exports.sendShopifyAnalytics = sendShopifyAnalytics;
	exports.storefrontApiCustomScalars = storefrontApiCustomScalars;
	exports.useCart = useCart;
	exports.useCartLine = useCartLine;
	exports.useLoadScript = useLoadScript;
	exports.useMoney = useMoney;
	exports.useProduct = useProduct;
	exports.useSelectedOptionInUrlParam = useSelectedOptionInUrlParam;
	exports.useShop = useShop;
	exports.useShopifyCookies = useShopifyCookies;
});
