// @shopify/hydrogen-react@2026.4.2/dist/browser-dev/analytics-utils.mjs
//#region src/analytics-utils.ts
/**
* Builds a Shopify Monorail event from a Shopify Monorail payload and a schema ID.
* @param payload - The Monorail payload
* @param schemaId - The schema ID to use
* @returns The formatted payload
**/
function schemaWrapper(schemaId, payload) {
	return {
		schema_id: schemaId,
		payload,
		metadata: { event_created_at_ms: Date.now() }
	};
}
/**
* Parses global id (gid) and returns the resource type and id.
* @see https://shopify.dev/api/usage/gids
* @param gid - A shopify GID (string)
*
* @example
* ```ts
* const {id, resource} = parseGid('gid://shopify/Order/123')
* // => id = "123", resource = 'Order'
*
*  * const {id, resource} = parseGid('gid://shopify/Cart/abc123')
* // => id = "abc123", resource = 'Cart'
* ```
* @publicDocs
**/
function parseGid(gid) {
	const defaultReturn = {
		id: "",
		resource: null,
		resourceId: null,
		search: "",
		searchParams: new URLSearchParams(),
		hash: ""
	};
	if (typeof gid !== "string") return defaultReturn;
	try {
		const { search, searchParams, pathname, hash } = new URL(gid);
		const pathnameParts = pathname.split("/");
		const lastPathnamePart = pathnameParts[pathnameParts.length - 1];
		const resourcePart = pathnameParts[pathnameParts.length - 2];
		if (!lastPathnamePart || !resourcePart) return defaultReturn;
		return {
			id: `${lastPathnamePart}${search}${hash}` || "",
			resource: resourcePart ?? null,
			resourceId: lastPathnamePart || null,
			search,
			searchParams,
			hash
		};
	} catch {
		return defaultReturn;
	}
}
/**
* Filters properties from an object and returns a new object with only the properties that have a truthy value.
* @param keyValuePairs - An object of key-value pairs
* @param formattedData - An object which will hold the truthy values
* @returns The formatted object
**/
function addDataIf(keyValuePairs, formattedData) {
	if (typeof keyValuePairs !== "object") return {};
	Object.entries(keyValuePairs).forEach(([key, value]) => {
		if (value) formattedData[key] = value;
	});
	return formattedData;
}
/**
* Utility that errors if a function is called on the server.
* @param fnName - The name of the function
* @returns A boolean
**/
function errorIfServer(fnName) {
	if (typeof document === "undefined") {
		console.error(`${fnName} should only be used within the useEffect callback or event handlers`);
		return true;
	}
	return false;
}
//#endregion
export { addDataIf, errorIfServer, parseGid, schemaWrapper };
