// @shopify/hydrogen-react@2026.4.2/dist/browser-dev/cart-queries.mjs
//#region src/cart-queries.ts
function getInContextVariables(includeVisitorConsent) {
	const base = `$country: CountryCode = ZZ
    $language: LanguageCode`;
	return includeVisitorConsent ? `${base}
    $visitorConsent: VisitorConsent` : base;
}
function getInContextDirective(includeVisitorConsent) {
	return includeVisitorConsent ? `@inContext(
    country: $country
    language: $language
    visitorConsent: $visitorConsent
  )` : `@inContext(
    country: $country
    language: $language
  )`;
}
var CartLineAdd = (cartFragment, options = {}) => `
  mutation CartLineAdd(
    $cartId: ID!
    $lines: [CartLineInput!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesAdd(cartId: $cartId, lines: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartCreate = (cartFragment, options = {}) => `
  mutation CartCreate(
    $input: CartInput!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartCreate(input: $input) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartLineRemove = (cartFragment, options = {}) => `
  mutation CartLineRemove(
    $cartId: ID!
    $lines: [ID!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesRemove(cartId: $cartId, lineIds: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartLineUpdate = (cartFragment, options = {}) => `
  mutation CartLineUpdate(
    $cartId: ID!
    $lines: [CartLineUpdateInput!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartLinesUpdate(cartId: $cartId, lines: $lines) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartNoteUpdate = (cartFragment, options = {}) => `
  mutation CartNoteUpdate(
    $cartId: ID!
    $note: String!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartNoteUpdate(cartId: $cartId, note: $note) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartBuyerIdentityUpdate = (cartFragment, options = {}) => `
  mutation CartBuyerIdentityUpdate(
    $cartId: ID!
    $buyerIdentity: CartBuyerIdentityInput!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartBuyerIdentityUpdate(cartId: $cartId, buyerIdentity: $buyerIdentity) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartAttributesUpdate = (cartFragment, options = {}) => `
  mutation CartAttributesUpdate(
    $attributes: [AttributeInput!]!
    $cartId: ID!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartAttributesUpdate(attributes: $attributes, cartId: $cartId) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartDiscountCodesUpdate = (cartFragment, options = {}) => `
  mutation CartDiscountCodesUpdate(
    $cartId: ID!
    $discountCodes: [String!]!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cartDiscountCodesUpdate(cartId: $cartId, discountCodes: $discountCodes) {
      cart {
        ...CartFragment
      }
    }
  }

  ${cartFragment}
`;
var CartQuery = (cartFragment, options = {}) => `
  query CartQuery(
    $id: ID!
    $numCartLines: Int = 250
    ${getInContextVariables(options.includeVisitorConsent ?? false)}
  )
  ${getInContextDirective(options.includeVisitorConsent ?? false)} {
    cart(id: $id) {
      ...CartFragment
    }
  }

  ${cartFragment}
`;
var defaultCartFragment = `
  fragment CartFragment on Cart {
    id
    checkoutUrl
    totalQuantity
    buyerIdentity {
      countryCode
      customer {
        id
        email
        firstName
        lastName
        displayName
      }
      email
      phone
    }
    lines(first: $numCartLines) {
      edges {
        node {
          id
          quantity
          attributes {
            key
            value
          }
          cost {
            totalAmount {
              amount
              currencyCode
            }
            compareAtAmountPerQuantity {
              amount
              currencyCode
            }
          }
          merchandise {
            ... on ProductVariant {
              id
              availableForSale
              compareAtPrice {
                ...MoneyFragment
              }
              price {
                ...MoneyFragment
              }
              requiresShipping
              title
              image {
                ...ImageFragment
              }
              product {
                handle
                title
                id
              }
              selectedOptions {
                name
                value
              }
            }
          }
        }
      }
    }
    cost {
      subtotalAmount {
        ...MoneyFragment
      }
      totalAmount {
        ...MoneyFragment
      }
      totalDutyAmount {
        ...MoneyFragment
      }
      totalTaxAmount {
        ...MoneyFragment
      }
    }
    note
    attributes {
      key
      value
    }
    discountCodes {
      code
      applicable
    }
  }

  fragment MoneyFragment on MoneyV2 {
    currencyCode
    amount
  }
  fragment ImageFragment on Image {
    id
    url
    altText
    width
    height
  }
`;
//#endregion
export { CartAttributesUpdate, CartBuyerIdentityUpdate, CartCreate, CartDiscountCodesUpdate, CartLineAdd, CartLineRemove, CartLineUpdate, CartNoteUpdate, CartQuery, defaultCartFragment };
