// @shopify/hydrogen-react@2026.4.2/dist/browser-dev/storefront-client.mjs
import { SFAPI_VERSION } from "./storefront-api-constants.mjs";
//#region src/storefront-client.ts
var MOCK_SHOP_DOMAIN = "mock.shop";
var isMockShop = (domain) => domain.includes(MOCK_SHOP_DOMAIN);
/**
* The `createStorefrontClient()` function creates helpers that enable you to quickly query the Shopify Storefront API.
*
* When used on the server, it is recommended to use the `privateStorefrontToken` prop. When used on the client, it is recommended to use the `publicStorefrontToken` prop.
* @publicDocs
*/
function createStorefrontClient({ storeDomain, privateStorefrontToken, publicStorefrontToken, storefrontApiVersion = SFAPI_VERSION, contentType }) {
	if (!storeDomain) {
		storeDomain = MOCK_SHOP_DOMAIN;
		warnOnce(`storeDomain missing, defaulting to ${MOCK_SHOP_DOMAIN}`, "info");
	}
	if (storefrontApiVersion !== "2026-04") warnOnce(`The Storefront API version that you're using is different than the version this build of Hydrogen React is targeting.\nYou may run into unexpected errors if these versions don't match. Received version: "${storefrontApiVersion}"; expected version "${SFAPI_VERSION}"`);
	if (!privateStorefrontToken && !globalThis.document && !isMockShop(storeDomain)) warnOnce("Using a private storefront token is recommended for server environments.\nRefer to the authentication https://shopify.dev/api/storefront#authentication documentation for more details.");
	if (privateStorefrontToken && globalThis.document) warnOnce("You are attempting to use a private token in an environment where it can be easily accessed by anyone.\nThis is a security risk; please use the public token and the `publicStorefrontToken` prop");
	const getShopifyDomain = (overrideProps) => {
		const domain = overrideProps?.storeDomain ?? storeDomain;
		return domain.includes("://") ? domain : `https://${domain}`;
	};
	return {
		getShopifyDomain,
		getStorefrontApiUrl(overrideProps) {
			const domain = getShopifyDomain(overrideProps);
			return `${domain + (domain.endsWith("/") ? "api" : "/api")}/${overrideProps?.storefrontApiVersion ?? storefrontApiVersion}/graphql.json`;
		},
		getPrivateTokenHeaders(overrideProps) {
			if (!privateStorefrontToken && !overrideProps?.privateStorefrontToken && !isMockShop(storeDomain)) throw new Error(H2_PREFIX_ERROR + "You did not pass in a `privateStorefrontToken` while using `createStorefrontClient()` or `getPrivateTokenHeaders()`");
			if (!overrideProps?.buyerIp) warnOnce("It is recommended to pass in the `buyerIp` property which improves analytics and data in the admin.");
			return {
				"content-type": (overrideProps?.contentType ?? contentType) === "graphql" ? "application/graphql" : "application/json",
				"X-SDK-Variant": "hydrogen-react",
				"X-SDK-Variant-Source": "react",
				"X-SDK-Version": storefrontApiVersion,
				"Shopify-Storefront-Private-Token": overrideProps?.privateStorefrontToken ?? privateStorefrontToken ?? "",
				...overrideProps?.buyerIp ? { "Shopify-Storefront-Buyer-IP": overrideProps.buyerIp } : {}
			};
		},
		getPublicTokenHeaders(overrideProps) {
			if (!publicStorefrontToken && !overrideProps?.publicStorefrontToken && !isMockShop(storeDomain)) throw new Error(H2_PREFIX_ERROR + "You did not pass in a `publicStorefrontToken` while using `createStorefrontClient()` or `getPublicTokenHeaders()`");
			return getPublicTokenHeadersRaw(overrideProps?.contentType ?? contentType ?? "json", storefrontApiVersion, overrideProps?.publicStorefrontToken ?? publicStorefrontToken ?? "");
		}
	};
}
function getPublicTokenHeadersRaw(contentType, storefrontApiVersion, accessToken) {
	return {
		"content-type": contentType === "graphql" ? "application/graphql" : "application/json",
		"X-SDK-Variant": "hydrogen-react",
		"X-SDK-Variant-Source": "react",
		"X-SDK-Version": storefrontApiVersion,
		"X-Shopify-Storefront-Access-Token": accessToken
	};
}
var warnings = /* @__PURE__ */ new Set();
var H2_PREFIX_ERROR = "[h2:error:createStorefrontClient] ";
var warnOnce = (string, type = "warn") => {
	if (!warnings.has(string)) {
		console[type](`[h2:${type}:createStorefrontClient] ` + string);
		warnings.add(string);
	}
};
//#endregion
export { createStorefrontClient, getPublicTokenHeadersRaw };
