// @shopify/hydrogen-react@2026.4.2/dist/browser-prod/ShopifyProvider.mjs
import { SFAPI_VERSION } from "./storefront-api-constants.mjs";
import { getPublicTokenHeadersRaw } from "./storefront-client.mjs";
import { createContext, useContext, useMemo } from "react";
import { jsx } from "react/jsx-runtime";
//#region src/ShopifyProvider.tsx
var defaultShopifyContext = {
	storeDomain: "test",
	storefrontToken: "abc123",
	storefrontApiVersion: SFAPI_VERSION,
	countryIsoCode: "US",
	languageIsoCode: "EN",
	getStorefrontApiUrl() {
		return "";
	},
	getPublicTokenHeaders() {
		return {};
	},
	getShopifyDomain() {
		return "";
	}
};
var ShopifyContext = createContext(defaultShopifyContext);
/**
* Hydrogen server sets this server timing key when the SFAPI proxy is enabled.
* Read it automatically in the browser for apps using frontend cart in full-stack Hydrogen,
* but don't export this utility yet since we don't want to make this a public convention.
*/
function isSfapiProxyEnabled() {
	if (typeof window === "undefined") return false;
	try {
		return !!(window.performance?.getEntriesByType?.("navigation")[0])?.serverTiming?.some((entry) => entry.name === "_sfapi_proxy");
	} catch (e) {
		return false;
	}
}
/**
* The `<ShopifyProvider/>` component enables use of the `useShop()` hook. The component should wrap your app.
* @publicDocs
*/
function ShopifyProvider({ children, ...shopifyConfig }) {
	if (!shopifyConfig.countryIsoCode || !shopifyConfig.languageIsoCode || !shopifyConfig.storeDomain || !shopifyConfig.storefrontToken || !shopifyConfig.storefrontApiVersion) throw new Error(`Please provide the necessary props to '<ShopifyProvider/>'`);
	if (shopifyConfig.storefrontApiVersion !== "2026-04") console.warn(`<ShopifyProvider/>: This version of Hydrogen React is built for Shopify's Storefront API version ${SFAPI_VERSION}, but it looks like you're using version ${shopifyConfig.storefrontApiVersion}. There may be issues or bugs if you use a mismatched version of Hydrogen React and the Storefront API.`);
	const finalConfig = useMemo(() => {
		const sameDomainForStorefrontApi = shopifyConfig.sameDomainForStorefrontApi ?? isSfapiProxyEnabled();
		function getShopifyDomain(overrideProps) {
			const domain = overrideProps?.storeDomain ?? shopifyConfig.storeDomain;
			return domain.includes("://") ? domain : `https://${domain}`;
		}
		return {
			...shopifyConfig,
			sameDomainForStorefrontApi,
			getPublicTokenHeaders(overrideProps) {
				return getPublicTokenHeadersRaw(overrideProps.contentType, shopifyConfig.storefrontApiVersion, overrideProps.storefrontToken ?? shopifyConfig.storefrontToken);
			},
			getShopifyDomain,
			getStorefrontApiUrl(overrideProps) {
				const finalDomainUrl = sameDomainForStorefrontApi && typeof window !== "undefined" ? window.location.origin : getShopifyDomain({ storeDomain: overrideProps?.storeDomain ?? shopifyConfig.storeDomain });
				return `${finalDomainUrl}${finalDomainUrl.endsWith("/") ? "" : "/"}api/${overrideProps?.storefrontApiVersion ?? shopifyConfig.storefrontApiVersion}/graphql.json`;
			}
		};
	}, [shopifyConfig]);
	return /* @__PURE__ */ jsx(ShopifyContext.Provider, {
		value: finalConfig,
		children
	});
}
/**
* Provides access to the `shopifyConfig` prop of `<ShopifyProvider/>`. Must be a descendent of `<ShopifyProvider/>`.
* @publicDocs
*/
function useShop() {
	const shopContext = useContext(ShopifyContext);
	if (!shopContext) throw new Error(`'useShop()' must be a descendent of <ShopifyProvider/>`);
	return shopContext;
}
//#endregion
export { ShopifyProvider, defaultShopifyContext, useShop };
