// @shopify/hydrogen-react@2026.4.2/dist/browser-prod/useMoney.mjs
import { useShop } from "./ShopifyProvider.mjs";
import { useMemo } from "react";
//#region src/useMoney.tsx
/**
* The `useMoney` hook takes a [MoneyV2 object from the Storefront API](https://shopify.dev/docs/api/storefront/2026-04/objects/MoneyV2)
* or a [MoneyV2 object from the Customer Account API](https://shopify.dev/docs/api/customer/2026-04/objects/moneyv2) and returns a
* default-formatted string of the amount with the correct currency indicator, along with some of the parts provided by
* [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat).
* Uses `locale` from `ShopifyProvider`
* &nbsp;
* @see {@link https://shopify.dev/api/hydrogen/hooks/usemoney}
* @example initialize the money object
* ```ts
* const money = useMoney({
*   amount: '100.00',
*   currencyCode: 'USD'
* })
* ```
* &nbsp;
*
* @example basic usage, outputs: $100.00
* ```ts
* money.localizedString
* ```
* &nbsp;
*
* @example without currency, outputs: 100.00
* ```ts
* money.amount
* ```
* &nbsp;
*
* @example without trailing zeros, outputs: $100
* ```ts
* money.withoutTrailingZeros
* ```
* &nbsp;
*
* @example currency name, outputs: US dollars
* ```ts
* money.currencyCode
* ```
* &nbsp;
*
* @example currency symbol, outputs: $
* ```ts
* money.currencySymbol
* ```
* &nbsp;
*
* @example without currency and without trailing zeros, outputs: 100
* ```ts
* money.withoutTrailingZerosAndCurrency
* ```
* @publicDocs
*/
function useMoney(money) {
	const { countryIsoCode, languageIsoCode } = useShop();
	const locale = languageIsoCode.includes("_") ? languageIsoCode.replace("_", "-") : `${languageIsoCode}-${countryIsoCode}`;
	if (!locale) throw new Error(`useMoney(): Unable to get 'locale' from 'useShop()', which means that 'locale' was not passed to '<ShopifyProvider/>'. 'locale' is required for 'useMoney()' to work`);
	const amount = parseFloat(money.amount);
	let isCurrencySupported = true;
	try {
		new Intl.NumberFormat(locale, {
			style: "currency",
			currency: money.currencyCode
		});
	} catch (e) {
		if (e instanceof RangeError && e.message.includes("currency")) isCurrencySupported = false;
	}
	const { defaultFormatter, nameFormatter, narrowSymbolFormatter, withoutTrailingZerosFormatter, withoutCurrencyFormatter, withoutTrailingZerosOrCurrencyFormatter } = useMemo(() => {
		const options = isCurrencySupported ? {
			style: "currency",
			currency: money.currencyCode
		} : {
			style: "decimal",
			minimumFractionDigits: 2,
			maximumFractionDigits: 2
		};
		return {
			defaultFormatter: getLazyFormatter(locale, options),
			nameFormatter: getLazyFormatter(locale, {
				...options,
				currencyDisplay: "name"
			}),
			narrowSymbolFormatter: getLazyFormatter(locale, {
				...options,
				currencyDisplay: "narrowSymbol"
			}),
			withoutTrailingZerosFormatter: getLazyFormatter(locale, {
				...options,
				minimumFractionDigits: 0,
				maximumFractionDigits: 0
			}),
			withoutCurrencyFormatter: getLazyFormatter(locale, {
				minimumFractionDigits: 2,
				maximumFractionDigits: 2
			}),
			withoutTrailingZerosOrCurrencyFormatter: getLazyFormatter(locale, {
				minimumFractionDigits: 0,
				maximumFractionDigits: 0
			})
		};
	}, [
		money.currencyCode,
		locale,
		isCurrencySupported
	]);
	const isPartCurrency = (part) => part.type === "currency";
	const lazyFormatters = useMemo(() => ({
		original: () => money,
		currencyCode: () => money.currencyCode,
		localizedString: () => {
			const formatted = defaultFormatter().format(amount);
			return isCurrencySupported ? formatted : `${formatted} ${money.currencyCode}`;
		},
		parts: () => {
			const parts = defaultFormatter().formatToParts(amount);
			if (!isCurrencySupported) parts.push({
				type: "literal",
				value: " "
			}, {
				type: "currency",
				value: money.currencyCode
			});
			return parts;
		},
		withoutTrailingZeros: () => {
			const formatted = amount % 1 === 0 ? withoutTrailingZerosFormatter().format(amount) : defaultFormatter().format(amount);
			return isCurrencySupported ? formatted : `${formatted} ${money.currencyCode}`;
		},
		withoutTrailingZerosAndCurrency: () => amount % 1 === 0 ? withoutTrailingZerosOrCurrencyFormatter().format(amount) : withoutCurrencyFormatter().format(amount),
		currencyName: () => nameFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? money.currencyCode,
		currencySymbol: () => defaultFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? money.currencyCode,
		currencyNarrowSymbol: () => narrowSymbolFormatter().formatToParts(amount).find(isPartCurrency)?.value ?? "",
		amount: () => defaultFormatter().formatToParts(amount).filter((part) => [
			"decimal",
			"fraction",
			"group",
			"integer",
			"literal"
		].includes(part.type)).map((part) => part.value).join("")
	}), [
		money,
		amount,
		isCurrencySupported,
		nameFormatter,
		defaultFormatter,
		narrowSymbolFormatter,
		withoutCurrencyFormatter,
		withoutTrailingZerosFormatter,
		withoutTrailingZerosOrCurrencyFormatter
	]);
	return useMemo(() => new Proxy(lazyFormatters, { get: (target, key) => Reflect.get(target, key)?.call(null) }), [lazyFormatters]);
}
var formatterCache = /* @__PURE__ */ new Map();
function getLazyFormatter(locale, options) {
	const key = JSON.stringify([locale, options]);
	return function() {
		let formatter = formatterCache.get(key);
		if (!formatter) {
			try {
				formatter = new Intl.NumberFormat(locale, options);
			} catch (error) {
				if (error instanceof RangeError && error.message.includes("currency")) {
					const fallbackOptions = { ...options };
					delete fallbackOptions.currency;
					delete fallbackOptions.currencyDisplay;
					delete fallbackOptions.currencySign;
					formatter = new Intl.NumberFormat(locale, fallbackOptions);
				} else throw error;
			}
			formatterCache.set(key, formatter);
		}
		return formatter;
	};
}
//#endregion
export { useMoney };
