// @shopify/hydrogen-react@2026.4.2/dist/browser-prod/useShopifyCookies.mjs
import { SHOPIFY_S, SHOPIFY_Y } from "./cart-constants.mjs";
import { SHOPIFY_UNIQUE_TOKEN_HEADER, SHOPIFY_VISIT_TOKEN_HEADER, getTrackingValues } from "./tracking-utils.mjs";
import { buildUUID } from "./cookies-utils.mjs";
import { useEffect, useRef, useState } from "react";
import { stringify } from "worktop/cookie";
//#region src/useShopifyCookies.tsx
var longTermLength = 3600 * 24 * 360 * 1;
var shortTermLength = 1800;
/**
* Sets the `shopify_y` and `shopify_s` cookies in the browser based on user consent
* for backward compatibility support.
*
* If `fetchTrackingValues` is true, it makes a request to Storefront API
* to fetch or refresh Shopiy analytics and marketing cookies and tracking values.
* Generally speaking, this should only be needed if you're not using Hydrogen's
* built-in analytics components and hooks that already handle this automatically.
* For example, set it to `true` if you are using `hydrogen-react` only with
* a different framework and still need to make a same-domain request to
* Storefront API to set cookies.
*
* If `ignoreDeprecatedCookies` is true, it skips setting the deprecated cookies entirely.
* Useful when you only want to use the newer tracking values and not rely on the deprecated ones.
*
* @returns `true` when cookies are set and ready.
* @publicDocs
*/
function useShopifyCookies(options) {
	const { hasUserConsent, domain = "", checkoutDomain = "", storefrontAccessToken, fetchTrackingValues, ignoreDeprecatedCookies = false } = options || {};
	const coreCookiesReady = useCoreShopifyCookies({
		storefrontAccessToken,
		fetchTrackingValues,
		checkoutDomain
	});
	useEffect(() => {
		if (ignoreDeprecatedCookies || !coreCookiesReady) return;
		/**
		* Setting cookie with domain
		*
		* If no domain is provided, the cookie will be set for the current host.
		* For Shopify, we need to ensure this domain is set with a leading dot.
		*/
		let currentDomain = domain || window.location.host;
		if (checkoutDomain) {
			const checkoutDomainParts = checkoutDomain.split(".").reverse();
			const currentDomainParts = currentDomain.split(".").reverse();
			const sameDomainParts = [];
			checkoutDomainParts.forEach((part, index) => {
				if (part === currentDomainParts[index]) sameDomainParts.push(part);
			});
			currentDomain = sameDomainParts.reverse().join(".");
		}
		if (/^localhost/.test(currentDomain)) currentDomain = "";
		const domainWithLeadingDot = currentDomain ? /^\./.test(currentDomain) ? currentDomain : `.${currentDomain}` : "";
		/**
		* Set user and session cookies and refresh the expiry time
		*/
		if (hasUserConsent) {
			const trackingValues = getTrackingValues();
			if ((trackingValues.uniqueToken || trackingValues.visitToken || "").startsWith("00000000-")) return;
			setCookie(SHOPIFY_Y, trackingValues.uniqueToken || buildUUID(), longTermLength, domainWithLeadingDot);
			setCookie(SHOPIFY_S, trackingValues.visitToken || buildUUID(), shortTermLength, domainWithLeadingDot);
		} else {
			setCookie(SHOPIFY_Y, "", 0, domainWithLeadingDot);
			setCookie(SHOPIFY_S, "", 0, domainWithLeadingDot);
		}
	}, [
		coreCookiesReady,
		hasUserConsent,
		domain,
		checkoutDomain,
		ignoreDeprecatedCookies
	]);
	return coreCookiesReady;
}
function setCookie(name, value, maxage, domain) {
	document.cookie = stringify(name, value, {
		maxage,
		domain,
		samesite: "Lax",
		path: "/"
	});
}
async function fetchTrackingValuesFromBrowser(storefrontAccessToken, storefrontApiDomain = "") {
	const { uniqueToken, visitToken } = getTrackingValues();
	const response = await fetch(`${storefrontApiDomain.replace(/\/+$/, "")}/api/unstable/graphql.json`, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
			...storefrontAccessToken && { "X-Shopify-Storefront-Access-Token": storefrontAccessToken },
			...visitToken || uniqueToken ? {
				[SHOPIFY_VISIT_TOKEN_HEADER]: visitToken,
				[SHOPIFY_UNIQUE_TOKEN_HEADER]: uniqueToken
			} : void 0
		},
		body: JSON.stringify({ query: "query ensureCookies { consentManagement { cookies(visitorConsent:{}) { cookieDomain } } }" })
	});
	if (!response.ok) throw new Error(`Failed to fetch consent from browser: ${response.status} ${response.statusText}`);
	await response.json();
	getTrackingValues();
}
/**
* Gets http-only cookies from Storefront API via same-origin fetch request.
* Falls back to checkout domain if provided to at least obtain the tracking
* values via server-timing headers.
*/
function useCoreShopifyCookies({ checkoutDomain, storefrontAccessToken, fetchTrackingValues = false }) {
	const [cookiesReady, setCookiesReady] = useState(!fetchTrackingValues);
	const hasFetchedTrackingValues = useRef(false);
	useEffect(() => {
		if (!fetchTrackingValues) {
			setCookiesReady(true);
			return;
		}
		if (hasFetchedTrackingValues.current) return;
		hasFetchedTrackingValues.current = true;
		fetchTrackingValuesFromBrowser(storefrontAccessToken).catch((error) => checkoutDomain ? fetchTrackingValuesFromBrowser(storefrontAccessToken, checkoutDomain) : Promise.reject(error)).catch((error) => {
			console.warn("[h2:warn:useShopifyCookies] Failed to fetch tracking values from browser: " + (error instanceof Error ? error.message : String(error)));
		}).finally(() => {
			setCookiesReady(true);
		});
	}, [
		checkoutDomain,
		fetchTrackingValues,
		storefrontAccessToken
	]);
	return cookiesReady;
}
//#endregion
export { useShopifyCookies };
