// @shopify/hydrogen-react@2026.4.2/dist/node-prod/AddToCartButton.js
const require_runtime = require("./_virtual/_rolldown/runtime.js");
const require_CartProvider = require("./CartProvider.js");
const require_ProductProvider = require("./ProductProvider.js");
const require_BaseButton = require("./BaseButton.js");
let react = require("react");
react = require_runtime.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/AddToCartButton.tsx
/**
* The `AddToCartButton` component renders a button that adds an item to the cart when pressed.
* It must be a descendent of the `CartProvider` component.
* @publicDocs
*/
function AddToCartButton(props) {
	const [addingItem, setAddingItem] = (0, react.useState)(false);
	const { variantId: explicitVariantId, quantity = 1, attributes, sellingPlanId, onClick, children, accessibleAddingToCartLabel, parent, ...passthroughProps } = props;
	const { status, linesAdd } = require_CartProvider.useCart();
	const { selectedVariant } = require_ProductProvider.useProduct();
	const variantId = explicitVariantId ?? selectedVariant?.id ?? "";
	const disabled = explicitVariantId === null || variantId === "" || selectedVariant === null || addingItem || passthroughProps.disabled;
	(0, react.useEffect)(() => {
		if (addingItem && status === "idle") setAddingItem(false);
	}, [status, addingItem]);
	const handleAddItem = (0, react.useCallback)(() => {
		setAddingItem(true);
		linesAdd([{
			quantity,
			merchandiseId: variantId || "",
			attributes,
			parent,
			sellingPlanId
		}]);
	}, [
		linesAdd,
		quantity,
		variantId,
		attributes,
		sellingPlanId,
		parent
	]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_BaseButton.BaseButton, {
		...passthroughProps,
		disabled,
		onClick,
		defaultOnClick: handleAddItem,
		children
	}), accessibleAddingToCartLabel ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", {
		style: {
			position: "absolute",
			width: "1px",
			height: "1px",
			padding: "0",
			margin: "-1px",
			overflow: "hidden",
			clip: "rect(0, 0, 0, 0)",
			whiteSpace: "nowrap",
			borderWidth: "0"
		},
		role: "alert",
		"aria-live": "assertive",
		children: addingItem ? accessibleAddingToCartLabel : null
	}) : null] });
}
//#endregion
exports.AddToCartButton = AddToCartButton;
