// @shopify/hydrogen-react@2026.4.2/dist/node-prod/AddToCartButton.mjs
import { useCart } from "./CartProvider.mjs";
import { useProduct } from "./ProductProvider.mjs";
import { BaseButton } from "./BaseButton.mjs";
import { useCallback, useEffect, useState } from "react";
import { Fragment as Fragment$1, jsx, jsxs } from "react/jsx-runtime";
//#region src/AddToCartButton.tsx
/**
* The `AddToCartButton` component renders a button that adds an item to the cart when pressed.
* It must be a descendent of the `CartProvider` component.
* @publicDocs
*/
function AddToCartButton(props) {
	const [addingItem, setAddingItem] = useState(false);
	const { variantId: explicitVariantId, quantity = 1, attributes, sellingPlanId, onClick, children, accessibleAddingToCartLabel, parent, ...passthroughProps } = props;
	const { status, linesAdd } = useCart();
	const { selectedVariant } = useProduct();
	const variantId = explicitVariantId ?? selectedVariant?.id ?? "";
	const disabled = explicitVariantId === null || variantId === "" || selectedVariant === null || addingItem || passthroughProps.disabled;
	useEffect(() => {
		if (addingItem && status === "idle") setAddingItem(false);
	}, [status, addingItem]);
	const handleAddItem = useCallback(() => {
		setAddingItem(true);
		linesAdd([{
			quantity,
			merchandiseId: variantId || "",
			attributes,
			parent,
			sellingPlanId
		}]);
	}, [
		linesAdd,
		quantity,
		variantId,
		attributes,
		sellingPlanId,
		parent
	]);
	return /* @__PURE__ */ jsxs(Fragment$1, { children: [/* @__PURE__ */ jsx(BaseButton, {
		...passthroughProps,
		disabled,
		onClick,
		defaultOnClick: handleAddItem,
		children
	}), accessibleAddingToCartLabel ? /* @__PURE__ */ jsx("p", {
		style: {
			position: "absolute",
			width: "1px",
			height: "1px",
			padding: "0",
			margin: "-1px",
			overflow: "hidden",
			clip: "rect(0, 0, 0, 0)",
			whiteSpace: "nowrap",
			borderWidth: "0"
		},
		role: "alert",
		"aria-live": "assertive",
		children: addingItem ? accessibleAddingToCartLabel : null
	}) : null] });
}
//#endregion
export { AddToCartButton };
