// @shopify/hydrogen-react@2026.4.2/dist/node-prod/analytics-schema-trekkie-storefront-page-view.js
const require_analytics_constants = require("./analytics-constants.js");
const require_analytics_utils = require("./analytics-utils.js");
const require_cookies_utils = require("./cookies-utils.js");
//#region src/analytics-schema-trekkie-storefront-page-view.ts
var SCHEMA_ID = "trekkie_storefront_page_view/1.4";
var OXYGEN_DOMAIN = "myshopify.dev";
function pageView(payload) {
	const pageViewPayload = payload;
	const { id, resource } = require_analytics_utils.parseGid(pageViewPayload.resourceId);
	const resourceType = resource ? resource.toLowerCase() : void 0;
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		pageType: pageViewPayload.pageType,
		customerId: parseInt(require_analytics_utils.parseGid(pageViewPayload.customerId).id || "0"),
		resourceType,
		resourceId: parseInt(id)
	}, formatPayload(pageViewPayload)))];
}
function formatPayload(payload) {
	return {
		appClientId: payload.shopifySalesChannel ? require_analytics_constants.ShopifyAppId[payload.shopifySalesChannel] : require_analytics_constants.ShopifyAppId.headless,
		isMerchantRequest: isMerchantRequest(payload.url),
		hydrogenSubchannelId: payload.storefrontId || payload.hydrogenSubchannelId || "0",
		isPersistentCookie: payload.hasUserConsent,
		uniqToken: payload.uniqueToken,
		visitToken: payload.visitToken,
		microSessionId: require_cookies_utils.buildUUID(),
		microSessionCount: 1,
		url: payload.url,
		path: payload.path,
		search: payload.search,
		referrer: payload.referrer,
		title: payload.title,
		shopId: parseInt(require_analytics_utils.parseGid(payload.shopId).id),
		currency: payload.currency,
		contentLanguage: payload.acceptedLanguage || "en"
	};
}
function isMerchantRequest(url) {
	if (typeof url !== "string") return false;
	const hostname = new URL(url).hostname;
	if (hostname.indexOf(OXYGEN_DOMAIN) !== -1 || hostname === "localhost") return true;
	return false;
}
//#endregion
exports.pageView = pageView;
