// @shopify/hydrogen-react@2026.4.2/dist/node-prod/cart-hooks.js
require("./_virtual/_rolldown/runtime.js");
const require_ShopifyProvider = require("./ShopifyProvider.js");
const require_cart_constants = require("./cart-constants.js");
const require_tracking_utils = require("./tracking-utils.js");
let react = require("react");
//#region src/cart-hooks.tsx
function useCartFetch() {
	const { storefrontId, getPublicTokenHeaders, getStorefrontApiUrl, sameDomainForStorefrontApi } = require_ShopifyProvider.useShop();
	return (0, react.useCallback)(({ query, variables }) => {
		const headers = getPublicTokenHeaders({ contentType: "json" });
		if (storefrontId) headers[require_cart_constants.SHOPIFY_STOREFRONT_ID_HEADER] = storefrontId;
		if (!sameDomainForStorefrontApi) {
			const { uniqueToken, visitToken } = require_tracking_utils.getTrackingValues();
			if (uniqueToken) {
				headers[require_cart_constants.SHOPIFY_STOREFRONT_Y_HEADER] = uniqueToken;
				headers[require_tracking_utils.SHOPIFY_UNIQUE_TOKEN_HEADER] = uniqueToken;
			}
			if (visitToken) {
				headers[require_cart_constants.SHOPIFY_STOREFRONT_S_HEADER] = visitToken;
				headers[require_tracking_utils.SHOPIFY_VISIT_TOKEN_HEADER] = visitToken;
			}
		}
		return fetch(getStorefrontApiUrl(), {
			method: "POST",
			headers,
			body: JSON.stringify({
				query: query.toString(),
				variables
			})
		}).then((res) => res.json()).catch((error) => {
			return {
				data: void 0,
				errors: error?.toString()
			};
		});
	}, [
		getPublicTokenHeaders,
		storefrontId,
		getStorefrontApiUrl,
		sameDomainForStorefrontApi
	]);
}
//#endregion
exports.useCartFetch = useCartFetch;
