// @shopify/hydrogen-react@2026.4.2/dist/node-prod/getProductOptions.mjs
import { isOptionValueCombinationInEncodedVariant } from "./optionValueDecoder.mjs";
//#region src/getProductOptions.ts
/**
* Creates a mapping of product options to their index for matching encoded values
* For example, a product option of
*  [
*    \{
*      name: 'Color',
*      optionValues: [\{name: 'Red'\}, \{name: 'Blue'\}]
*    \},
*    \{
*      name: 'Size',
*      optionValues: [\{name: 'Small'\}, \{name: 'Medium'\}, \{name: 'Large'\}]
*    \}
*  ]
* Would return
*  \{
*    'Color': \{Red: 0, Blue: 1\},
*    'Size': \{Small: 0, Medium: 1, Large: 2\}
*  \}
*/
function mapProductOptions(options) {
	return Object.assign({}, ...options.map((option) => {
		return { [option.name]: Object.assign({}, ...option?.optionValues ? option.optionValues.map((value, index) => {
			return { [value.name]: index };
		}) : []) };
	}));
}
/**
* Converts the product option into an Object\<key, value\> for building query params
* For example, a selected product option of
*  [
*    \{
*      name: 'Color',
*      value: 'Red',
*    \},
*    \{
*      name: 'Size',
*      value: 'Medium',
*    \}
*  ]
* Would return
*  \{
*    Color: 'Red',
*    Size: 'Medium',
*  \}
* @publicDocs
*/
function mapSelectedProductOptionToObject(options) {
	return Object.assign({}, ...options.map((key) => {
		return { [key.name]: key.value };
	}));
}
/**
* Returns the JSON stringify result of mapSelectedProductOptionToObject
*/
function mapSelectedProductOptionToObjectAsString(options) {
	return JSON.stringify(mapSelectedProductOptionToObject(options));
}
/**
* Encode the selected product option as a key for mapping to the encoded variants
* For example, a selected product option of
*  [
*    \{
*      name: 'Color',
*      value: 'Red',
*    \},
*    \{
*      name: 'Size',
*      value: 'Medium',
*    \}
*  ]
* Would return
*  JSON.stringify(\{
*    Color: 'Red',
*    Size: 'Medium',
*  \})
*/
function encodeSelectedProductOptionAsKey(selectedOption) {
	if (Array.isArray(selectedOption)) return JSON.stringify(Object.assign({}, ...selectedOption.map((option) => ({ [option.name]: option.value }))));
	else return JSON.stringify(selectedOption);
}
/**
* Build the encoding array for the given selected options. For example, if we have
* the following productOptionMappings:
*
*  \{
*    'Color': \{Red: 0, Blue: 1\},
*    'Size': \{Small: 0, Medium: 1, Large: 2\}
*  \}
*
* A selectedOption of
*
* \{
*    Color: 'Red',
*    Size: 'Medium',
* \}
*
* `buildEncodingArrayFromSelectedOptions` will produce
*
* [0,1]
*
* If in the case where a selected option doesn't exists in the mapping array, for example:
*
* \{
*    Color: 'Red',
*    Fabric: 'Cotton',
*    Size: 'Medium',
* \}
*
* `buildEncodingArrayFromSelectedOptions` will still produce
*
*  [0,1]
*
* This can be caused by when we do not have all the product
* option information for the loading optimistic variant
*/
function buildEncodingArrayFromSelectedOptions(selectedOption, productOptionMappings) {
	return Object.keys(selectedOption).map((key) => {
		return productOptionMappings[key] ? productOptionMappings[key][selectedOption[key]] : null;
	}).filter((code) => code !== null);
}
/**
* Takes an array of product variants and maps them to an object with the encoded selected option values as the key.
* For example, a product variant of
* [
*  \{
*    id: 1,
*    selectedOptions: [
*      \{name: 'Color', value: 'Red'\},
*      \{name: 'Size', value: 'Small'\},
*    ],
*  \},
*  \{
*    id: 2,
*    selectedOptions: [
*      \{name: 'Color', value: 'Red'\},
*      \{name: 'Size', value: 'Medium'\},
*    ],
*  \}
* ]
* Would return
* \{
*    '[0,0]': \{id: 1, selectedOptions: [\{name: 'Color', value: 'Red'\}, \{name: 'Size', value: 'Small'\}]\},
*    '[0,1]': \{id: 2, selectedOptions: [\{name: 'Color', value: 'Red'\}, \{name: 'Size', value: 'Medium'\}]\},
* \}
*/
function mapVariants(variants) {
	return Object.assign({}, ...variants.map((variant) => {
		return { [encodeSelectedProductOptionAsKey(variant.selectedOptions || [])]: variant };
	}));
}
var PRODUCT_INPUTS = [
	"options",
	"selectedOrFirstAvailableVariant",
	"adjacentVariants"
];
var PRODUCT_INPUTS_EXTRA = [
	"handle",
	"encodedVariantExistence",
	"encodedVariantAvailability"
];
function logErrorAndReturnFalse(key) {
	console.error(`[h2:error:getProductOptions] product.${key} is missing. Make sure you query for this field from the Storefront API.`);
	return false;
}
function checkProductParam(product, checkAll = false) {
	let validParam = true;
	const productKeys = Object.keys(product);
	(checkAll ? [...PRODUCT_INPUTS, ...PRODUCT_INPUTS_EXTRA] : PRODUCT_INPUTS).forEach((key) => {
		if (!productKeys.includes(key)) validParam = logErrorAndReturnFalse(key);
	});
	if (product.options) {
		const firstOption = product?.options[0];
		if (checkAll && !firstOption?.name) validParam = logErrorAndReturnFalse("options.name");
		if (product?.options[0]?.optionValues) {
			let firstOptionValues = product.options[0].optionValues[0];
			if (checkAll && !firstOptionValues?.name) validParam = logErrorAndReturnFalse("options.optionValues.name");
			firstOptionValues = product.options[0].optionValues.filter((value) => !!value?.firstSelectableVariant)[0];
			if (firstOptionValues?.firstSelectableVariant) validParam = checkProductVariantParam(firstOptionValues.firstSelectableVariant, "options.optionValues.firstSelectableVariant", validParam, checkAll);
		} else validParam = logErrorAndReturnFalse("options.optionValues");
	}
	if (product.selectedOrFirstAvailableVariant) validParam = checkProductVariantParam(product.selectedOrFirstAvailableVariant, "selectedOrFirstAvailableVariant", validParam, checkAll);
	if (!!product.adjacentVariants && product.adjacentVariants[0]) validParam = checkProductVariantParam(product.adjacentVariants[0], "adjacentVariants", validParam, checkAll);
	return validParam ? product : {};
}
function checkProductVariantParam(variant, key, currentValidParamState, checkAll) {
	let validParam = currentValidParamState;
	if (checkAll && !variant.product?.handle) validParam = logErrorAndReturnFalse(`${key}.product.handle`);
	if (variant.selectedOptions) {
		const firstSelectedOption = variant.selectedOptions[0];
		if (!firstSelectedOption?.name) validParam = logErrorAndReturnFalse(`${key}.selectedOptions.name`);
		if (!firstSelectedOption?.value) validParam = logErrorAndReturnFalse(`${key}.selectedOptions.value`);
	} else validParam = logErrorAndReturnFalse(`${key}.selectedOptions`);
	return validParam;
}
/**
* Finds all the variants provided by adjacentVariants, options.optionValues.firstAvailableVariant,
* and selectedOrFirstAvailableVariant and return them in a single array
* @publicDocs
*/
function getAdjacentAndFirstAvailableVariants(product) {
	const checkedProduct = checkProductParam(product);
	if (!checkedProduct.options) return [];
	const availableVariants = {};
	checkedProduct.options.map((option) => {
		option.optionValues?.map((value) => {
			if (value.firstSelectableVariant) {
				const variantKey = mapSelectedProductOptionToObjectAsString(value.firstSelectableVariant.selectedOptions);
				availableVariants[variantKey] = value.firstSelectableVariant;
			}
		});
	});
	checkedProduct.adjacentVariants.map((variant) => {
		const variantKey = mapSelectedProductOptionToObjectAsString(variant.selectedOptions);
		availableVariants[variantKey] = variant;
	});
	const selectedVariant = checkedProduct.selectedOrFirstAvailableVariant;
	if (selectedVariant) {
		const variantKey = mapSelectedProductOptionToObjectAsString(selectedVariant.selectedOptions);
		availableVariants[variantKey] = selectedVariant;
	}
	return Object.values(availableVariants);
}
/**
* Returns a product options array with its relevant information
* about the variant
* @publicDocs
*/
function getProductOptions(product) {
	const checkedProduct = checkProductParam(product, true);
	if (!checkedProduct.options) return [];
	const { options, selectedOrFirstAvailableVariant: selectedVariant, adjacentVariants, encodedVariantExistence, encodedVariantAvailability, handle: productHandle } = checkedProduct;
	const selectedOptionKeys = selectedVariant?.selectedOptions.map((option) => option.name);
	const filteredOptions = options.filter((option) => {
		return selectedOptionKeys && selectedOptionKeys.indexOf(option.name) >= 0;
	});
	const productOptionMappings = mapProductOptions(options);
	const variants = mapVariants(selectedVariant ? [selectedVariant, ...adjacentVariants] : adjacentVariants);
	const selectedOptions = mapSelectedProductOptionToObject(selectedVariant ? selectedVariant.selectedOptions : []);
	return filteredOptions.map((option, optionIndex) => {
		return {
			...option,
			optionValues: option.optionValues.map((value) => {
				const targetOptionParams = { ...selectedOptions };
				targetOptionParams[option.name] = value.name;
				const targetKey = encodeSelectedProductOptionAsKey(targetOptionParams || []);
				const topDownKey = buildEncodingArrayFromSelectedOptions(targetOptionParams || [], productOptionMappings).slice(0, optionIndex + 1);
				const exists = isOptionValueCombinationInEncodedVariant(topDownKey, encodedVariantExistence || "");
				const available = isOptionValueCombinationInEncodedVariant(topDownKey, encodedVariantAvailability || "");
				const variant = variants[targetKey] || value.firstSelectableVariant;
				let variantOptionParam = {};
				if (variant) variantOptionParam = mapSelectedProductOptionToObject(variant.selectedOptions || []);
				const searchParams = new URLSearchParams(variantOptionParam);
				const handle = variant?.product?.handle || productHandle;
				return {
					...value,
					variant,
					handle,
					variantUriQuery: searchParams.toString(),
					selected: selectedOptions[option.name] === value.name,
					exists,
					available,
					isDifferentProduct: handle !== productHandle
				};
			})
		};
	});
}
//#endregion
export { getAdjacentAndFirstAvailableVariants, getProductOptions, mapSelectedProductOptionToObject };
