// @shopify/hydrogen-react@2026.4.2/dist/node-prod/ProductProvider.js
require("./_virtual/_rolldown/runtime.js");
const require_flatten_connection = require("./flatten-connection.js");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/ProductProvider.tsx
var ProductOptionsContext = (0, react.createContext)(null);
/**
* `<ProductProvider />` is a context provider that enables use of the `useProduct()` hook.
*
* It helps manage selected options and variants for a product.
* @publicDocs
*/
function ProductProvider({ children, data: product, initialVariantId: explicitVariantId }) {
	const variants = (0, react.useMemo)(() => require_flatten_connection.flattenConnection(product.variants ?? {}), [product.variants]);
	if (!isProductVariantArray(variants)) throw new Error(`<ProductProvider/> requires 'product.variants.nodes' or 'product.variants.edges'`);
	const options = (0, react.useMemo)(() => getOptions(variants), [variants]);
	/**
	* Track the selectedVariant within the provider.
	*/
	const [selectedVariant, setSelectedVariant] = (0, react.useState)(() => getVariantBasedOnIdProp(explicitVariantId, variants));
	/**
	* Track the selectedOptions within the provider. If a `initialVariantId`
	* is passed, use that to select initial options.
	*/
	const [selectedOptions, setSelectedOptions] = (0, react.useState)(() => getSelectedOptions(selectedVariant));
	/**
	* When the initialVariantId changes, we need to make sure we
	* update the selected variant and selected options. If not,
	* then the selected variant and options will reference incorrect
	* values.
	*/
	(0, react.useEffect)(() => {
		const newSelectedVariant = getVariantBasedOnIdProp(explicitVariantId, variants);
		setSelectedVariant(newSelectedVariant);
		setSelectedOptions(getSelectedOptions(newSelectedVariant));
	}, [explicitVariantId, variants]);
	/**
	* Allow the developer to select an option.
	*/
	const setSelectedOption = (0, react.useCallback)((name, value) => {
		setSelectedOptions((selectedOptions) => {
			const opts = {
				...selectedOptions,
				[name]: value
			};
			setSelectedVariant(getSelectedVariant(variants, opts));
			return opts;
		});
	}, [setSelectedOptions, variants]);
	const isOptionInStock = (0, react.useCallback)((option, value) => {
		return getSelectedVariant(variants, {
			...selectedOptions,
			[option]: value
		})?.availableForSale ?? true;
	}, [selectedOptions, variants]);
	const sellingPlanGroups = (0, react.useMemo)(() => require_flatten_connection.flattenConnection(product.sellingPlanGroups ?? {}).map((sellingPlanGroup) => ({
		...sellingPlanGroup,
		sellingPlans: require_flatten_connection.flattenConnection(sellingPlanGroup?.sellingPlans ?? {})
	})), [product.sellingPlanGroups]);
	/**
	* Track the selectedSellingPlan within the hook. If `initialSellingPlanId`
	* is passed, use that as an initial value. Look it up from the `selectedVariant`, since
	* that is also a requirement.
	*/
	const [selectedSellingPlan, setSelectedSellingPlan] = (0, react.useState)(void 0);
	const selectedSellingPlanAllocation = (0, react.useMemo)(() => {
		if (!selectedVariant || !selectedSellingPlan) return;
		if (!selectedVariant.sellingPlanAllocations?.nodes && !selectedVariant.sellingPlanAllocations?.edges) throw new Error(`<ProductProvider/>: You must include 'sellingPlanAllocations.nodes' or 'sellingPlanAllocations.edges' in your variants in order to calculate selectedSellingPlanAllocation`);
		return require_flatten_connection.flattenConnection(selectedVariant.sellingPlanAllocations).find((allocation) => allocation?.sellingPlan?.id === selectedSellingPlan.id);
	}, [selectedVariant, selectedSellingPlan]);
	const value = (0, react.useMemo)(() => ({
		product,
		variants,
		variantsConnection: product.variants,
		options,
		selectedVariant,
		setSelectedVariant,
		selectedOptions,
		setSelectedOption,
		setSelectedOptions,
		isOptionInStock,
		selectedSellingPlan,
		setSelectedSellingPlan,
		selectedSellingPlanAllocation,
		sellingPlanGroups,
		sellingPlanGroupsConnection: product.sellingPlanGroups
	}), [
		product,
		isOptionInStock,
		options,
		selectedOptions,
		selectedSellingPlan,
		selectedSellingPlanAllocation,
		selectedVariant,
		sellingPlanGroups,
		setSelectedOption,
		variants
	]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ProductOptionsContext.Provider, {
		value,
		children
	});
}
/**
* Provides access to the context value provided by `<ProductProvider />`. Must be a descendent of `<ProductProvider />`.
* @publicDocs
*/
function useProduct() {
	const context = (0, react.useContext)(ProductOptionsContext);
	if (!context) throw new Error(`'useProduct' must be a child of <ProductProvider />`);
	return context;
}
function getSelectedVariant(variants, choices) {
	/**
	* Ensure the user has selected all the required options, not just some.
	*/
	if (!variants.length || variants?.[0]?.selectedOptions?.length !== Object.keys(choices).length) return;
	return variants?.find((variant) => {
		return Object.entries(choices).every(([name, value]) => {
			return variant?.selectedOptions?.some((option) => option?.name === name && option?.value === value);
		});
	});
}
function getOptions(variants) {
	const map = variants.reduce((memo, variant) => {
		if (!variant.selectedOptions) throw new Error(`'getOptions' requires 'variant.selectedOptions'`);
		variant?.selectedOptions?.forEach((opt) => {
			memo[opt?.name ?? ""] = memo[opt?.name ?? ""] || /* @__PURE__ */ new Set();
			memo[opt?.name ?? ""].add(opt?.value ?? "");
		});
		return memo;
	}, {});
	return Object.keys(map).map((option) => {
		return {
			name: option,
			values: Array.from(map[option])
		};
	});
}
function getVariantBasedOnIdProp(explicitVariantId, variants) {
	if (explicitVariantId) {
		const foundVariant = variants.find((variant) => variant?.id === explicitVariantId);
		if (!foundVariant) console.warn(`<ProductProvider/> received a 'initialVariantId' prop, but could not actually find a variant with that ID`);
		return foundVariant;
	}
	if (explicitVariantId === null) return null;
	if (explicitVariantId === void 0) return variants.find((variant) => variant?.availableForSale) || variants[0];
}
function getSelectedOptions(selectedVariant) {
	return selectedVariant?.selectedOptions ? selectedVariant.selectedOptions.reduce((memo, optionSet) => {
		memo[optionSet?.name ?? ""] = optionSet?.value ?? "";
		return memo;
	}, {}) : {};
}
function isProductVariantArray(maybeVariantArray) {
	if (!maybeVariantArray || !Array.isArray(maybeVariantArray)) return false;
	return true;
}
//#endregion
exports.ProductProvider = ProductProvider;
exports.useProduct = useProduct;
