// @shopify/hydrogen-react@2026.4.2/dist/node-prod/RichText.components.js
require("./_virtual/_rolldown/runtime.js");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/RichText.components.tsx
var RichTextComponents = {
	root: Root,
	heading: Heading,
	paragraph: Paragraph,
	text: Text,
	link: RichTextLink,
	list: List,
	"list-item": ListItem
};
function Root({ node }) {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: node.children });
}
function Heading({ node }) {
	return (0, react.createElement)(`h${node.level ?? "1"}`, null, node.children);
}
function Paragraph({ node }) {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", { children: node.children });
}
function Text({ node }) {
	if (node.bold && node.italic) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("em", { children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("strong", { children: node.value }) });
	if (node.bold) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("strong", { children: node.value });
	if (node.italic) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("em", { children: node.value });
	return node.value;
}
function RichTextLink({ node }) {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("a", {
		href: node.url,
		title: node.title,
		target: node.target,
		children: node.children
	});
}
function List({ node }) {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(node.listType === "unordered" ? "ul" : "ol", { children: node.children });
}
function ListItem({ node }) {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("li", { children: node.children });
}
//#endregion
exports.RichTextComponents = RichTextComponents;
