// @shopify/hydrogen-react@2026.4.2/dist/node-prod/ShopifyProvider.js
require("./_virtual/_rolldown/runtime.js");
const require_storefront_api_constants = require("./storefront-api-constants.js");
const require_storefront_client = require("./storefront-client.js");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/ShopifyProvider.tsx
var defaultShopifyContext = {
	storeDomain: "test",
	storefrontToken: "abc123",
	storefrontApiVersion: require_storefront_api_constants.SFAPI_VERSION,
	countryIsoCode: "US",
	languageIsoCode: "EN",
	getStorefrontApiUrl() {
		return "";
	},
	getPublicTokenHeaders() {
		return {};
	},
	getShopifyDomain() {
		return "";
	}
};
var ShopifyContext = (0, react.createContext)(defaultShopifyContext);
/**
* Hydrogen server sets this server timing key when the SFAPI proxy is enabled.
* Read it automatically in the browser for apps using frontend cart in full-stack Hydrogen,
* but don't export this utility yet since we don't want to make this a public convention.
*/
function isSfapiProxyEnabled() {
	if (typeof window === "undefined") return false;
	try {
		return !!(window.performance?.getEntriesByType?.("navigation")[0])?.serverTiming?.some((entry) => entry.name === "_sfapi_proxy");
	} catch (e) {
		return false;
	}
}
/**
* The `<ShopifyProvider/>` component enables use of the `useShop()` hook. The component should wrap your app.
* @publicDocs
*/
function ShopifyProvider({ children, ...shopifyConfig }) {
	if (!shopifyConfig.countryIsoCode || !shopifyConfig.languageIsoCode || !shopifyConfig.storeDomain || !shopifyConfig.storefrontToken || !shopifyConfig.storefrontApiVersion) throw new Error(`Please provide the necessary props to '<ShopifyProvider/>'`);
	if (shopifyConfig.storefrontApiVersion !== "2026-04") console.warn(`<ShopifyProvider/>: This version of Hydrogen React is built for Shopify's Storefront API version ${require_storefront_api_constants.SFAPI_VERSION}, but it looks like you're using version ${shopifyConfig.storefrontApiVersion}. There may be issues or bugs if you use a mismatched version of Hydrogen React and the Storefront API.`);
	const finalConfig = (0, react.useMemo)(() => {
		const sameDomainForStorefrontApi = shopifyConfig.sameDomainForStorefrontApi ?? isSfapiProxyEnabled();
		function getShopifyDomain(overrideProps) {
			const domain = overrideProps?.storeDomain ?? shopifyConfig.storeDomain;
			return domain.includes("://") ? domain : `https://${domain}`;
		}
		return {
			...shopifyConfig,
			sameDomainForStorefrontApi,
			getPublicTokenHeaders(overrideProps) {
				return require_storefront_client.getPublicTokenHeadersRaw(overrideProps.contentType, shopifyConfig.storefrontApiVersion, overrideProps.storefrontToken ?? shopifyConfig.storefrontToken);
			},
			getShopifyDomain,
			getStorefrontApiUrl(overrideProps) {
				const finalDomainUrl = sameDomainForStorefrontApi && typeof window !== "undefined" ? window.location.origin : getShopifyDomain({ storeDomain: overrideProps?.storeDomain ?? shopifyConfig.storeDomain });
				return `${finalDomainUrl}${finalDomainUrl.endsWith("/") ? "" : "/"}api/${overrideProps?.storefrontApiVersion ?? shopifyConfig.storefrontApiVersion}/graphql.json`;
			}
		};
	}, [shopifyConfig]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ShopifyContext.Provider, {
		value: finalConfig,
		children
	});
}
/**
* Provides access to the `shopifyConfig` prop of `<ShopifyProvider/>`. Must be a descendent of `<ShopifyProvider/>`.
* @publicDocs
*/
function useShop() {
	const shopContext = (0, react.useContext)(ShopifyContext);
	if (!shopContext) throw new Error(`'useShop()' must be a descendent of <ShopifyProvider/>`);
	return shopContext;
}
//#endregion
exports.ShopifyProvider = ShopifyProvider;
exports.defaultShopifyContext = defaultShopifyContext;
exports.useShop = useShop;
