// @shopify/hydrogen-react@2026.4.2/dist/node-prod/useCartActions.js
require("./_virtual/_rolldown/runtime.js");
const require_cart_queries = require("./cart-queries.js");
const require_cart_hooks = require("./cart-hooks.js");
let react = require("react");
//#region src/useCartActions.tsx
/**
* The `useCartActions` hook returns helper graphql functions for Storefront Cart API
*
* See [cart API graphql mutations](https://shopify.dev/api/storefront/2026-04/objects/Cart)
*/
function useCartActions({ numCartLines, cartFragment, countryCode = "US", languageCode = "EN" }) {
	const fetchCart = require_cart_hooks.useCartFetch();
	const cartFetch = (0, react.useCallback)((cartId) => {
		return fetchCart({
			query: require_cart_queries.CartQuery(cartFragment),
			variables: {
				id: cartId,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		fetchCart,
		cartFragment,
		numCartLines,
		countryCode,
		languageCode
	]);
	const cartCreate = (0, react.useCallback)((cart) => {
		return fetchCart({
			query: require_cart_queries.CartCreate(cartFragment),
			variables: {
				input: cart,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const cartLineAdd = (0, react.useCallback)((cartId, lines) => {
		return fetchCart({
			query: require_cart_queries.CartLineAdd(cartFragment),
			variables: {
				cartId,
				lines,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const cartLineUpdate = (0, react.useCallback)((cartId, lines) => {
		return fetchCart({
			query: require_cart_queries.CartLineUpdate(cartFragment),
			variables: {
				cartId,
				lines,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const cartLineRemove = (0, react.useCallback)((cartId, lines) => {
		return fetchCart({
			query: require_cart_queries.CartLineRemove(cartFragment),
			variables: {
				cartId,
				lines,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const noteUpdate = (0, react.useCallback)((cartId, note) => {
		return fetchCart({
			query: require_cart_queries.CartNoteUpdate(cartFragment),
			variables: {
				cartId,
				note,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		fetchCart,
		cartFragment,
		numCartLines,
		countryCode,
		languageCode
	]);
	const buyerIdentityUpdate = (0, react.useCallback)((cartId, buyerIdentity) => {
		return fetchCart({
			query: require_cart_queries.CartBuyerIdentityUpdate(cartFragment),
			variables: {
				cartId,
				buyerIdentity,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const cartAttributesUpdate = (0, react.useCallback)((cartId, attributes) => {
		return fetchCart({
			query: require_cart_queries.CartAttributesUpdate(cartFragment),
			variables: {
				cartId,
				attributes,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	const discountCodesUpdate = (0, react.useCallback)((cartId, discountCodes) => {
		return fetchCart({
			query: require_cart_queries.CartDiscountCodesUpdate(cartFragment),
			variables: {
				cartId,
				discountCodes,
				numCartLines,
				country: countryCode,
				language: languageCode
			}
		});
	}, [
		cartFragment,
		countryCode,
		fetchCart,
		numCartLines,
		languageCode
	]);
	return (0, react.useMemo)(() => ({
		cartFetch,
		cartCreate,
		cartLineAdd,
		cartLineUpdate,
		cartLineRemove,
		noteUpdate,
		buyerIdentityUpdate,
		cartAttributesUpdate,
		discountCodesUpdate,
		cartFragment
	}), [
		cartFetch,
		cartCreate,
		cartLineAdd,
		cartLineUpdate,
		cartLineRemove,
		noteUpdate,
		buyerIdentityUpdate,
		cartAttributesUpdate,
		discountCodesUpdate,
		cartFragment
	]);
}
//#endregion
exports.useCartActions = useCartActions;
