// @shopify/hydrogen-react@2026.4.2/dist/node-dev/analytics-schema-custom-storefront-customer-tracking.js
const require_analytics_constants = require("./analytics-constants.js");
const require_analytics_utils = require("./analytics-utils.js");
const require_cookies_utils = require("./cookies-utils.js");
require("./packages/hydrogen-react/package.js");
//#region src/analytics-schema-custom-storefront-customer-tracking.ts
var SCHEMA_ID = "custom_storefront_customer_tracking/1.2";
var PAGE_RENDERED_EVENT_NAME = "page_rendered";
var COLLECTION_PAGE_RENDERED_EVENT_NAME = "collection_page_rendered";
var PRODUCT_PAGE_RENDERED_EVENT_NAME = "product_page_rendered";
var PRODUCT_ADDED_TO_CART_EVENT_NAME = "product_added_to_cart";
var SEARCH_SUBMITTED_EVENT_NAME = "search_submitted";
function prepareAdditionalPayload(payload) {
	return {
		canonical_url: payload.canonicalUrl || payload.url,
		customer_id: parseInt(require_analytics_utils.parseGid(payload.customerId).id || "0")
	};
}
function pageView(payload) {
	const pageViewPayload = payload;
	const additionalPayload = prepareAdditionalPayload(pageViewPayload);
	const pageType = pageViewPayload.pageType;
	const pageViewEvents = [];
	pageViewEvents.push(require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: PAGE_RENDERED_EVENT_NAME,
		...additionalPayload
	}, formatPayload(pageViewPayload))));
	switch (pageType) {
		case require_analytics_constants.AnalyticsPageType.collection:
			pageViewEvents.push(require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
				event_name: COLLECTION_PAGE_RENDERED_EVENT_NAME,
				...additionalPayload,
				collection_name: pageViewPayload.collectionHandle,
				collection_id: parseInt(require_analytics_utils.parseGid(pageViewPayload.collectionId).id)
			}, formatPayload(pageViewPayload))));
			break;
		case require_analytics_constants.AnalyticsPageType.product:
			pageViewEvents.push(require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
				event_name: PRODUCT_PAGE_RENDERED_EVENT_NAME,
				...additionalPayload,
				products: formatProductPayload(pageViewPayload.products),
				total_value: pageViewPayload.totalValue
			}, formatPayload(pageViewPayload))));
			break;
		case require_analytics_constants.AnalyticsPageType.search:
			pageViewEvents.push(require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
				event_name: SEARCH_SUBMITTED_EVENT_NAME,
				...additionalPayload,
				search_string: pageViewPayload.searchString
			}, formatPayload(pageViewPayload))));
			break;
	}
	return pageViewEvents;
}
function pageView2(payload) {
	const pageViewPayload = payload;
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: PAGE_RENDERED_EVENT_NAME,
		...prepareAdditionalPayload(pageViewPayload)
	}, formatPayload(pageViewPayload)))];
}
function collectionView(payload) {
	const pageViewPayload = payload;
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: COLLECTION_PAGE_RENDERED_EVENT_NAME,
		...prepareAdditionalPayload(pageViewPayload),
		collection_name: pageViewPayload.collectionHandle,
		collection_id: parseInt(require_analytics_utils.parseGid(pageViewPayload.collectionId).id)
	}, formatPayload(pageViewPayload)))];
}
function productView(payload) {
	const pageViewPayload = payload;
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: PRODUCT_PAGE_RENDERED_EVENT_NAME,
		...prepareAdditionalPayload(pageViewPayload),
		products: formatProductPayload(pageViewPayload.products),
		total_value: pageViewPayload.totalValue
	}, formatPayload(pageViewPayload)))];
}
function searchView(payload) {
	const pageViewPayload = payload;
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: SEARCH_SUBMITTED_EVENT_NAME,
		...prepareAdditionalPayload(pageViewPayload),
		search_string: pageViewPayload.searchString
	}, formatPayload(pageViewPayload)))];
}
function addToCart(payload) {
	const addToCartPayload = payload;
	const cartToken = require_analytics_utils.parseGid(addToCartPayload.cartId);
	return [require_analytics_utils.schemaWrapper(SCHEMA_ID, require_analytics_utils.addDataIf({
		event_name: PRODUCT_ADDED_TO_CART_EVENT_NAME,
		customerId: addToCartPayload.customerId,
		cart_token: cartToken?.id ? `${cartToken.id}` : null,
		total_value: addToCartPayload.totalValue,
		products: formatProductPayload(addToCartPayload.products),
		customer_id: parseInt(require_analytics_utils.parseGid(addToCartPayload.customerId).id || "0")
	}, formatPayload(addToCartPayload)))];
}
function formatPayload(payload) {
	const payloadWithPrivacy = payload;
	return {
		source: payload.shopifySalesChannel || require_analytics_constants.ShopifySalesChannel.headless,
		asset_version_id: payload.assetVersionId || "2026.4.2",
		hydrogenSubchannelId: payload.storefrontId || payload.hydrogenSubchannelId || "0",
		is_persistent_cookie: payload.hasUserConsent,
		deprecated_visit_token: payload.visitToken,
		unique_token: payload.uniqueToken,
		event_time: Date.now(),
		event_id: require_cookies_utils.buildUUID(),
		event_source_url: payload.url,
		referrer: payload.referrer,
		user_agent: payload.userAgent,
		navigation_type: payload.navigationType,
		navigation_api: payload.navigationApi,
		shop_id: parseInt(require_analytics_utils.parseGid(payload.shopId).id),
		currency: payload.currency,
		ccpa_enforced: payloadWithPrivacy.ccpaEnforced || false,
		gdpr_enforced: payloadWithPrivacy.gdprEnforced || false,
		gdpr_enforced_as_string: payloadWithPrivacy.gdprEnforced ? "true" : "false",
		analytics_allowed: payload.analyticsAllowed || false,
		marketing_allowed: payload.marketingAllowed || false,
		sale_of_data_allowed: payload.saleOfDataAllowed || false
	};
}
function formatProductPayload(products) {
	return products ? products.map((p) => {
		const product = require_analytics_utils.addDataIf({
			variant_gid: p.variantGid,
			category: p.category,
			sku: p.sku,
			product_id: parseInt(require_analytics_utils.parseGid(p.productGid).id),
			variant_id: parseInt(require_analytics_utils.parseGid(p.variantGid).id)
		}, {
			product_gid: p.productGid,
			name: p.name,
			variant: p.variantName || "",
			brand: p.brand,
			price: parseFloat(p.price),
			quantity: Number(p.quantity || 0)
		});
		return JSON.stringify(product);
	}) : [];
}
//#endregion
exports.addToCart = addToCart;
exports.collectionView = collectionView;
exports.pageView = pageView;
exports.pageView2 = pageView2;
exports.productView = productView;
exports.searchView = searchView;
