// @shopify/hydrogen-react@2026.4.2/dist/node-dev/analytics.mjs
import { getTrackingValues } from "./tracking-utils.mjs";
import { AnalyticsEventName } from "./analytics-constants.mjs";
import { errorIfServer } from "./analytics-utils.mjs";
import { pageView } from "./analytics-schema-trekkie-storefront-page-view.mjs";
import { addToCart, collectionView, pageView as pageView$1, pageView2, productView, searchView } from "./analytics-schema-custom-storefront-customer-tracking.mjs";
//#region src/analytics.ts
/**
* Set user and session cookies and refresh the expiry time. If `event.payload.hasUserConsent` is false, no analytics event will happen.
* @param event - The analytics event.
* @param shopDomain - The Online Store domain to sent Shopify analytics under the same
*   top level domain.
* @publicDocs
*/
function sendShopifyAnalytics(event, shopDomain) {
	const { eventName, payload } = event;
	if (!payload.hasUserConsent) return Promise.resolve();
	let events = [];
	const pageViewPayload = payload;
	if (eventName === AnalyticsEventName.PAGE_VIEW) events = events.concat(pageView(pageViewPayload), pageView$1(pageViewPayload));
	else if (eventName === AnalyticsEventName.ADD_TO_CART) events = events.concat(addToCart(payload));
	else if (eventName === AnalyticsEventName.PAGE_VIEW_2) events = events.concat(pageView(pageViewPayload), pageView2(pageViewPayload));
	else if (eventName === AnalyticsEventName.COLLECTION_VIEW) events = events.concat(collectionView(pageViewPayload));
	else if (eventName === AnalyticsEventName.PRODUCT_VIEW) events = events.concat(productView(pageViewPayload));
	else if (eventName === AnalyticsEventName.SEARCH_VIEW) events = events.concat(searchView(pageViewPayload));
	if (events.length) return sendToShopify(events, shopDomain);
	else return Promise.resolve();
}
function isLighthouseUserAgent() {
	if (typeof window === "undefined" || !window.navigator) return false;
	return /Chrome-Lighthouse/.test(window.navigator.userAgent);
}
var ERROR_MESSAGE = "sendShopifyAnalytics request is unsuccessful";
function sendToShopify(events, shopDomain) {
	if (isLighthouseUserAgent()) return Promise.resolve();
	const eventsToBeSent = {
		events,
		metadata: { event_sent_at_ms: Date.now() }
	};
	try {
		return fetch(shopDomain ? `https://${shopDomain}/.well-known/shopify/monorail/unstable/produce_batch` : "https://monorail-edge.shopifysvc.com/unstable/produce_batch", {
			method: "post",
			headers: { "content-type": "text/plain" },
			body: JSON.stringify(eventsToBeSent)
		}).then((response) => {
			if (!response.ok) throw new Error("Response failed");
			return response.text();
		}).then((data) => {
			if (data) JSON.parse(data).result.forEach((eventResponse) => {
				if (eventResponse.status !== 200) console.error(ERROR_MESSAGE, "\n\n", eventResponse.message);
			});
		}).catch((err) => {
			console.error(ERROR_MESSAGE, err);
			throw new Error(ERROR_MESSAGE);
		});
	} catch (error) {
		return Promise.resolve();
	}
}
/**
* If executed on server, this method will return empty string for each field.
* @publicDocs
*/
function getClientBrowserParameters() {
	if (errorIfServer("getClientBrowserParameters")) return {
		uniqueToken: "",
		visitToken: "",
		url: "",
		path: "",
		search: "",
		referrer: "",
		title: "",
		userAgent: "",
		navigationType: "",
		navigationApi: ""
	};
	const [navigationType, navigationApi] = getNavigationType();
	const trackingValues = getTrackingValues();
	return {
		uniqueToken: trackingValues.uniqueToken,
		visitToken: trackingValues.visitToken,
		url: location.href,
		path: location.pathname,
		search: location.search,
		referrer: document.referrer,
		title: document.title,
		userAgent: navigator.userAgent,
		navigationType,
		navigationApi
	};
}
function getNavigationTypeExperimental() {
	try {
		const navigationEntries = performance?.getEntriesByType && performance?.getEntriesByType("navigation");
		if (navigationEntries && navigationEntries[0]) {
			const rawType = window.performance.getEntriesByType("navigation")[0]["type"];
			return rawType && rawType.toString();
		}
	} catch (err) {}
}
function getNavigationTypeLegacy() {
	try {
		if (PerformanceNavigation && performance?.navigation?.type !== null && performance?.navigation?.type !== void 0) {
			const rawType = performance.navigation.type;
			switch (rawType) {
				case PerformanceNavigation.TYPE_NAVIGATE: return "navigate";
				case PerformanceNavigation.TYPE_RELOAD: return "reload";
				case PerformanceNavigation.TYPE_BACK_FORWARD: return "back_forward";
				default: return `unknown: ${rawType}`;
			}
		}
	} catch (err) {}
}
function getNavigationType() {
	try {
		let navApi = "PerformanceNavigationTiming";
		let navType = getNavigationTypeExperimental();
		if (!navType) {
			navType = getNavigationTypeLegacy();
			navApi = "performance.navigation";
		}
		if (navType) return [navType, navApi];
		else return ["unknown", "unknown"];
	} catch (err) {}
	return ["error", "error"];
}
//#endregion
export { getClientBrowserParameters, sendShopifyAnalytics };
