// @shopify/hydrogen-react@2026.4.2/dist/node-dev/cart-hooks.mjs
import { useShop } from "./ShopifyProvider.mjs";
import { SHOPIFY_STOREFRONT_ID_HEADER, SHOPIFY_STOREFRONT_S_HEADER, SHOPIFY_STOREFRONT_Y_HEADER } from "./cart-constants.mjs";
import { SHOPIFY_UNIQUE_TOKEN_HEADER, SHOPIFY_VISIT_TOKEN_HEADER, getTrackingValues } from "./tracking-utils.mjs";
import { useCallback } from "react";
//#region src/cart-hooks.tsx
function useCartFetch() {
	const { storefrontId, getPublicTokenHeaders, getStorefrontApiUrl, sameDomainForStorefrontApi } = useShop();
	return useCallback(({ query, variables }) => {
		const headers = getPublicTokenHeaders({ contentType: "json" });
		if (storefrontId) headers[SHOPIFY_STOREFRONT_ID_HEADER] = storefrontId;
		if (!sameDomainForStorefrontApi) {
			const { uniqueToken, visitToken } = getTrackingValues();
			if (uniqueToken) {
				headers[SHOPIFY_STOREFRONT_Y_HEADER] = uniqueToken;
				headers[SHOPIFY_UNIQUE_TOKEN_HEADER] = uniqueToken;
			}
			if (visitToken) {
				headers[SHOPIFY_STOREFRONT_S_HEADER] = visitToken;
				headers[SHOPIFY_VISIT_TOKEN_HEADER] = visitToken;
			}
		}
		return fetch(getStorefrontApiUrl(), {
			method: "POST",
			headers,
			body: JSON.stringify({
				query: query.toString(),
				variables
			})
		}).then((res) => res.json()).catch((error) => {
			return {
				data: void 0,
				errors: error?.toString()
			};
		});
	}, [
		getPublicTokenHeaders,
		storefrontId,
		getStorefrontApiUrl,
		sameDomainForStorefrontApi
	]);
}
//#endregion
export { useCartFetch };
