// @shopify/hydrogen-react@2026.4.2/dist/node-dev/CartProvider.js
require("./_virtual/_rolldown/runtime.js");
const require_cart_queries = require("./cart-queries.js");
const require_ShopifyProvider = require("./ShopifyProvider.js");
const require_cart_constants = require("./cart-constants.js");
const require_useCartAPIStateMachine = require("./useCartAPIStateMachine.js");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/CartProvider.tsx
var CartContext = (0, react.createContext)(null);
/**
* The `useCart` hook provides access to the cart object. It must be a descendent of a `CartProvider` component.
* @publicDocs
*/
function useCart() {
	const context = (0, react.useContext)(CartContext);
	if (!context) throw new Error("Expected a Cart Context, but no Cart Context was found");
	return context;
}
/**
* The `CartProvider` component synchronizes the state of the Storefront API Cart and a customer's cart,
* and allows you to more easily manipulate the cart by adding, removing, and updating it.
* It could be placed at the root of your app so that your whole app is able to use the `useCart()` hook anywhere.
*
* There are props that trigger when a call to the Storefront API is made, such as `onLineAdd={}` when a line is added to the cart.
* There are also props that trigger when a call to the Storefront API is completed, such as `onLineAddComplete={}` when the fetch request for adding a line to the cart completes.
*
* The `CartProvider` component must be a descendant of the `ShopifyProvider` component.
* @publicDocs
*/
function CartProvider({ children, numCartLines, onCreate, onLineAdd, onLineRemove, onLineUpdate, onNoteUpdate, onBuyerIdentityUpdate, onAttributesUpdate, onDiscountCodesUpdate, onCreateComplete, onLineAddComplete, onLineRemoveComplete, onLineUpdateComplete, onNoteUpdateComplete, onBuyerIdentityUpdateComplete, onAttributesUpdateComplete, onDiscountCodesUpdateComplete, data: cart, cartFragment = require_cart_queries.defaultCartFragment, customerAccessToken, countryCode, languageCode }) {
	const shop = require_ShopifyProvider.useShop();
	if (!shop) throw new Error("<CartProvider> needs to be a descendant of <ShopifyProvider>");
	countryCode = (countryCode ?? shop.countryIsoCode ?? "US").toUpperCase();
	languageCode = (languageCode ?? shop.languageIsoCode ?? "EN").toUpperCase();
	if (countryCode) countryCode = countryCode.toUpperCase();
	const [prevCountryCode, setPrevCountryCode] = (0, react.useState)(countryCode);
	const [prevCustomerAccessToken, setPrevCustomerAccessToken] = (0, react.useState)(customerAccessToken);
	const customerOverridesCountryCode = (0, react.useRef)(false);
	if (prevCountryCode !== countryCode || prevCustomerAccessToken !== customerAccessToken) {
		setPrevCountryCode(countryCode);
		setPrevCustomerAccessToken(customerAccessToken);
		customerOverridesCountryCode.current = false;
	}
	const [cartState, cartSend] = require_useCartAPIStateMachine.useCartAPIStateMachine({
		numCartLines,
		data: cart,
		cartFragment,
		countryCode,
		languageCode,
		onCartActionEntry(_, event) {
			try {
				switch (event.type) {
					case "CART_CREATE": return onCreate?.();
					case "CARTLINE_ADD": return onLineAdd?.();
					case "CARTLINE_REMOVE": return onLineRemove?.();
					case "CARTLINE_UPDATE": return onLineUpdate?.();
					case "NOTE_UPDATE": return onNoteUpdate?.();
					case "BUYER_IDENTITY_UPDATE": return onBuyerIdentityUpdate?.();
					case "CART_ATTRIBUTES_UPDATE": return onAttributesUpdate?.();
					case "DISCOUNT_CODES_UPDATE": return onDiscountCodesUpdate?.();
				}
			} catch (error) {
				console.error("Cart entry action failed", error);
			}
		},
		onCartActionOptimisticUI(context, event) {
			if (!context.cart) return { ...context };
			switch (event.type) {
				case "CARTLINE_REMOVE": return {
					...context,
					cart: {
						...context.cart,
						lines: context?.cart?.lines?.filter((line) => line?.id && !event.payload.lines.includes(line?.id))
					}
				};
				case "CARTLINE_UPDATE": return {
					...context,
					cart: {
						...context.cart,
						lines: context?.cart?.lines?.map((line) => {
							const updatedLine = event.payload.lines.find(({ id }) => id === line?.id);
							if (updatedLine && updatedLine.quantity) return {
								...line,
								quantity: updatedLine.quantity
							};
							return line;
						})
					}
				};
			}
			return { ...context };
		},
		onCartActionComplete(context, event) {
			const cartActionEvent = event.payload.cartActionEvent;
			try {
				switch (event.type) {
					case "RESOLVE": switch (cartActionEvent.type) {
						case "CART_CREATE": return onCreateComplete?.();
						case "CARTLINE_ADD": return onLineAddComplete?.();
						case "CARTLINE_REMOVE": return onLineRemoveComplete?.();
						case "CARTLINE_UPDATE": return onLineUpdateComplete?.();
						case "NOTE_UPDATE": return onNoteUpdateComplete?.();
						case "BUYER_IDENTITY_UPDATE":
							if (countryCodeNotUpdated(context, cartActionEvent)) customerOverridesCountryCode.current = true;
							return onBuyerIdentityUpdateComplete?.();
						case "CART_ATTRIBUTES_UPDATE": return onAttributesUpdateComplete?.();
						case "DISCOUNT_CODES_UPDATE": return onDiscountCodesUpdateComplete?.();
					}
				}
			} catch (error) {
				console.error("onCartActionComplete failed", error);
			}
		}
	});
	const cartReady = (0, react.useRef)(false);
	const [isCartReady, setIsCartReady] = (0, react.useState)(false);
	const cartCompleted = cartState.matches("cartCompleted");
	const countryChanged = (cartState.value === "idle" || cartState.value === "error" || cartState.value === "cartCompleted") && countryCode !== cartState?.context?.cart?.buyerIdentity?.countryCode && !cartState.context.errors;
	const fetchingFromStorage = (0, react.useRef)(false);
	/**
	* Initializes cart with priority in this order:
	* 1. cart props
	* 2. localStorage cartId
	*/
	(0, react.useEffect)(() => {
		if (!cartReady.current && !fetchingFromStorage.current) {
			if (!cart && storageAvailable("localStorage")) {
				fetchingFromStorage.current = true;
				try {
					const cartId = window.localStorage.getItem(require_cart_constants.CART_ID_STORAGE_KEY);
					if (cartId) cartSend({
						type: "CART_FETCH",
						payload: { cartId }
					});
				} catch (error) {
					console.warn("error fetching cartId");
					console.warn(error);
				}
			}
			cartReady.current = true;
			setIsCartReady(true);
		}
	}, [
		cart,
		cartReady,
		cartSend
	]);
	(0, react.useEffect)(() => {
		if (!countryChanged || customerOverridesCountryCode.current) return;
		cartSend({
			type: "BUYER_IDENTITY_UPDATE",
			payload: { buyerIdentity: {
				countryCode,
				customerAccessToken
			} }
		});
	}, [
		countryCode,
		customerAccessToken,
		countryChanged,
		customerOverridesCountryCode,
		cartSend
	]);
	const onCartReadySend = (0, react.useCallback)((cartEvent) => {
		if (!cartReady.current) return console.warn("Cart isn't ready yet");
		cartSend(cartEvent);
	}, [cartSend]);
	(0, react.useEffect)(() => {
		if (cartState?.context?.cart?.id && storageAvailable("localStorage")) try {
			window.localStorage.setItem(require_cart_constants.CART_ID_STORAGE_KEY, cartState.context.cart?.id);
		} catch (error) {
			console.warn("Failed to save cartId to localStorage", error);
		}
	}, [cartState?.context?.cart?.id]);
	(0, react.useEffect)(() => {
		if (cartCompleted && storageAvailable("localStorage")) try {
			window.localStorage.removeItem(require_cart_constants.CART_ID_STORAGE_KEY);
		} catch (error) {
			console.warn("Failed to delete cartId from localStorage", error);
		}
	}, [cartCompleted]);
	const cartCreate = (0, react.useCallback)((cartInput) => {
		if (countryCode && !cartInput.buyerIdentity?.countryCode) {
			if (cartInput.buyerIdentity == null) cartInput.buyerIdentity = {};
			cartInput.buyerIdentity.countryCode = countryCode;
		}
		if (customerAccessToken && !cartInput.buyerIdentity?.customerAccessToken) {
			if (cartInput.buyerIdentity == null) cartInput.buyerIdentity = {};
			cartInput.buyerIdentity.customerAccessToken = customerAccessToken;
		}
		onCartReadySend({
			type: "CART_CREATE",
			payload: cartInput
		});
	}, [
		countryCode,
		customerAccessToken,
		onCartReadySend
	]);
	const cartDisplayState = useDelayedStateUntilHydration(cartState);
	const cartContextValue = (0, react.useMemo)(() => {
		return {
			...cartDisplayState?.context?.cart ?? {
				lines: [],
				attributes: []
			},
			status: transposeStatus(cartDisplayState.value),
			error: cartDisplayState?.context?.errors,
			totalQuantity: cartDisplayState?.context?.cart?.totalQuantity ?? 0,
			cartCreate,
			cartReady: isCartReady,
			linesAdd(lines) {
				if (cartDisplayState?.context?.cart?.id) onCartReadySend({
					type: "CARTLINE_ADD",
					payload: { lines }
				});
				else cartCreate({ lines });
			},
			linesRemove(lines) {
				onCartReadySend({
					type: "CARTLINE_REMOVE",
					payload: { lines }
				});
			},
			linesUpdate(lines) {
				onCartReadySend({
					type: "CARTLINE_UPDATE",
					payload: { lines }
				});
			},
			noteUpdate(note) {
				onCartReadySend({
					type: "NOTE_UPDATE",
					payload: { note }
				});
			},
			buyerIdentityUpdate(buyerIdentity) {
				onCartReadySend({
					type: "BUYER_IDENTITY_UPDATE",
					payload: { buyerIdentity }
				});
			},
			cartAttributesUpdate(attributes) {
				onCartReadySend({
					type: "CART_ATTRIBUTES_UPDATE",
					payload: { attributes }
				});
			},
			discountCodesUpdate(discountCodes) {
				onCartReadySend({
					type: "DISCOUNT_CODES_UPDATE",
					payload: { discountCodes }
				});
			},
			cartFragment
		};
	}, [
		cartCreate,
		isCartReady,
		cartDisplayState?.context?.cart,
		cartDisplayState?.context?.errors,
		cartDisplayState.value,
		cartFragment,
		onCartReadySend
	]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CartContext.Provider, {
		value: cartContextValue,
		children
	});
}
function transposeStatus(status) {
	switch (status) {
		case "uninitialized":
		case "initializationError": return "uninitialized";
		case "idle":
		case "cartCompleted":
		case "error": return "idle";
		case "cartFetching": return "fetching";
		case "cartCreating": return "creating";
		case "cartLineAdding":
		case "cartLineRemoving":
		case "cartLineUpdating":
		case "noteUpdating":
		case "buyerIdentityUpdating":
		case "cartAttributesUpdating":
		case "discountCodesUpdating": return "updating";
	}
}
/**
* Delays a state update until hydration finishes. Useful for preventing suspense boundaries errors when updating a context
* @remarks this uses startTransition and waits for it to finish.
*/
function useDelayedStateUntilHydration(state) {
	const [isPending, startTransition] = (0, react.useTransition)();
	const [delayedState, setDelayedState] = (0, react.useState)(state);
	const firstTimePending = (0, react.useRef)(false);
	if (isPending) firstTimePending.current = true;
	const firstTimePendingFinished = (0, react.useRef)(false);
	if (!isPending && firstTimePending.current) firstTimePendingFinished.current = true;
	(0, react.useEffect)(() => {
		startTransition(() => {
			if (!firstTimePendingFinished.current) setDelayedState(state);
		});
	}, [state]);
	return firstTimePendingFinished.current ? state : delayedState;
}
/**
* Check for storage availability function obtained from
* @see https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API
*/
function storageAvailable(type) {
	let storage;
	try {
		storage = window[type];
		const x = "__storage_test__";
		storage.setItem(x, x);
		storage.removeItem(x);
		return true;
	} catch (e) {
		return !!(e instanceof DOMException && (e.code === 22 || e.code === 1014 || e.name === "QuotaExceededError" || e.name === "NS_ERROR_DOM_QUOTA_REACHED") && storage && storage.length !== 0);
	}
}
function countryCodeNotUpdated(context, event) {
	return !!(event.payload.buyerIdentity.countryCode && context.cart?.buyerIdentity?.countryCode !== event.payload.buyerIdentity.countryCode);
}
//#endregion
exports.CartProvider = CartProvider;
exports.useCart = useCart;
