// @shopify/hydrogen-react@2026.4.2/dist/node-dev/useCartAPIStateMachine.js
require("./_virtual/_rolldown/runtime.js");
const require_index = require("./node_modules/.pnpm/@xstate_fsm@2.0.0/node_modules/@xstate/fsm/es/index.js");
const require_useMachine = require("./useMachine.js");
const require_flatten_connection = require("./flatten-connection.js");
const require_useCartActions = require("./useCartActions.js");
let react = require("react");
//#region src/useCartAPIStateMachine.tsx
function invokeCart(action, options) {
	return {
		entry: [
			...options?.entryActions || [],
			require_index.i({ lastValidCart: (context) => context?.cart }),
			"onCartActionEntry",
			"onCartActionOptimisticUI",
			action
		],
		on: {
			RESOLVE: {
				target: options?.resolveTarget || "idle",
				actions: [require_index.i({
					prevCart: (context) => context?.lastValidCart,
					cart: (_, event) => event?.payload?.cart,
					rawCartResult: (_, event) => event?.payload?.rawCartResult,
					errors: (_) => void 0
				})]
			},
			ERROR: {
				target: options?.errorTarget || "error",
				actions: [require_index.i({
					prevCart: (context) => context?.lastValidCart,
					cart: (context) => context?.lastValidCart,
					errors: (_, event) => event?.payload?.errors
				})]
			},
			CART_COMPLETED: {
				target: "cartCompleted",
				actions: require_index.i({
					prevCart: (_) => void 0,
					cart: (_) => void 0,
					lastValidCart: (_) => void 0,
					rawCartResult: (_) => void 0,
					errors: (_) => void 0
				})
			}
		},
		exit: ["onCartActionComplete", ...options?.exitActions || []]
	};
}
var INITIALIZING_CART_EVENTS = {
	CART_FETCH: { target: "cartFetching" },
	CART_CREATE: { target: "cartCreating" },
	CART_SET: {
		target: "idle",
		actions: [require_index.i({
			rawCartResult: (_, event) => event.payload.cart,
			cart: (_, event) => cartFromGraphQL(event.payload.cart)
		})]
	}
};
var UPDATING_CART_EVENTS = {
	CARTLINE_ADD: { target: "cartLineAdding" },
	CARTLINE_UPDATE: { target: "cartLineUpdating" },
	CARTLINE_REMOVE: { target: "cartLineRemoving" },
	NOTE_UPDATE: { target: "noteUpdating" },
	BUYER_IDENTITY_UPDATE: { target: "buyerIdentityUpdating" },
	CART_ATTRIBUTES_UPDATE: { target: "cartAttributesUpdating" },
	DISCOUNT_CODES_UPDATE: { target: "discountCodesUpdating" }
};
function createCartMachine(initialCart) {
	return require_index.s({
		id: "Cart",
		initial: initialCart ? "idle" : "uninitialized",
		context: { cart: initialCart && cartFromGraphQL(initialCart) },
		states: {
			uninitialized: { on: INITIALIZING_CART_EVENTS },
			cartCompleted: { on: INITIALIZING_CART_EVENTS },
			initializationError: { on: INITIALIZING_CART_EVENTS },
			idle: { on: {
				...INITIALIZING_CART_EVENTS,
				...UPDATING_CART_EVENTS
			} },
			error: { on: {
				...INITIALIZING_CART_EVENTS,
				...UPDATING_CART_EVENTS
			} },
			cartFetching: invokeCart("cartFetchAction", { errorTarget: "initializationError" }),
			cartCreating: invokeCart("cartCreateAction", { errorTarget: "initializationError" }),
			cartLineRemoving: invokeCart("cartLineRemoveAction"),
			cartLineUpdating: invokeCart("cartLineUpdateAction"),
			cartLineAdding: invokeCart("cartLineAddAction"),
			noteUpdating: invokeCart("noteUpdateAction"),
			buyerIdentityUpdating: invokeCart("buyerIdentityUpdateAction"),
			cartAttributesUpdating: invokeCart("cartAttributesUpdateAction"),
			discountCodesUpdating: invokeCart("discountCodesUpdateAction")
		}
	});
}
function useCartAPIStateMachine({ numCartLines, onCartActionEntry, onCartActionOptimisticUI, onCartActionComplete, data: cart, cartFragment, countryCode, languageCode }) {
	const { cartFetch, cartCreate, cartLineAdd, cartLineUpdate, cartLineRemove, noteUpdate, buyerIdentityUpdate, cartAttributesUpdate, discountCodesUpdate } = require_useCartActions.useCartActions({
		numCartLines,
		cartFragment,
		countryCode,
		languageCode
	});
	const [state, send, service] = require_useMachine.useMachine((0, react.useMemo)(() => createCartMachine(cart), [cart]), { actions: {
		cartFetchAction: async (_, event) => {
			if (event.type !== "CART_FETCH") return;
			const { data, errors } = await cartFetch(event?.payload?.cartId);
			send(eventFromFetchResult(event, data?.cart, errors));
		},
		cartCreateAction: async (_, event) => {
			if (event.type !== "CART_CREATE") return;
			const { data, errors } = await cartCreate(event?.payload);
			send(eventFromFetchResult(event, data?.cartCreate?.cart, errors));
		},
		cartLineAddAction: async (context, event) => {
			if (event.type !== "CARTLINE_ADD" || !context?.cart?.id) return;
			const { data, errors } = await cartLineAdd(context.cart.id, event.payload.lines);
			send(eventFromFetchResult(event, data?.cartLinesAdd?.cart, errors));
		},
		cartLineUpdateAction: async (context, event) => {
			if (event.type !== "CARTLINE_UPDATE" || !context?.cart?.id) return;
			const { data, errors } = await cartLineUpdate(context.cart.id, event.payload.lines);
			send(eventFromFetchResult(event, data?.cartLinesUpdate?.cart, errors));
		},
		cartLineRemoveAction: async (context, event) => {
			if (event.type !== "CARTLINE_REMOVE" || !context?.cart?.id) return;
			const { data, errors } = await cartLineRemove(context.cart.id, event.payload.lines);
			send(eventFromFetchResult(event, data?.cartLinesRemove?.cart, errors));
		},
		noteUpdateAction: async (context, event) => {
			if (event.type !== "NOTE_UPDATE" || !context?.cart?.id) return;
			const { data, errors } = await noteUpdate(context.cart.id, event.payload.note);
			send(eventFromFetchResult(event, data?.cartNoteUpdate?.cart, errors));
		},
		buyerIdentityUpdateAction: async (context, event) => {
			if (event.type !== "BUYER_IDENTITY_UPDATE" || !context?.cart?.id) return;
			const { data, errors } = await buyerIdentityUpdate(context.cart.id, event.payload.buyerIdentity);
			send(eventFromFetchResult(event, data?.cartBuyerIdentityUpdate?.cart, errors));
		},
		cartAttributesUpdateAction: async (context, event) => {
			if (event.type !== "CART_ATTRIBUTES_UPDATE" || !context?.cart?.id) return;
			const { data, errors } = await cartAttributesUpdate(context.cart.id, event.payload.attributes);
			send(eventFromFetchResult(event, data?.cartAttributesUpdate?.cart, errors));
		},
		discountCodesUpdateAction: async (context, event) => {
			if (event.type !== "DISCOUNT_CODES_UPDATE" || !context?.cart?.id) return;
			const { data, errors } = await discountCodesUpdate(context.cart.id, event.payload.discountCodes);
			send(eventFromFetchResult(event, data?.cartDiscountCodesUpdate?.cart, errors));
		},
		...onCartActionEntry && { onCartActionEntry: (context, event) => {
			if (isCartActionEvent(event)) onCartActionEntry(context, event);
		} },
		...onCartActionOptimisticUI && { onCartActionOptimisticUI: require_index.i((context, event) => {
			return onCartActionOptimisticUI(context, event);
		}) },
		...onCartActionComplete && { onCartActionComplete: (context, event) => {
			if (isCartFetchResultEvent(event)) onCartActionComplete(context, event);
		} }
	} });
	return (0, react.useMemo)(() => [
		state,
		send,
		service
	], [
		state,
		send,
		service
	]);
}
function cartFromGraphQL(cart) {
	return {
		...cart,
		lines: require_flatten_connection.flattenConnection(cart?.lines),
		note: cart.note ?? void 0
	};
}
function eventFromFetchResult(cartActionEvent, cart, errors) {
	if (errors) return {
		type: "ERROR",
		payload: {
			errors,
			cartActionEvent
		}
	};
	if (!cart) return {
		type: "CART_COMPLETED",
		payload: { cartActionEvent }
	};
	return {
		type: "RESOLVE",
		payload: {
			cart: cartFromGraphQL(cart),
			rawCartResult: cart,
			cartActionEvent
		}
	};
}
function isCartActionEvent(event) {
	return event.type === "CART_CREATE" || event.type === "CARTLINE_ADD" || event.type === "CARTLINE_UPDATE" || event.type === "CARTLINE_REMOVE" || event.type === "NOTE_UPDATE" || event.type === "BUYER_IDENTITY_UPDATE" || event.type === "CART_ATTRIBUTES_UPDATE" || event.type === "DISCOUNT_CODES_UPDATE";
}
function isCartFetchResultEvent(event) {
	return event.type === "RESOLVE" || event.type === "ERROR" || event.type === "CART_COMPLETED";
}
//#endregion
exports.useCartAPIStateMachine = useCartAPIStateMachine;
